---
title: "HSCR QC and Filtering"
author: "Liz Vincent"
date: "`r format(Sys.time(), '%B %d, %Y')`"
output: pdf_document
---
```{r include=FALSE}
knitr::opts_chunk$set(message = FALSE, warning = FALSE, error = FALSE, fig.align = "center")
source("init.R")
```

# Import Data
### Data from Cuffnorm run on all cells
```{r Import_Data, results='hide'}
# Gene FPKMs
fpkms <- read.delim("all_cells/genes.fpkm_table", row.names = 1, stringsAsFactors = F)
colnames(fpkms) <- unlist(strsplit(as.character(colnames(fpkms)), "_0$")) # Remove "_0" after cell ids

# Isoform FPKMs
isoform_fpkms <- read.delim("all_cells/isoforms.fpkm_table", row.names = 1, stringsAsFactors = F)
colnames(isoform_fpkms) <- unlist(strsplit(as.character(colnames(isoform_fpkms)), "_0$")) # Remove "_0" after cell ids

# Sample Annotation
sample_ann <- read.delim("all_cells/samples.table", stringsAsFactors = F)
sample_ann <- sample_ann[, c(-2, -5)] # Remove file path and external scale (external scale = 1 for all cells)
names(sample_ann) <- c("cell_id", names(sample_ann)[2:3]) # Rename sample_id column as cell_id
sample_ann$cell_id <- unlist(strsplit(as.character(sample_ann$cell_id), "_0$")) # Remove "_0" after cell ids

master_cell_sheet <- read.delim("sample_info.txt", stringsAsFactors = F, row.names = 1)
sample_info <- merge(sample_ann, master_cell_sheet, by = "cell_id")
alignment_rates <- read.delim("reads.txt", stringsAsFactors = F)
alignment_rates$unmapped_reads <- alignment_rates$unmapped_mates/2
sample_info <- merge(sample_info, alignment_rates, by = "cell_id")
rownames(sample_info) <- sample_info[, 1] # Set row names to cell_ids

# Gene Annotation
gene_ann <- read.delim("all_cells/genes.attr_table", row.names = 1, stringsAsFactors = F)

fd <- new("AnnotatedDataFrame", data = gene_ann)
pd <- new("AnnotatedDataFrame", data = sample_info)

# Create cell data set object
dat.relative <- newCellDataSet(cellData = as.matrix(fpkms),
                      phenoData = pd,
                      featureData = fd)
```

```{r}
pca.1 <- prcomp(t(exprs(dat.relative)), scale = F, center = T)

ggbiplot(pca.1, groups = pData(dat.relative)$genotype, var.axes = F, ellipse) +
  scale_color_viridis(discrete = T) +
  scale_x_continuous(limits = c(-0.07, 0)) +
  scale_y_continuous(limits = c(-0.05, 0.07)) +
  theme(aspect.ratio = 1)
```


# Remove Outliers
### Remove cells with low alignment rates
Cells with total reads in the bottom 5% and an alignment rate below 50% were removed (**Figure 1**).
```{r}
n <- floor(nrow(pData(dat.relative)) * 0.05)  # 5% of 1369 cells = 68.45 cells, round down to 68
read_min <- (pData(dat.relative)[order(pData(dat.relative)$total_reads),"total_reads"])[n]  # 5% of cells have < 21810 total reads
```

```{r include = F}
pdf("AlignmentByReads.pdf", height = 4, width = 6, useDingbats = F)
ggplot(pData(dat.relative)[!pData(dat.relative)$num_cells == 2,]) + 
  geom_point(aes(x=log10(total_reads),
                 y=alignment_percentage, 
                 col = as.factor(num_cells)), size = 0.2) +
  geom_vline(xintercept = log10(read_min), lty = "dashed") +
  geom_hline(yintercept = 50, lty = "dashed") +
  guides(color = guide_legend(title = "Well", override.aes = list(size = 2.5))) +
  labs(x = "Log10(Total Number of Reads)", y = "% Reads Aligned") +
  scale_color_manual(values = c("red", "black"), breaks = c("1", "0"), labels = c("Cell", "Empty")) +
  theme(aspect.ratio = 1,
        legend.position = "none",
        axis.title = element_text(size = 12),
        axis.text = element_text(size = 10)
  )
dev.off()

p1 <- pData(dat.relative) %>%
  mutate(group = paste(age, sex, genotype)) %>%
  ggplot(.) + 
  geom_point(aes(x = total_reads, y = alignment_percentage, col = group)) +
  stat_ellipse(aes(x = total_reads, y = alignment_percentage, col = group), type = "norm") +
  labs(x = "Total Number of Reads", y = "% Reads Aligned", color = "Group")

p2 <- ggplot(pData(dat.relative)) + 
  geom_point(aes(x = total_reads, y = alignment_percentage, col = genotype)) +
  stat_ellipse(aes(x = total_reads, y = alignment_percentage, col = genotype), type = "norm") +
  labs(x = "Total Number of Reads", y = "% Reads Aligned", color = "Genotype") +
  scale_color_brewer(palette = "Set1")

p3 <- ggplot(pData(dat.relative)) + 
  geom_point(aes(x = total_reads, y = alignment_percentage, col = sex)) +
  stat_ellipse(aes(x = total_reads, y = alignment_percentage, col = sex), type = "norm") +
  labs(x = "Total Number of Reads", y = "% Reads Aligned", color = "Sex") +
  scale_color_brewer(palette = "Set1")

p4 <- ggplot(pData(dat.relative)) + 
  geom_point(aes(x = total_reads, y = alignment_percentage, col = age)) +
  stat_ellipse(aes(x = total_reads, y = alignment_percentage, col = age), type = "norm") +
  labs(x = "Total Number of Reads", y = "% Reads Aligned", color = "Age") +
  scale_color_brewer(palette = "Set1")

p5 <- ggplot(pData(dat.relative)) + 
  geom_point(aes(x = total_reads, y = alignment_percentage, col = batch)) +
  stat_ellipse(aes(x = total_reads, y = alignment_percentage, col = batch), type = "norm") +
  labs(x = "Total Number of Reads", y = "% Reads Aligned", color = "Batch") +
  scale_color_brewer(palette = "Set1")

p6 <- ggplot(pData(dat.relative)) + 
  geom_point(aes(x = total_reads, y = alignment_percentage, col = split_plate)) +
  stat_ellipse(aes(x = total_reads, y = alignment_percentage, col = split_plate), type = "norm") +
  labs(x = "Total Number of Reads", y = "% Reads Aligned", color = "Rearrayed Plate")

legend <- unique(pData(dat.relative)[, c("source_plate", "age", "sex", "genotype")] %>%
       mutate(group = paste(source_plate, age, sex, genotype)))

p7 <- ggplot(pData(dat.relative)) + 
  geom_point(aes(x = total_reads, y = alignment_percentage, col = source_plate)) +
  stat_ellipse(aes(x = total_reads, y = alignment_percentage, col = source_plate), type = "norm") +
  labs(x = "Total Number of Reads", y = "% Reads Aligned", color = "Embryo") +
  scale_color_discrete(breaks = legend$source_plate,
                       labels = legend$group)


p8 <- ggplot(pData(dat.relative)[!pData(dat.relative)$num_cells == 2,]) + 
  geom_point(aes(x = log10(total_reads), y = alignment_percentage, col = factor(num_cells))) +
  stat_ellipse(aes(x = log10(total_reads), y = alignment_percentage, col = factor(num_cells)), type = "norm") +
  labs(x = "Total Number of Reads", y = "% Reads Aligned", color = "Number of Cells per Well")

p9 <- ggplot(pData(dat.relative)) + 
  geom_point(aes(x = total_reads, y = alignment_percentage, col = factor(sort_date))) +
  stat_ellipse(aes(x = total_reads, y = alignment_percentage, col = factor(sort_date)), type = "norm") +
  labs(x = "Total Number of Reads", y = "% Reads Aligned", color = "Sort Date")

# manually grid plots to force same size regardless of legends/axis text
#grid.draw(
#  rbind(
#    cbind(ggplotGrob(p1),
#          ggplotGrob(p2),
#          ggplotGrob(p3),
#          ggplotGrob(p4),
#          size="last"),
#    cbind(ggplotGrob(p5),
#          ggplotGrob(p6),
#          ggplotGrob(p7),
#          ggplotGrob(p8),
#          size="last"),
#    cbind(ggplotGrob(p9),
#          ggplotGrob(p10),
#          ggplotGrob(p11),
#          ggplotGrob(p12),
#          size="last"),
#    size="last")
#)

pdf("Alignment_distribution_by_condition.pdf", height = 5, width = 8, useDingbats = F)
pData(dat.relative) %>%
  mutate(group = paste(age, sex)) %>%
  ggplot(.) + 
  geom_density(aes(x = alignment_percentage, fill = genotype), alpha = 0.4) +
  facet_wrap(~group) +
  scale_fill_manual(breaks = c("het", "hom"), values = c("gray10", "darkred")) +
  scale_y_continuous(limits = c(0, 0.1), expand = c(0,0), breaks = c(0, 0.05, 0.1)) +
  scale_x_continuous(limits = c(0, 100), expand = c(0,0)) +
  theme_bw() +
  theme(aspect.ratio = 1) +
  ggtitle("Alignment Distributions")
dev.off()


pdf("Read_distribution_by_condition.pdf", height = 3, width = 8, useDingbats = F)
pData(dat.relative) %>%
  mutate(group = paste(age, sex)) %>%
  ggplot(.) + 
  geom_density(aes(x = log10(total_reads), fill = genotype), alpha = 0.4) +
  facet_wrap(~group, nrow = 1) +
  scale_fill_manual(breaks = c("het", "hom"), values = c("gray10", "darkred")) +
  scale_x_continuous(expand = c(0,0)) +
  scale_y_continuous(expand = c(0,0)) +
  theme_bw() +
  theme(aspect.ratio = 1) +
  ggtitle("Total Reads Distributions")
dev.off()

```

```{r}
remove <- pData(dat.relative)[pData(dat.relative)$total_reads < read_min & pData(dat.relative)$alignment_percentage < 50,"cell_id"]

tbl1 <- table(pData(dat.relative[, remove])$age, pData(dat.relative[, remove])$sex, pData(dat.relative[, remove])$genotype)

dat.relative.filtered <- dat.relative[, !(row.names(pData(dat.relative)) %in% remove)]
```

### Iteratively run PCA and manually remove outliers
```{r PCA_filtering}
# PCA on FPKM values
dat.relative.pca <- prcomp(t(exprs(dat.relative.filtered)), scale = F, center = T)
```

```{r figures, echo = F}
p1 <- ggbiplot(dat.relative.pca, var.axes = F, groups = (dat.relative.pca$x[,1] > 5e+6)) + 
  theme(aspect.ratio = 1, legend.position = "none") + 
  scale_color_manual(values = c("black", "red")) +
  ggtitle("1. PC1 vs. PC2") 
```

Remove PC 1 and 2 outliers and rerun PCA
```{r remove_outliers}
remove<-c(remove,names(which(dat.relative.pca$x[,1] > 5e+06)))
dat.relative.filtered<-dat.relative[,!(row.names(pData(dat.relative)) %in% remove)]

dat.relative.filtered.pca<-prcomp(t(exprs(dat.relative.filtered)),scale=F,center=T)
```

Repeat iteratively until there are no obvious outliers
```{r echo = F, fig.width=7.5, fig.height = 10}
p2 <- ggbiplot(dat.relative.filtered.pca, var.axes = F, groups = (dat.relative.filtered.pca$x[,1] > 5e+06 | dat.relative.filtered.pca$x[,2] > 3e+06)) + 
  theme(aspect.ratio = 1, legend.position = "none") +  
  scale_color_manual(values = c("black", "red")) +
  ggtitle("2. PC1 vs. PC2") 

remove <- c(remove,names(which(dat.relative.filtered.pca$x[,1] > 5e+06 | dat.relative.filtered.pca$x[,2] > 3e+06)))
dat.relative.filtered <- dat.relative[, !(row.names(pData(dat.relative)) %in% remove)]

dat.relative.filtered.pca <- prcomp(t(exprs(dat.relative.filtered)), scale = F, center = T)

p3 <- ggbiplot(dat.relative.filtered.pca, var.axes = F, groups = (dat.relative.filtered.pca$x[,1] > 3e+06 | dat.relative.filtered.pca$x[,2] < -2e+06)) + 
  theme(aspect.ratio = 1, legend.position = "none") + 
  scale_color_manual(values = c("black", "red")) +
  ggtitle("3. PC1 vs. PC2")

remove <- c(remove,names(which(dat.relative.filtered.pca$x[,1] > 3e+06 | dat.relative.filtered.pca$x[,2] < -2e+06)))
dat.relative.filtered <- dat.relative[, !(row.names(pData(dat.relative)) %in% remove)]

dat.relative.filtered.pca <- prcomp(t(exprs(dat.relative.filtered)), scale = F, center = T)

p4 <- ggbiplot(dat.relative.filtered.pca, var.axes = F, groups = (dat.relative.filtered.pca$x[,1] < -2e+06 | dat.relative.filtered.pca$x[,2] > 2e+06)) + 
  theme(aspect.ratio = 1, legend.position = "none") + 
  scale_color_manual(values = c("black", "red")) +
  ggtitle("4. PC1 vs. PC2")

remove <- c(remove,names(which(dat.relative.filtered.pca$x[,1] < -2e+06 | dat.relative.filtered.pca$x[,2] > 2e+06)))
dat.relative.filtered <- dat.relative[, !(row.names(pData(dat.relative)) %in% remove)]

dat.relative.filtered.pca <- prcomp(t(exprs(dat.relative.filtered)), scale = F, center = T)

p5 <- ggbiplot(dat.relative.filtered.pca, var.axes = F, groups = (dat.relative.filtered.pca$x[,1] < -1e+06 | dat.relative.filtered.pca$x[,2] > 1e+06)) + 
  theme(aspect.ratio = 1,legend.position = "none") + 
  scale_color_manual(values = c("black", "red")) +
  ggtitle("5. PC1 vs. PC2")

remove <- c(remove, names(which(dat.relative.filtered.pca$x[,1] < -1e+06 | dat.relative.filtered.pca$x[,2] > 1e+06)))
dat.relative.filtered <- dat.relative[, !(row.names(pData(dat.relative)) %in% remove)]

dat.relative.filtered.pca <- prcomp(t(exprs(dat.relative.filtered)), scale = F, center = T)

p6 <- ggbiplot(dat.relative.filtered.pca, var.axes = F, groups = (dat.relative.filtered.pca$x[,1] < -1e+06 | dat.relative.filtered.pca$x[,2] > 1e+06)) + 
  theme(aspect.ratio = 1, legend.position = "none") + 
  scale_color_manual(values = c("black", "red")) +
  ggtitle("6. PC1 vs. PC2")

remove <- c(remove,names(which(dat.relative.filtered.pca$x[,1] < -1e+06 | dat.relative.filtered.pca$x[,2] > 1e+06)))
dat.relative.filtered <- dat.relative[, !(row.names(pData(dat.relative)) %in% remove)]

dat.relative.filtered.pca <- prcomp(t(exprs(dat.relative.filtered)), scale = F, center = T)

p7 <- ggbiplot(dat.relative.filtered.pca, var.axes = F, groups = (dat.relative.filtered.pca$x[,1] < -1e+06 | dat.relative.filtered.pca$x[,2] > 5e+05)) +
  theme(aspect.ratio = 1, legend.position = "none") + 
  scale_color_manual(values = c("black", "red")) +
  ggtitle("7. PC1 vs. PC2")

remove <- c(remove, names(which(dat.relative.filtered.pca$x[,1] < -1e+06 | dat.relative.filtered.pca$x[,2] > 5e+05)))
dat.relative.filtered <- dat.relative[, !(row.names(pData(dat.relative)) %in% remove)]

dat.relative.filtered.pca <- prcomp(t(exprs(dat.relative.filtered)), scale = F, center = T)

p8 <- ggbiplot(dat.relative.filtered.pca, var.axes = F) +
  theme(aspect.ratio = 1) + 
  ggtitle("8. PC1 vs. PC2")

write.table(x = remove, "outliers.txt", quote = F, sep = "\n", row.names = F, col.names = F)

grid.arrange(p1, p2, p3, p4, p5, p6, p7, p8, ncol = 4)

pg1 <- ggplot(pData(dat.relative)) +
  geom_point(aes(x = total_reads, y = alignment_percentage, col = ifelse(pData(dat.relative)$cell_id %in% remove, "Removed", "Kept"))) +
  labs(x = "Total Number of Reads", y = "% Reads Aligned", color = "Status") + 
  scale_color_manual(values = c("black","red")) +
  theme(legend.position = c(0.85, 0.94))
     

pg2 <- pData(dat.relative) %>%
  mutate(group = ifelse(pData(dat.relative)$cell_id %in% remove, "Removed", "Kept")) %>%
  ggplot(.) + 
  stat_density(aes(x = total_reads), fill = "gray70", color = "gray60", alpha = 0.3) +
  facet_wrap(~group) +
  scale_fill_brewer(palette = "Set1") +
  scale_color_brewer(palette = "Set1") +
  theme_bw() +
  xlab("Total Reads")

pg3 <- pData(dat.relative) %>%
  mutate(group = ifelse(pData(dat.relative)$cell_id %in% remove, "Removed", "Kept")) %>%
  ggplot(.) + 
  stat_density(aes(x = alignment_percentage), fill = "gray70", color = "gray60", alpha = 0.3) +
  facet_wrap(~group) +
  scale_fill_brewer(palette = "Set1") +
  scale_color_brewer(palette = "Set1") +
  theme_bw() +
  xlab("% Aligned Reads")

pg4 <- pData(dat.relative) %>%
  filter(cell_id %in% remove) %>%
  mutate(group = paste(age, sex, genotype)) %>%
  ggplot(., aes(x = group, fill = sex, alpha=as.numeric(factor(genotype)))) +
    geom_bar() +
    scale_alpha("Genotype", range = c(0.4, 1), breaks = c(2, 1), labels = c("hom","het")) +
    scale_fill_brewer("Sex", palette = "Set1") +
    scale_y_continuous(limits = c(NA, 27), expand = c(0, 0)) +
    theme(axis.ticks.x = element_blank(), axis.title.x = element_blank(), legend.position = c(0.9, 0.9)) +
    geom_text(stat = "count", aes(label = ..count..), vjust = -0.5, alpha = 1)
#  labs(x="Genotype and Sex",y="# Cells",title="Cells Remaining")

grid1 <- plot_grid(pg2, pg3, ncol = 2)
grid2 <- plot_grid(pg1, pg4, ncol = 2)
pg <- plot_grid(grid2, grid1, nrow = 2, rel_heights = c(2, 1))
save_plot("Figure 3.pdf", pg, base_height = 16, base_width = 22)
```

# Rerun Cuffnorm on remaining 1296 cells and import data
```{r Post_rerunning_Cuffnorm}
# Gene FPKMs
fpkms <- read.delim("genes.fpkm_table", row.names = 1, stringsAsFactors = F)
colnames(fpkms) <- unlist(strsplit(as.character(colnames(fpkms)), "_0$"))

# Isoform FPKMs
isoform_fpkms <- read.delim("isoforms.fpkm_table", row.names = 1, stringsAsFactors = F)
colnames(isoform_fpkms) <- unlist(strsplit(as.character(colnames(isoform_fpkms)), "_0$"))

# Sample Annotation
sample_ann <- read.delim("samples.table", stringsAsFactors = F)
sample_ann <- sample_ann[, c(-2, -5)] # Remove file path and external scale (external scale = 1 for all cells)
names(sample_ann) <- c("cell_id", names(sample_ann)[2:3]) # Rename sample_id column as cell_id
sample_ann$cell_id <- unlist(strsplit(as.character(sample_ann$cell_id), "_0$")) # Remove "_0" after cell ids
sample_info <- merge(sample_ann, master_cell_sheet, by = 'cell_id')
sample_info <- merge(sample_info, alignment_rates, by = "cell_id")
rownames(sample_info) <- sample_info[, 1]

# Gene Annotation
gene_ann <- read.delim("genes.attr_table", row.names = 1, stringsAsFactors = F)

fd <- new("AnnotatedDataFrame", data = gene_ann)
pd <- new("AnnotatedDataFrame", data = sample_info)

# Create cell data set object
dat.relative.1296 <- newCellDataSet(cellData = as.matrix(fpkms),
                      phenoData = pd,
                      featureData = fd)

fData(dat.relative.1296)$num_cells_expressed <- rowSums(exprs(dat.relative.1296) > 0)
pData(dat.relative.1296)$num_genes_expressed <- colSums(exprs(dat.relative.1296) > 0)
```

```{r}
pData(dat.relative.1296) %>%
  mutate(group=paste(age, sex, genotype)) %>%
  ggplot(.) +
  geom_density(aes(x = num_genes_expressed, color = group), alpha = 0.2) +
  geom_vline(xintercept = 1000, lty = "dashed") +
  theme(legend.position = c(0.75, 0.85))

# Remove cells expressing very few genes
dat.relative.filtered <- dat.relative.1296[, pData(dat.relative.1296)$num_genes_expressed > 1000]
```

### Run PCA on second round of cuffnorm data
```{r}
dat.relative.filtered.pca <- prcomp(t(exprs(dat.relative.filtered)), scale = F, center = T)
```

```{r echo = F}
p1 <- ggbiplot(dat.relative.filtered.pca, var.axes = F, groups = (dat.relative.filtered.pca$x[,1] > 5e+05 | dat.relative.filtered.pca$x[,2] > 5e+05)) + 
  theme(aspect.ratio = 1, legend.position = "none") +  
  scale_color_manual(values = c("black", "red")) +
  ggtitle("1. PC1 vs. PC2") 

remove <- names(which(dat.relative.filtered.pca$x[,1] > 5e+05 | dat.relative.filtered.pca$x[,2] > 5e+05))
dat.relative.filtered <- dat.relative.filtered[, !(row.names(pData(dat.relative.filtered)) %in% remove)]

dat.relative.filtered.pca <- prcomp(t(exprs(dat.relative.filtered)), scale = F, center = T)

p2 <- ggbiplot(dat.relative.filtered.pca, var.axes = F, groups = (dat.relative.filtered.pca$x[,1] > 5e+05 | dat.relative.filtered.pca$x[,2] > 5e+05)) + 
  theme(aspect.ratio = 1, legend.position = "none") +  
  scale_color_manual(values = c("black", "red")) +
  ggtitle("2. PC1 vs. PC2") 

remove <- c(remove, names(which(dat.relative.filtered.pca$x[,1] > 5e+05 | dat.relative.filtered.pca$x[,2] > 5e+05)))
dat.relative.filtered <- dat.relative.filtered[, !(row.names(pData(dat.relative.filtered)) %in% remove)]

dat.relative.filtered.pca <- prcomp(t(exprs(dat.relative.filtered)), scale = F, center = T)

p3 <- ggbiplot(dat.relative.filtered.pca, var.axes = F, groups = (dat.relative.filtered.pca$x[,1] > 5e+05 | dat.relative.filtered.pca$x[,2] > 5e+05)) + 
  theme(aspect.ratio = 1, legend.position = "none") +  
  scale_color_manual(values = c("black", "red")) +
  ggtitle("3. PC1 vs. PC2") 

grid.arrange(p1, p2, p3, nrow = 1)
```

### Convert FPKM to CPC
```{r run_relative2abs}
isoform_t_estimate <- estimate_t(isoform_fpkms)

fpkm_matrix_adj <- relative2abs(dat.relative.filtered, cores = detectCores() - 1, t_estimate = isoform_t_estimate)

# Create new cell data set with CPC values
dat.filtered <- newCellDataSet(as.matrix(fpkm_matrix_adj),
                      phenoData = pd[rownames(pd) %in% colnames(fpkm_matrix_adj)],
                      featureData = fd,
                      expressionFamily = negbinomial.size(),
                      lowerDetectionLimit = 1)
```

### Add and format metadata
```{r}
pData(dat.filtered)$num_genes_expressed <- colSums(exprs(dat.filtered) > 0)
pData(dat.filtered)$total_mRNAs <- colSums(exprs(dat.filtered))
pData(dat.filtered)$mean_expr <- esApply(dat.filtered, 2, function(x){mean(x)})
pData(dat.filtered)$sd_expr <- esApply(dat.filtered, 2, function(x){sd(x)})
pData(dat.filtered)$genotype <- factor(pData(dat.filtered)$genotype)
pData(dat.filtered)$sex <- factor(pData(dat.filtered)$sex)
pData(dat.filtered)$batch <- factor(pData(dat.filtered)$batch)
pData(dat.filtered)$source_plate <- factor(pData(dat.filtered)$source_plate)
pData(dat.filtered)$split_plate <- factor(pData(dat.filtered)$split_plate)
pData(dat.filtered)$RT_date <- factor(pData(dat.filtered)$RT_date)

fData(dat.filtered)$num_cells_expressed <- rowSums(exprs(dat.filtered) > 0)
fData(dat.filtered)$gene_id <- rownames(fData(dat.filtered))
fData(dat.filtered)$mean_expr <- esApply(dat.filtered, 1, function(x){mean(x)})
fData(dat.filtered)$sd_expr <- esApply(dat.filtered, 1, function(x){sd(x)})
fData(dat.filtered)$bcv <- (fData(dat.filtered)$sd_expr/fData(dat.filtered)$mean_expr)**2
fData(dat.filtered)$percent_detection <- (fData(dat.filtered)$num_cells_expressed/dim(dat.filtered)[2])*100

dat.filtered@dim_reduce_type <- "DDRTree"
dat.filtered@auxOrderingData <- new.env()
```

# Determine Expressed Genes
A gene is considered expressed if it is detected at non-zero CPC in at least 20 cells (approx. 10% of cells in one condition), and the standard deviation is not zero.
```{r Determine_expressed_genes }
expressed_genes <- rownames(fData(dat.filtered)[fData(dat.filtered)$num_cells_expressed >= 20 & fData(dat.filtered)$sd_expr > 1e-20,])
# Set sd lower limit effectively to 0 - this excludes 3 more genes than setting it exactly to zero, but these genes have sd < 2e-35, which prcomp considers to be constant/zero variance
#14,319 genes

fData(dat.filtered)$expressed <- FALSE
fData(dat.filtered)[expressed_genes, "expressed"] <- TRUE


ggplot(fData(dat.filtered)) +
  geom_density(aes(x = num_cells_expressed)) +
  geom_vline(xintercept = 20, lty = "dashed", color = "red") +
  scale_y_continuous(expand = c(0,0)) +
  scale_x_continuous(expand = c(0,0)) +
  ggtitle("Density of genes expressed in a given numer of cells")
```

```{r remove_low_lib_quality}
dat_withmito <- dat.filtered

mito_genes <- fData(dat_withmito)[str_detect(fData(dat_withmito)$gene_short_name, pattern = "mt-"), "gene_short_name"]
mito_genes <- lookupGeneId(dat_withmito, mito_genes)

mito <- exprs(dat_withmito[mito_genes,])
notmito <- exprs(dat_withmito[expressed_genes[!(expressed_genes %in% mito_genes)],])

mt_ratio <- apply(exprs(dat_withmito), 2, function(x){
  sum(x[mito_genes])/sum(x[!(expressed_genes %in% mito_genes)])
})

cutoff <- mean(mt_ratio) + 2*sd(mt_ratio)
length(mt_ratio[mt_ratio < cutoff]) #1003 cells
keep <- names(mt_ratio[mt_ratio < cutoff])

dat.filtered <- dat_withmito[,keep] # removes 28 cells

ggplot(as.data.frame(mt_ratio)) +
  geom_density(aes(x = mt_ratio)) +
  geom_vline(xintercept = cutoff, lty = "dashed", color = "red") +
  scale_y_continuous(expand = c(0,0)) +
  scale_x_continuous(expand = c(0,0)) +
  ggtitle("Distribution of the ratio of mt to genomic CPCs")
```

# PCA on Cleaned Data (Expressed Genes)
```{r}
dat.filtered.pca <- prcomp(t(exprs(dat.filtered[expressed_genes,])), scale = T, center = T)
```

```{r echo = F,fig.height=6.5,fig.width=7.5}
screeplot(dat.filtered.pca, npcs = 20, type = "lines")

ggbiplot(dat.filtered.pca, var.axes = F, groups = pData(dat.filtered)$genotype, alpha = 0) + 
  theme(aspect.ratio = 1,plot.title = element_text(size = 12), axis.title = element_text(size = 8), legend.title = element_text(size = 8), legend.text = element_text(size = 7)) + 
  ggtitle("PC1 vs. PC2") + 
  scale_color_brewer("Genotype", palette = "Set1") +  
  geom_point(aes(color = pData(dat.filtered)$genotype), size = 0.6)

p1 <- ggbiplot(dat.filtered.pca, 
               var.axes = F, 
               groups = log10(pData(dat.filtered)$total_reads)) + 
  scale_color_viridis("# reads") +
  ggtitle("PC1 vs. PC2") + 
  theme(aspect.ratio = 1,
        plot.title = element_text(size = 12), 
        axis.title = element_text(size = 8), 
        legend.title = element_text(size = 8), 
        legend.text = element_text(size = 7))

pdf("PC1vsPC2-batch.pdf", height = 4, width = 5, useDingbats = F)
ggbiplot(dat.filtered.pca, var.axes = F, groups = pData(dat.filtered)$batch, alpha = 0, ellipse = T) +
  ggtitle("PC1 vs. PC2") + 
  scale_color_viridis("Batch", discrete = T, option = "B") +  
  geom_point(aes(color = pData(dat.filtered)$batch), size = 0.6) +
  theme(aspect.ratio = 1, 
        legend.position = c(0.7, 0.2),
        plot.title = element_text(size = 12), 
        axis.title = element_text(size = 10), 
        axis.text = element_text(size = 10),
        legend.title = element_text(size = 10), 
        legend.text = element_text(size = 10))
dev.off()

p3 <- ggbiplot(dat.filtered.pca, var.axes = F, groups = pData(dat.filtered)$sex, alpha = 0) + 
  theme(aspect.ratio = 1, plot.title = element_text(size = 12), axis.title = element_text(size = 8), legend.title = element_text(size = 8), legend.text = element_text(size = 7)) + 
  ggtitle("PC1 vs. PC2") + 
  scale_color_brewer("Sex", palette = "Set1") +  
  geom_point(aes(color = pData(dat.filtered)$sex), size = 0.6)

p4 <- ggbiplot(dat.filtered.pca, var.axes = F, groups = pData(dat.filtered)$total_mRNAs, alpha = 0) + 
  theme(aspect.ratio = 1, plot.title = element_text(size = 12), axis.title = element_text(size =8), legend.title = element_text(size = 8), legend.text = element_text(size = 7)) + 
  ggtitle("PC1 vs. PC2") + 
  scale_color_continuous(name = "Total\nmRNAs") +  
  geom_point(aes(color = pData(dat.filtered)$total_mRNAs), size = 0.6)

grid.draw(rbind(cbind(ggplotGrob(p1), ggplotGrob(p2), size = "last"), cbind(ggplotGrob(p3), ggplotGrob(p4), size = "last"), size = "last"))
```

# Plots of Cleaned Cell Data Set
```{r cleaned_figures, echo = F, fig.height=5,fig.width=5}
ggplot(pData(dat.filtered), aes(x = genotype)) + 
  geom_bar(aes(fill = batch)) + 
  facet_wrap(~age + sex, nrow = 1) + 
  scale_fill_brewer("Batch", palette="Set1") +
  geom_text(stat = "count", aes(label = ..count..), vjust = -0.5) +
  labs(x = "Genotype and Sex", y = "# Cells", title = "Cells Passing Filters")  +
  theme(plot.title = element_text(size = 12), axis.title = element_text(size = 10), legend.text = element_text(size = 8)) +
  scale_y_continuous(expand = expand_scale(mult = c(0, .05)))
```

```{r}
pdf("Read_distribution_by_condition_filtered.pdf", height = 3, width = 8, useDingbats = F)
pData(dat.filtered) %>%
  mutate(group = paste(age, sex)) %>%
  ggplot(.) + 
  geom_density(aes(x = log10(total_reads), fill = genotype), alpha = 0.4) +
  facet_wrap(~group, nrow = 1) +
  scale_fill_manual(breaks = c("het", "hom"), values = c("gray10", "darkred")) +
  scale_x_continuous(expand = c(0,0)) +
  scale_y_continuous(expand = c(0,0)) +
  theme_bw() +
  theme(aspect.ratio = 1) +
  ggtitle("Total Reads Distributions")
dev.off()

```

```{r saveCDS, echo = F}
saveRDS(dat.filtered, file = "datfiltered_postQC.rds")
saveRDS(expressed_genes, file = "expressed_genes.rds")
```

# Session Info
```{r echo = FALSE}
sessionInfo()
```