---
title: "HSCR tSNE"
author: "Liz Vincent"
date: "`r format(Sys.time(), '%B %d, %Y')`"
output: pdf_document
---
```{r init, include = F}
knitr::opts_chunk$set(message = FALSE, warning = FALSE, error = FALSE)
#setwd("/Users/liz/Documents/Goff Lab/Chakravarti/Chakravarti_HD")

source('init.R')

dat.filtered <- readRDS("datfiltered_postQC.rds")
expressed_genes <- readRDS("expressed_genes.rds")
```

## Determine Overdispersed Genes and Run PCA on Overdispersed Genes
Subset the uncorrected Cell Data Set by batch to determine overdispersed genes within each batch and take the insection of genes overdispersed in both batches.
```{r subset_by_batch, include = F}
dat.batch1 <- dat.filtered[, pData(dat.filtered)$batch == "HD01"]
dat.batch1 <- monocle:::estimateSizeFactorsForCellDataSet(dat.batch1)
dat.batch1 <- monocle:::estimateDispersionsForCellDataSet(dat.batch1, modelFormulaStr = "~1", relative_expr = TRUE, removeOutliers = TRUE,
  min_cells_detected = 1)
disp_table.batch1 <- dispersionTable(dat.batch1)
disp_table.batch1$gene_id<-as.character(disp_table.batch1$gene_id)
unsup_clustering_genes.batch1 <- disp_table.batch1[disp_table.batch1$dispersion_empirical >= 1.5*(disp_table.batch1$dispersion_fit),'gene_id']
dat.batch1 <- setOrderingFilter(dat.batch1, unsup_clustering_genes.batch1)
p.1<-plot_ordering_genes(dat.batch1) + ggtitle("Genes Overdispersed in Batch 1")

dat.batch2<-dat.filtered[,pData(dat.filtered)$batch == "HD02"]
dat.batch2<-estimateSizeFactors(dat.batch2)
dat.batch2<-estimateDispersions(dat.batch2,cores=detectCores()-1)
disp_table.batch2 <- dispersionTable(dat.batch2)
disp_table.batch2$gene_id<-as.character(disp_table.batch2$gene_id)
unsup_clustering_genes.batch2 <- disp_table.batch2[disp_table.batch2$dispersion_empirical >= 1.5*(disp_table.batch2$dispersion_fit),'gene_id']
dat.batch2 <- setOrderingFilter(dat.batch2, unsup_clustering_genes.batch2)
p.2<-plot_ordering_genes(dat.batch2) + ggtitle("Genes Overdispersed in Batch 2")

dat.batch3<-dat.filtered[,pData(dat.filtered)$batch == "HD03_pool_1"]
dat.batch3<-estimateSizeFactors(dat.batch3)
dat.batch3<-estimateDispersions(dat.batch3,cores=detectCores()-1)
disp_table.batch3 <- dispersionTable(dat.batch3)
disp_table.batch3$gene_id<-as.character(disp_table.batch3$gene_id)
unsup_clustering_genes.batch3 <- disp_table.batch3[disp_table.batch3$dispersion_empirical >= 1.5*(disp_table.batch3$dispersion_fit),'gene_id']
dat.batch3 <- setOrderingFilter(dat.batch3, unsup_clustering_genes.batch3)
p.3<-plot_ordering_genes(dat.batch3) + ggtitle("Genes Overdispersed in Batch 3")

dat.batch4<-dat.filtered[,pData(dat.filtered)$batch == "HD03_pool_2"]
dat.batch4<-estimateSizeFactors(dat.batch4)
dat.batch4<-estimateDispersions(dat.batch4,cores=detectCores()-1)
disp_table.batch4 <- dispersionTable(dat.batch4)
disp_table.batch4$gene_id<-as.character(disp_table.batch4$gene_id)
unsup_clustering_genes.batch4 <- disp_table.batch4[disp_table.batch4$dispersion_empirical >= 1.5*(disp_table.batch4$dispersion_fit),'gene_id']
dat.batch4 <- setOrderingFilter(dat.batch4, unsup_clustering_genes.batch4)
p.4<-plot_ordering_genes(dat.batch4) + ggtitle("Genes Overdispersed in Batch 4")

dat.batch5<-dat.filtered[,pData(dat.filtered)$batch == "HD03_pool_3"]
dat.batch5<-estimateSizeFactors(dat.batch5)
dat.batch5<-estimateDispersions(dat.batch5,cores=detectCores()-1)
disp_table.batch5 <- dispersionTable(dat.batch5)
disp_table.batch5$gene_id<-as.character(disp_table.batch5$gene_id)
unsup_clustering_genes.batch5 <- disp_table.batch5[disp_table.batch5$dispersion_empirical >= 1.5*(disp_table.batch5$dispersion_fit),'gene_id']
dat.batch5 <- setOrderingFilter(dat.batch5, unsup_clustering_genes.batch5)
p.5<-plot_ordering_genes(dat.batch5) + ggtitle("Genes Overdispersed in Batch 5")

grid.arrange(p.1,p.2,p.3,p.4,p.5,nrow=2)

#calculate overall dispersions
dat.filtered<-estimateSizeFactors(dat.filtered)
dat.filtered<-estimateDispersions(dat.filtered,cores=detectCores()-1)
disp_table <- dispersionTable(dat.filtered)
disp_table$gene_id <- as.character(disp_table$gene_id)

#limit to genes overdispersed in both batches and in the expressed gene set calculated earlier
tmp1 <- intersect(unsup_clustering_genes.batch1, unsup_clustering_genes.batch2)
tmp2 <- intersect(tmp1, unsup_clustering_genes.batch3)
tmp3 <- intersect(tmp2, unsup_clustering_genes.batch4)
unsup_clustering_genes_intersect <- intersect(tmp3, unsup_clustering_genes.batch5)
#241

unsup_clustering_expressed_genes<-intersect(unsup_clustering_genes_intersect, expressed_genes)
# 241 genes

dat.filtered <- setOrderingFilter(dat.filtered, unsup_clustering_genes_intersect)
p.intersect<-plot_ordering_genes(dat.filtered) + ggtitle("Genes Overdispersed in Both Batches")
dat.filtered <- setOrderingFilter(dat.filtered, unsup_clustering_expressed_genes)
p.expressed<-plot_ordering_genes(dat.filtered) + ggtitle("Genes Overdispersed in Both Batches\nExpressed Genes")

grid.arrange(p.1,p.2,p.3,p.4,p.5,p.intersect,p.expressed,ncol=3)
```

```{r pca_overdispersion}
dat.filtered.pca<-
  prcomp(t(log10(exprs(dat.filtered)[unsup_clustering_expressed_genes,] + 1)),
         scale=T,
         center=T)
```

```{r pca_overdisperdsion_plots,echo=F,fig.width=12, fig.height=10}
screeplot(dat.filtered.pca, npcs = 20, type = "lines", main = "PCA on high-variance expressed genes")

ggbiplot(dat.filtered.pca,var.axes=F,groups=pData(dat.filtered)$batch,ellipse=T) + 
  scale_color_brewer("Batch", palette="Set1") + 
  theme(aspect.ratio=1) +
  ggtitle("PCA on high-variance expressed genes")
```

# Umap
```{r Umap}
embedding <- umap(t(log10(exprs(dat.filtered[expressed_genes,]) + 1)),n_neighbors = 15L)

pData(dat.filtered)$UMAP1 <- embedding$UMAP1
pData(dat.filtered)$UMAP2 <- embedding$UMAP2

ggplot(pData(dat.filtered), aes(x = UMAP1, y = UMAP2, color = batch)) +
  geom_point(size = 1) +
  scale_color_brewer("Batch", palette = "Set1") +
  theme_classic() +
  ggtitle("Umap on 14,319 expressed genes")

embedding <- umap(t(log10(exprs(dat.filtered[unsup_clustering_expressed_genes,]) + 1)), n_neighbors = 20L, random_state = 100L)

pData(dat.filtered)$UMAP1 <- embedding$UMAP1
pData(dat.filtered)$UMAP2 <- embedding$UMAP2

ggplot(pData(dat.filtered), aes(x = UMAP1, y = UMAP2, color = batch)) +
  geom_point(size = 1) +
  scale_color_brewer(palette = "Set1") +
  theme_classic() +
  ggtitle("Umap on 292 high-variance expressed genes")
```

# Normalization by sequencing depth
```{r normalize_by_read_depth}
# sum across expressed genes for each cell and divide all gene values by sum to normalize by read depth
dat.filtered.expressed <- dat.filtered[expressed_genes,]
dat.filtered.expressed.normalized <- dat.filtered.expressed
exprs(dat.filtered.expressed.normalized) <- apply(exprs(dat.filtered.expressed), 2, function(x){x/sum(x)})*1e+5
#exprs(dat.filtered.expressed.normalized) <- apply(exprs(dat.filtered.expressed), 2, function(x){x/sum(x)})

fData(dat.filtered.expressed.normalized)$sd_expr<-esApply(dat.filtered.expressed.normalized,1,function(x){sd(x)})
#fData(dat.filtered.expressed.normalized)$mean_expr<-esApply(dat.filtered.expressed.normalized,1,function(x){mean(x)})
```

```{r PCA_on_normalized_data}
normalized.pca <- prcomp(t(log10(exprs(dat.filtered.expressed.normalized[fData(dat.filtered.expressed.normalized)$sd_expr > .Machine$double.eps,]) + 1)), center =T, scale = T)

ggbiplot(normalized.pca,var.axes=F,groups=pData(dat.filtered.expressed.normalized)$batch,alpha=0, ellipse = T) + 
  theme(aspect.ratio = 1,plot.title = element_text(size=12), axis.title = element_text(size=8), legend.title = element_text(size=8), legend.text=element_text(size=7)) + 
  scale_color_brewer("Batch",palette="Set1") +  
  geom_point(aes(color=pData(dat.filtered.expressed.normalized)$batch), size = 0.6)
```

```{r subset_by_batch_normalized,include=F}
dat.batch1<-dat.filtered.expressed.normalized[,pData(dat.filtered.expressed.normalized)$batch == "HD01"]
#dat.batch1<-estimateSizeFactors(dat.batch1)
dat.batch1 <- monocle:::estimateDispersionsForCellDataSet(dat.batch1, modelFormulaStr = "~1", relative_expr = TRUE, removeOutliers = TRUE,
  min_cells_detected = 1)
disp_table.batch1 <- dispersionTable(dat.batch1)
disp_table.batch1$gene_id<-as.character(disp_table.batch1$gene_id)
unsup_clustering_genes.batch1 <- disp_table.batch1[disp_table.batch1$dispersion_empirical >= 1.4*(disp_table.batch1$dispersion_fit),'gene_id']
dat.batch1 <- setOrderingFilter(dat.batch1, unsup_clustering_genes.batch1)
p.1<-plot_ordering_genes(dat.batch1) + ggtitle("Genes Overdispersed in Batch 1") #2309

dat.batch2<-dat.filtered.expressed.normalized[,pData(dat.filtered.expressed.normalized)$batch == "HD02"]
#dat.batch2<-estimateSizeFactors(dat.batch2)
#dat.batch2<-estimateDispersions(dat.batch2,cores=detectCores()-1)
disp_table.batch2 <- dispersionTable(dat.batch2)
disp_table.batch2$gene_id<-as.character(disp_table.batch2$gene_id)
unsup_clustering_genes.batch2 <- disp_table.batch2[disp_table.batch2$dispersion_empirical >= 1.4*(disp_table.batch2$dispersion_fit),'gene_id']
dat.batch2 <- setOrderingFilter(dat.batch2, unsup_clustering_genes.batch2)
p.2<-plot_ordering_genes(dat.batch2) + ggtitle("Genes Overdispersed in Batch 2")

dat.batch3<-dat.filtered.expressed.normalized[,pData(dat.filtered.expressed.normalized)$batch == "HD03_pool_1"]
#dat.batch3<-estimateSizeFactors(dat.batch3)
#dat.batch3<-estimateDispersions(dat.batch3,cores=detectCores()-1)
disp_table.batch3 <- dispersionTable(dat.batch3)
disp_table.batch3$gene_id<-as.character(disp_table.batch3$gene_id)
unsup_clustering_genes.batch3 <- disp_table.batch3[disp_table.batch3$dispersion_empirical >= 1.4*(disp_table.batch3$dispersion_fit),'gene_id']
dat.batch3 <- setOrderingFilter(dat.batch3, unsup_clustering_genes.batch3)
p.3<-plot_ordering_genes(dat.batch3) + ggtitle("Genes Overdispersed in Batch 3")

dat.batch4<-dat.filtered.expressed.normalized[,pData(dat.filtered.expressed.normalized)$batch == "HD03_pool_2"]
#dat.batch4<-estimateSizeFactors(dat.batch4)
#dat.batch4<-estimateDispersions(dat.batch4,cores=detectCores()-1)
disp_table.batch4 <- dispersionTable(dat.batch4)
disp_table.batch4$gene_id<-as.character(disp_table.batch4$gene_id)
unsup_clustering_genes.batch4 <- disp_table.batch4[disp_table.batch4$dispersion_empirical >= 1.4*(disp_table.batch4$dispersion_fit),'gene_id']
dat.batch4 <- setOrderingFilter(dat.batch4, unsup_clustering_genes.batch4)
p.4<-plot_ordering_genes(dat.batch4) + ggtitle("Genes Overdispersed in Batch 4")

dat.batch5<-dat.filtered.expressed.normalized[,pData(dat.filtered.expressed.normalized)$batch == "HD03_pool_3"]
dat.batch5<-estimateSizeFactors(dat.batch5)
dat.batch5<-estimateDispersions(dat.batch5,cores=detectCores()-1)
disp_table.batch5 <- dispersionTable(dat.batch5)
disp_table.batch5$gene_id<-as.character(disp_table.batch5$gene_id)
unsup_clustering_genes.batch5 <- disp_table.batch5[disp_table.batch5$dispersion_empirical >= 1.4*(disp_table.batch5$dispersion_fit),'gene_id']
dat.batch5 <- setOrderingFilter(dat.batch5, unsup_clustering_genes.batch5)
p.5<-plot_ordering_genes(dat.batch5) + ggtitle("Genes Overdispersed in Batch 5")

#grid.arrange(p.1,p.2,p.3,p.4,p.5,nrow=2)

#calculate overall dispersions
dat.filtered.expressed.normalized<-estimateSizeFactors(dat.filtered.expressed.normalized)
dat.filtered.expressed.normalized<-estimateDispersions(dat.filtered.expressed.normalized,cores=detectCores()-1)
disp_table <- dispersionTable(dat.filtered.expressed.normalized)
disp_table$gene_id <- as.character(disp_table$gene_id)

#limit to genes overdispersed in both batches and in the expressed gene set calculated earlier
tmp1 <- intersect(unsup_clustering_genes.batch1, unsup_clustering_genes.batch2)
tmp2 <- intersect(tmp1, unsup_clustering_genes.batch3)
tmp3 <- intersect(tmp2, unsup_clustering_genes.batch4)
unsup_clustering_expressed_genes <- intersect(tmp3, unsup_clustering_genes.batch5)
#419

dat.filtered.expressed.normalized <- setOrderingFilter(dat.filtered.expressed.normalized, unsup_clustering_expressed_genes)
p.expressed<-plot_ordering_genes(dat.filtered.expressed.normalized) + ggtitle("Genes Overdispersed in Both Batches\nExpressed Genes")

#grid.arrange(p.1,p.2,p.3,p.4,p.5,p.expressed,ncol=3)

embedding <- umap(t(log10(exprs(dat.filtered.expressed.normalized[unsup_clustering_expressed_genes,]) + 1)), n_neighbors = 20L, random_state = 100L, min_dist = .001, spread = 15, n_epochs = 5000)

#embedding <- umap(t(log10(exprs(dat.filtered.expressed.normalized[unsup_clustering_expressed_genes,]) + 1)), n_neighbors = 20L, random_state = 100L)

pData(dat.filtered.expressed.normalized)$UMAP1 <- embedding$UMAP1
pData(dat.filtered.expressed.normalized)$UMAP2 <- embedding$UMAP2

ggplot(pData(dat.filtered.expressed.normalized), aes(x = UMAP1, y = UMAP2, color = genotype)) +
  geom_point(size = 1) +
  scale_color_brewer(palette = "Set1") +
  theme_classic() +
  ggtitle("Umap on 419 high-variance expressed genes")

pData(dat.filtered)$UMAP1 <- embedding$UMAP1
pData(dat.filtered)$UMAP2 <- embedding$UMAP2
```

```{r plot_by_condition}
pData(dat.filtered) %>%
  mutate(condition = paste(age, genotype, sex)) %>%
  ggplot(.) + 
  geom_point(aes(x = UMAP1,y = UMAP2, col = condition)) +
  scale_color_brewer("Condition", palette = "Paired") +
  theme(legend.position = c(0.6, 0.2), legend.text = element_text(size = 18), legend.title = element_text(size = 20))

pca.3 <- prcomp(t(exprs(dat.filtered.expressed.normalized[fData(dat.filtered.expressed.normalized)$use_for_ordering,])), scale = F, center = T)
```

```{r clustering}
m <- Mclust(data = pData(dat.filtered.expressed.normalized)[c("UMAP1", "UMAP2")]) # returns 9 clusters
plot(m)
pData(dat.filtered.expressed.normalized)$cluster <- as.factor(m$classification)

ggplot(pData(dat.filtered.expressed.normalized), aes(x = UMAP1, y = UMAP2, color = cluster)) +
  geom_point(size = 1) +
#  scale_color_brewer(palette = "Set1") +
  theme_classic()

pData(dat.filtered)$cluster <- as.factor(m$classification)

plotUMAP(dat.filtered, color = "cluster")

original_cluster <- 1:9
new_cluster <- c(3, 5, 4, 8, 2, 6, 7, 1, 9)
x <- data.frame(original = original_cluster, new = new_cluster)
dat.filtered$cluster <- x[as.numeric(pData(dat.filtered)$cluster), "new"] #reorder clusters to be sequential on UMAP

pData(dat.filtered)$cluster <- as.factor(dat.filtered$cluster)
plotUMAP(dat.filtered, color = "cluster")

pData(dat.filtered.expressed.normalized)$cluster <- pData(dat.filtered)$cluster

naiveProportion.ex <- (table(pData(dat.filtered)$genotype)/sum(table(pData(dat.filtered)$genotype)))[2]
clusterByGenotype.ex <- table(pData(dat.filtered)$cluster, pData(dat.filtered)$genotype)

clusterByGenotype.chisq.test.ex <- apply(clusterByGenotype.ex, MARGIN = 1, function(x){
  prop.test(x[2], sum(x), alternative = "two.sided", correct = TRUE, p = naiveProportion.ex)$p.value
})

clusterByGenotype.df.ex <- 
  data.frame(index = c(1:length(unique(pData(dat.filtered)$cluster))), 
             pval = clusterByGenotype.chisq.test.ex, sig = " ")
clusterByGenotype.df.ex$sig <- as.character(clusterByGenotype.df.ex$sig)
clusterByGenotype.df.ex$sig[clusterByGenotype.df.ex$pval <= 0.01/9] <- "*"

pdf("Cluster_proportion.pdf", height = 2, width = 2)
ggplot(pData(dat.filtered)) + 
  geom_bar(aes(x = cluster, fill = genotype), position = "fill") + 
  #scale_fill_brewer("Genotype", palette = "Set1", labels = c("RetCFP/+", "RetCFP/CFP")) + 
  scale_fill_brewer(palette = "Set1", guide = F) + 
  geom_hline(aes(yintercept = naiveProportion.ex), linetype = "dashed", col = "grey20") +
  geom_text(aes(x = index, 
                y = 1, 
                hjust = -1,
                vjust = 1,
                label = sig), 
            data = clusterByGenotype.df.ex, 
            angle = 90, 
            size = 2) + 
  coord_cartesian(clip = "off") +
  scale_y_continuous("Proportion", expand = c(0,0), breaks = c(0, 0.25, naiveProportion.ex, 0.5, 0.75, 1), labels = c(0, 0.25, "naive\nproportion", 0.5, 0.75, 1)) +
  scale_x_discrete("Cluster") +
  theme(plot.margin = margin(6, 0, -10, -12),
        axis.ticks.x = element_blank(), 
        axis.text = element_text(size = 6),
        axis.text.x = element_text(vjust = 2),
        axis.title = element_text(size = 8),
        axis.title.x = element_text(vjust = 4),
        axis.title.y = element_text(vjust = -7)) +
  monocle:::monocle_theme_opts()
dev.off()
```


```{r celltype_classifier}
model <- readRDS("../Linnarsson_enteric_data/RFmodel_subset.rds")

test_dat <- t(projection70$projection)
test_dat <- apply(test_dat, 2, function(x){ (x-mean(x))/sd(x)})

pData(dat.filtered)$celltype_pred <- predict(model, newdata = test_dat)
table(pData(dat.filtered)$celltype_pred) # No cells classified as ENMFB
pData(dat.filtered)$parentCelltype <- ifelse(str_detect(pData(dat.filtered)$celltype_pred, "ENTG"), "progenitor/glia", "neuron")

pdf("parentcelltype_classification.pdf", height = 2, width = 2, useDingbats = F)
plotUMAP(dat.filtered, color = "parentCelltype", size = 0.1) +
  scale_color_manual("Cell Type", values = cbPalette[1:2]) +
    theme(aspect.ratio = 1,
        plot.title = element_text(size = 6, face = "plain", hjust = 0.01, margin = margin(b = -12)),
        strip.background = element_blank(), 
        strip.text = element_blank(), 
        axis.title = element_text(size = 6),
        axis.title.x = element_text(margin = margin(t = -1)),
        axis.title.y = element_text(margin = margin(r = -1)),
        axis.text = element_text(size = 6),
        axis.text.x = element_text(margin = margin(t = -1)),
        axis.text.y = element_text(margin = margin(r = -1)),
        axis.ticks = element_blank(),
        legend.position = c(0.6, 0.2),
        legend.background = element_blank(),
        legend.key.height = unit(7, "pt"),
        legend.key.width = unit(5, "pt"),
        legend.text = element_text(size = 6),
        legend.title = element_text(size = 8)
        ) 
dev.off()

pdf("Cluster_proportion_celltype.pdf", height = 3, width = 3, useDingbats = F)
ggplot(pData(dat.filtered)) + 
  geom_bar(aes(x = cluster, fill = parentCelltype), position = "fill") + 
  geom_hline(aes(yintercept = 0.1), linetype = "dashed", col = "grey20") +
  geom_hline(aes(yintercept = 0.9), linetype = "dashed", col = "grey20") +
  scale_fill_manual(values = cbPalette[1:2], guide = F) +
  scale_y_continuous("Proportion", expand = c(0,0), breaks = c(0, 0.1, 0.25, 0.5, 0.75, 0.9, 1)) +
  scale_x_discrete("Cluster") +
  theme(plot.margin = margin(6, 0, -10, 0),
        axis.ticks.x = element_blank(), 
        axis.text = element_text(size = 6),
        axis.text.x = element_text(vjust = 2),
        axis.title = element_text(size = 8),
        axis.title.x = element_text(vjust = 4),
        axis.title.y = element_text(vjust = 0)) +
  monocle:::monocle_theme_opts()
dev.off()

pData(dat.filtered)$celltype <- 
apply(pData(dat.filtered), 1, function(x){
if(x[["cluster"]] %in% c(1:5)){
  "progenitor/glia"
}else if(x[["cluster"]] %in% c(7, 8)){
  "neuron"
} else(
  x[["parentCelltype"]]
)})

plotUMAP(dat.filtered, color = "celltype") +
  scale_color_manual(values = cbPalette[1:2]) +
  theme(aspect.ratio = 1)

pdf("UMAP_parentcelltype_neuron_glia.pdf", height = 2, width = 2, useDingbats = F)
plotUMAP(dat.filtered, color = "celltype", size = 0.1) +
  scale_color_manual("Cell Type", values = cbPalette[c(1:2)]) +
    theme(aspect.ratio = 1,
        plot.title = element_text(size = 6, face = "plain", hjust = 0.01, margin = margin(b = -12)),
        strip.background = element_blank(), 
        strip.text = element_blank(), 
        axis.title = element_text(size = 6),
        axis.title.x = element_text(margin = margin(t = -1)),
        axis.title.y = element_text(margin = margin(r = -1)),
        axis.text = element_text(size = 6),
        axis.text.x = element_text(margin = margin(t = -1)),
        axis.text.y = element_text(margin = margin(r = -1)),
        axis.ticks = element_blank(),
        legend.position = c(0.3, 0.2),
        legend.background = element_blank(),
        legend.key.height = unit(10, "pt"),
        legend.key.width = unit(5, "pt"),
        legend.text = element_text(size = 6),
        legend.title = element_text(size = 8)
        ) 
dev.off()

#Etv1, Gal, Nos1, Vip = Branch A = inhibitory MNs
grid.arrange(plotUMAP(dat.filtered, markers = c("Etv1", "Gal", "Nos1", "Vip")),
             plotUMAP(dat.filtered, color = "cluster"),
             nrow = 1)

#Bnc2, Dlx5, Mgat4c, Ndufa4l2 = Branch B = excitatory MNs
grid.arrange(plotUMAP(dat.filtered, markers = c("Bnc2", "Dlx5", "Mgat4c", "Ndufa4l2")),
             plotUMAP(dat.filtered, color = "cluster"),
             nrow = 1)

grid.arrange(plotUMAP(dat.filtered, markers = "Plp1"),
             plotUMAP(dat.filtered, color = "cluster"),
             nrow = 1)

pData(dat.filtered)$celltype <- 
apply(pData(dat.filtered), 1, function(x){
if(x[["cluster"]] == 7){
  "inhibitory motor neuron"
}else if(x[["cluster"]] == 8){
  "excitatory motor neuron"
} else(
  x[["celltype"]]
)})

pData(dat.filtered)$celltype <- factor(pData(dat.filtered)$celltype, levels = c("progenitor/glia", "neuron", "inhibitory motor neuron", "excitatory motor neuron"))

plotUMAP(dat.filtered, color = "celltype") +
  scale_color_manual(values = cbPalette[c(1:3, 6)]) +
  theme(aspect.ratio = 1)

pdf("UMAP_parentcelltype_neuron_glia.pdf", height = 2, width = 2, useDingbats = F)
plotUMAP(dat.filtered, color = "celltype", size = 0.1) +
  scale_color_manual("Cell Type", values = cbPalette[c(1:3, 6)]) +
    theme(aspect.ratio = 1,
        plot.title = element_text(size = 6, face = "plain", hjust = 0.01, margin = margin(b = -12)),
        strip.background = element_blank(), 
        strip.text = element_blank(), 
        axis.title = element_text(size = 6),
        axis.title.x = element_text(margin = margin(t = -1)),
        axis.title.y = element_text(margin = margin(r = -1)),
        axis.text = element_text(size = 6),
        axis.text.x = element_text(margin = margin(t = -1)),
        axis.text.y = element_text(margin = margin(r = -1)),
        axis.ticks = element_blank(),
        legend.position = c(0.3, 0.2),
        legend.background = element_blank(),
        legend.key.height = unit(10, "pt"),
        legend.key.width = unit(5, "pt"),
        legend.text = element_text(size = 6),
        legend.title = element_text(size = 8)
        ) 
dev.off()
```

```{r}
naiveProportion.ex <- (table(pData(dat.filtered)$genotype)/sum(table(pData(dat.filtered)$genotype)))[2]
E12.5naiveProportion.ex <- (table(pData(dat.filtered)[pData(dat.filtered)$age == "E12.5","genotype"])/sum(table(pData(dat.filtered)[pData(dat.filtered)$age == "E12.5","genotype"])))[2]

E12.5celltypeByGenotype.ex <- table(pData(dat.filtered)[pData(dat.filtered)$age == "E12.5","celltype"], pData(dat.filtered)[pData(dat.filtered)$age == "E12.5","genotype"])

E12.5celltypeByGenotype.chisq.test.ex <- apply(E12.5celltypeByGenotype.ex, MARGIN = 1, function(x){
  prop.test(x[2], sum(x), alternative = "two.sided", correct = TRUE, p = E12.5naiveProportion.ex)$p.value
})

E12.5celltypeByGenotype.df.ex <- 
  data.frame(index = c(1:length(unique(pData(dat.filtered)$celltype))), 
             pval = E12.5celltypeByGenotype.chisq.test.ex, sig = " ")
E12.5celltypeByGenotype.df.ex$sig <- as.character(E12.5celltypeByGenotype.df.ex$sig)
E12.5celltypeByGenotype.df.ex$sig[E12.5celltypeByGenotype.df.ex$pval <= 0.01/2] <- "*"


E14.5naiveProportion.ex <- (table(pData(dat.filtered)[pData(dat.filtered)$age == "E14.5","genotype"])/sum(table(pData(dat.filtered)[pData(dat.filtered)$age == "E14.5","genotype"])))[2]

E14.5celltypeByGenotype.ex <- table(pData(dat.filtered)[pData(dat.filtered)$age == "E14.5","celltype"], pData(dat.filtered)[pData(dat.filtered)$age == "E14.5","genotype"])

E14.5celltypeByGenotype.chisq.test.ex <- apply(E14.5celltypeByGenotype.ex, MARGIN = 1, function(x){
  prop.test(x[2], sum(x), alternative = "two.sided", correct = TRUE, p = E14.5naiveProportion.ex)$p.value
})

E14.5celltypeByGenotype.df.ex <- 
  data.frame(index = c(1:length(unique(pData(dat.filtered)$celltype))), 
             pval = E14.5celltypeByGenotype.chisq.test.ex, sig = " ")
E14.5celltypeByGenotype.df.ex$sig <- as.character(E14.5celltypeByGenotype.df.ex$sig)
E14.5celltypeByGenotype.df.ex$sig[E14.5celltypeByGenotype.df.ex$pval <= 0.01/2] <- "*"

pdf("celltype_proportion.pdf", height = 3, width = 3)
ggplot(pData(dat.filtered)) + 
  geom_bar(aes(x = celltype, color = celltype, fill = genotype), position = "fill") + 
  facet_wrap(~age) +
  #scale_fill_brewer("Genotype", palette = "Set1", labels = c("RetCFP/+", "RetCFP/CFP")) + 
  scale_fill_manual(values = c("gray70", "darkred"), guide = F) + 
  scale_color_manual(values = cbPalette[1:2], guide = F) + 
  geom_hline(aes(yintercept = E12.5naiveProportion.ex), linetype = "dashed", col = "black") +
  geom_hline(aes(yintercept = E14.5naiveProportion.ex), linetype = "dashed", col = "black") +
  geom_text(aes(x = index, 
                y = 1, 
                hjust = -1,
                vjust = 1,
                label = sig), 
            data = celltypeByGenotype.df.ex, 
            angle = 90, 
            size = 2) + 
  coord_cartesian(clip = "off") +
  scale_y_continuous("Proportion", expand = c(0,0), breaks = c(0, E14.5naiveProportion.ex, 0.25, E12.5naiveProportion.ex, 0.5, 0.75, 1), labels = c(0, "E14.5 naive\nproportion", 0.25, "E12.5 naive\nproportion", 0.5, 0.75, 1)) +
  scale_x_discrete("Cell Type") +
  theme(plot.margin = margin(6, 0, -10, -12),
        axis.ticks.x = element_blank(), 
        axis.text = element_text(size = 6),
        axis.text.x = element_text(vjust = 2),
        axis.title = element_text(size = 8),
        axis.title.x = element_text(vjust = 4),
        axis.title.y = element_text(vjust = -7)) +
  monocle:::monocle_theme_opts()
dev.off()
```


# Cluster Specificity
```{r specificity}
melted_df <- melt(exprs(dat.filtered[expressed_genes,]))
colnames(melted_df) <- c("gene_id","cell_id","value")
melted_df <- merge(melted_df, pData(dat.filtered)[, c("genotype", "sex", "age", "cluster")], by.x = "cell_id", by.y = 0)

mean_expr_cluster <- melted_df %>%
  group_by(cluster, gene_id) %>%
  summarise(mean_expression_cluster = mean(value)) %>%
  as.data.frame()

mean_expr_cluster <- dcast(mean_expr_cluster, gene_id ~ cluster)
rownames(mean_expr_cluster) <- mean_expr_cluster$gene_id
mean_expr_cluster <- mean_expr_cluster[,-1]

#Data frame of mean expression for each gene in each cluster

cluster_specificity <- cummeRbund:::.specificity(mean_expr_cluster)

tmp <- as.data.frame(lookupGeneName(dat.filtered, rownames(cluster_specificity)))
names(tmp) <- "gene_name"

write.csv(cbind(tmp, cluster_specificity), "cluster_specificity_scores.csv")

melted_cluster_specificity <- melt(cluster_specificity)
mean_expr_cluster_melted <- melt(mean_expr_cluster)
names(mean_expr_cluster_melted) <- c("cluster","mean_expr")
mean_expr_cluster_melted$gene_id <- rep(row.names(mean_expr_cluster), nlevels(pData(dat.filtered)$cluster))
melted_cluster_specificity$Var2 <- as.numeric(str_replace(melted_cluster_specificity$Var2, pattern = "_spec", replacement = ""))
names(melted_cluster_specificity) <- c("gene_id", "cluster", "specificity")
melted_df <- merge(melted_cluster_specificity, mean_expr_cluster_melted, by = c("gene_id", "cluster"))
melted_df$gene_short_name <- lookupGeneName(dat.filtered, as.character(melted_df$gene_id), unique = F)

index <- apply(cluster_specificity, 1, which.max)
gene_specificity <- list()
for(i in 1:nlevels(pData(dat.filtered)$cluster)){
  gene_specificity[[i]] <- lookupGeneName(dat.filtered, names(index[index == i]))
}

callouts <- list()
callouts[[1]] <- c("Gpr17", "Anxa2", "Cnn2", "Myl9", "Plp1", "S100a16", "Sox5", "Rgcc", "Fabp7", "Ccnd2", "Ednrb", "Cdkn1c")
callouts[[2]] <- c("Igf1", "Hist1h2ah", "Hist1h2ai", "Cyr61", "Hist1h2ap", "Hist1h1b", "Hist1h2ag", "Hes1", "Ccne2", "Tubb6", "Hist1h2ao")
callouts[[3]] <- c("Epcam", "Cela1", "Cpa2", "Ccnf", "Bub1", "Aurka", "Aurkb", "Top2a", "Cdca3", "Cdca8", "Ccna2", "Cdc20", "Cks2", "Cdk1", "Ccnb1", "Ube2c")
callouts[[4]] <- c("Hist1h2br", "Nfix", "Neurod4", "Clcn1")
callouts[[5]] <- c("Pcp2", "H2-Ab1", "hes5")
callouts[[6]] <- c("Btbd17", "St18", "Mfap4")
callouts[[7]] <- c("Agrp", "Nos1", "Kcnv1", "Ptgir", "Ntng1", "Nrg1", "Ache", "Myl1", "Sncg", "Vip", "Etv1", "Prph", "Ass1", "Stmn3", "Gal", "Cartpt")
callouts[[8]] <- c("Tac1", "Ecel1", "Ndufa4l2", "Mgat4c", "Bnc2", "Dlx5", "Npy", "Gfra2", "Slc18a3", "Chgb", "Nrxn", "Scn3b", "Pcp4", "Snap25", "Stmn2") 
callouts[[9]] <- c("Hoxc10", "Hoxa10", "Hoxc6", "Hoxa7", "Shox2", "Hoxc9", "Hoxa9", "Gata2", "Gfra3", "Maob", "Agtr2", "Slc18a1", "Hoxb7", "Cdkn1a", "Syn2", "Gata3", "Pmp22", "Phox2a", "Th", "Chga", "Maoa", "Ddc", "Dbh", "Nefl", "Nefm")


plot_specificity <- function(x){
#  ggplot(melted_df[melted_df$cluster == x,], aes(x = log10(mean_expr + 1), 
  df <- melted_df[melted_df$cluster == x & melted_df$gene_short_name %in% gene_specificity[[x]],]
  ggplot(df, aes(x = log10(mean_expr + 1), y = specificity)) +
    ggtitle(paste("Cluster", x)) +
    geom_point(aes(color = ifelse(ifelse(df$gene_short_name %in% callouts[[x]],
                                                       df$gene_short_name, 
                                                       "") == "", "no", "yes")),
                 size = 0.1) +
    scale_color_manual(values = c("no" = "black", "yes" = "red")) +
    scale_x_continuous("log10(mean expression + 1)", expand = c(0,0)) +
    scale_y_continuous("Specificity", expand = c(0,0), limits = c(0,1)) +
    geom_text(aes(label = ifelse(df$gene_short_name %in% callouts[[x]],
                                                       df$gene_short_name, 
                                                       "")),
                size = 1.4, 
                hjust = 0, 
                vjust = 0, 
                nudge_x = 0.02, 
                nudge_y = -0.004, 
                color = "black") +
    geom_smooth(method = "gam", formula = y ~ s(x, bs = "cs"), size = 0.4) +
    theme(legend.position = "none",
            plot.title = element_text(size = 10, hjust = 0.005, face = "plain"),
            plot.margin = margin(0, 2, 2, 2),
            axis.title = element_text(size= 8),
            axis.title.x = element_text(margin = margin(t = 0)),
            axis.title.y = element_text(margin = margin(r = 0)),
            axis.text = element_text(size = 6),
            aspect.ratio = 1
            )
}



p1 <- plot_specificity(1)
p2 <- plot_specificity(2)
p3 <- plot_specificity(3)
p4 <- plot_specificity(4)
p5 <- plot_specificity(5)
p6 <- plot_specificity(6)
p7 <- plot_specificity(7)
p8 <- plot_specificity(8)
p9 <- plot_specificity(9)

pdf("SF1-cluster_specificity_callouts.pdf", height = 12, width = 12, useDingbats = F)
grid.arrange(p1, p2, p3, p4, p5, p6, p7, p8, p9, ncol = 3)
dev.off()
```

```{r,include=F,eval=F}
saveRDS(dat.filtered,file="datfiltered_postCTH.rds")
```

# Session Info
```{r SessionInfo,echo=F}
sessionInfo()
```