---
title: "HSCR CoGAPS"
author: "Liz Vincent"
date: "`r format(Sys.time(), '%B %d, %Y')`"
output: pdf_document
---

```{r import}
load("/Users/Liz/Documents/Hopkins/Goff\ Lab/Chakravarti/CoGAPS/np25_dat.filtered.expressed.normalized_result.RData")
np25.gapsResult <- gapsResult
row.names(np25.gapsResult@sampleFactors) <- colnames(exprs(dat.filtered.expressed.normalized))
row.names(np25.gapsResult@featureLoadings) <- rownames(exprs(dat.filtered.expressed.normalized))
row.names(np25.gapsResult@sampleStdDev) <- colnames(exprs(dat.filtered.expressed.normalized))
row.names(np25.gapsResult@featureStdDev) <- rownames(exprs(dat.filtered.expressed.normalized))
rownames(np25.gapsResult@featureLoadings) <- gsub("\\..*", "", rownames(np25.gapsResult@featureLoadings))
rownames(np25.gapsResult@featureStdDev) <- gsub("\\..*", "", rownames(np25.gapsResult@featureStdDev))

np25cor <- cor(np25.gapsResult@sampleFactors)

pattern_order <- order.dendrogram(as.dendrogram(hclust(dist(t(np25cor)))))
np25.gapsResult@sampleFactors <- np25.gapsResult@sampleFactors[,pattern_order]
np25.gapsResult@sampleStdDev <- np25.gapsResult@sampleStdDev[,pattern_order]
np25.gapsResult@featureLoadings <- np25.gapsResult@featureLoadings[,pattern_order]
np25.gapsResult@featureStdDev <- np25.gapsResult@featureStdDev[,pattern_order]

#Rename patterns after clustering
colnames(np25.gapsResult@sampleFactors) <- paste0("Pattern_", c(1:25))
colnames(np25.gapsResult@sampleStdDev) <- paste0("Pattern_", c(1:25))
colnames(np25.gapsResult@featureLoadings) <- paste0("Pattern_", c(1:25))
colnames(np25.gapsResult@featureStdDev) <- paste0("Pattern_", c(1:25))

saveRDS(np25.gapsResult, "HSCR_np25.rds")
np25.gapsResult <- readRDS("HSCR_np25.rds")
```

```{r UMAP_by_pattern_weight}
plot_list = list()
for (i in 1:25) {
      p = plotUMAP(dat.filtered, color_by = np25.gapsResult@sampleFactors[,i]) + 
        ggtitle(paste("Pattern", i)) + 
        labs(color = "Pattern\nWeight") +
        theme_classic() + 
        theme(aspect.ratio = 1, 
              legend.position = c(0.85, 0.2),
              axis.text = element_blank(),
              axis.ticks = element_blank())
      plot_list[[i]] = p
}
pdf(paste0("np25.pdf"))
  for (i in seq_along(plot_list)) {
      print(plot_list[[i]])
}
dev.off()
```

```{r pattern_meta_corr}
np25cor <- cor(np25.gapsResult@sampleFactors)
plot(as.dendrogram(hclust(dist(t(np25cor)))))

#One-hot encode numeric traits

tmp <- pData(dat.filtered)[,c("genotype", "sex", "age")]
tmp$genotype <- ifelse(tmp$genotype == "het", 0, 1)
tmp$sex <- ifelse(tmp$sex == "male", 0, 1)
tmp$age <- ifelse(tmp$age == "E12.5", 0, 1)

tmp2 <- pData(dat.filtered) %>% 
  mutate(condition = paste(age, genotype, sex)) %>%
  select(celltype, batch, cluster, condition)

library(caret)
dmy <- caret::dummyVars("~.", data = tmp2)
cor_meta <- cbind(tmp, data.frame(predict(dmy, newdata = tmp2)))

cor_meta <- cbind(cor_meta, pData(dat.filtered)[,c("total_mass","total_reads", "alignment_percentage", "num_genes_expressed", "total_mRNAs", "mean_expr","sd_expr")])


np25.cor <- cor(cor_meta, np25.gapsResult@sampleFactors)
np25.cor <- np25.cor[order.dendrogram(as.dendrogram(hclust(dist(np25.cor)))),]
pmat <- cor.mtest(cbind(cor_meta, np25.gapsResult@sampleFactors))
pmat <- pmat$p
rownames(pmat) <- colnames(cbind(cor_meta, np25.gapsResult@sampleFactors))
colnames(pmat) <- colnames(cbind(cor_meta, np25.gapsResult@sampleFactors))
pmat <- pmat[rownames(np25.cor), colnames(np25.cor)] # subset & make sure they're in the same order

cutoff <- .05/(dim(np25.cor)[1]*dim(np25.cor)[2])

#tmp <- melt(np25.cor)
#tmp2 <- melt(pmat)
tmp3 <- merge(melt(np25.cor), melt(pmat), by = c("Var1", "Var2"))
names(tmp3) <- c("metadata", "pattern", "correlation", "pvalue")

meta_meta <- as.data.frame(colnames(cor_meta))
colnames(meta_meta) <- "metadata"
meta_meta$label <- c("Genotype", "Sex", "Age", "Progenitor/Glia", "Neuron", "Inhibitory MN", "Excitatory MN", "Batch 1", "Batch 2", "Batch 3", "Batch 4", "Batch 5", "Cluster 1", "Cluster 2", "Cluster 3", "Cluster 4", "Cluster 5", "Cluster 6", "Cluster 7", "Cluster 8", "Cluster 9", "E12.5 Het F", "E12.5 Het M", "E12.5 Hom F", "E12.5 Hom M", "E14.5 Het F", "E14.5 Het M", "E14.5 Hom F", "E14.5 Hom M", "Total mRNA Mass", "Reads", "Alignment %", "# Genes Expr", "# mRNAs", "Mean Expr","SD Expr")
meta_meta$category <- c(rep("Experimental Var", 3), rep("Cell Type", 4), rep("Batch", 5), rep("Cluster", 9), rep("Condition", 8), rep("Technical Vars", 7))
meta_meta$category <- factor(meta_meta$category, levels = c("Experimental Var", "Condition", "Cell Type", "Cluster", "Technical Vars", "Batch"))
meta_meta <- meta_meta[match(rownames(np25.cor), as.character(meta_meta$metadata)),]
meta_meta <- meta_meta[order(meta_meta$category),]
meta_meta$y <- seq(1, nrow(meta_meta))

tmp3$metadata <- as.character(tmp3$metadata)
meta_meta$metadata <- as.character(meta_meta$metadata)

tmp4 <- merge(tmp3, meta_meta, by = "metadata")

tmp4$pattern <- str_remove(tmp4$pattern, "Pattern_")
tmp4$pattern <- as.numeric(tmp4$pattern)

pdf("HSCR_meta_cor.pdf", width = 3.25, height = 4)
ggplot(tmp4) +
  geom_tile(aes(x = pattern, 
                y = -y, 
                fill = ifelse(tmp3$pvalue < cutoff, correlation, NA)), 
            color = "black") +
  scale_fill_viridis("Correlation", 
                     na.value = "grey80",
                     limits = c(-1,1),
                      option = "inferno")  +
#                     breaks = c(0, 0.5, 1)) +
  scale_x_continuous("HSCR Pattern", 
                     breaks = c(1:25), 
                     expand = c(0,0),
                     position = "top") +
  scale_y_continuous("Metadata", 
                     breaks = -unique(tmp4$y), 
                     labels = unique(tmp4$label),
#                     labels = unique(tmp3$metadata), 
                     expand = c(0,0)) +
  theme(axis.ticks = element_blank(), 
        axis.line = element_blank(), 
        axis.text = element_text(size = 4), 
        axis.text.x.top = element_text(margin = margin(b = -2)),
        axis.text.y = element_text(margin = margin(r = -2)), 
        axis.title = element_text(size = 8),
        axis.title.x.top = element_text(margin = margin(b = 0)),
        axis.title.y = element_text(margin = margin(r = -5)),
        legend.key.height = unit(5, "pt"),
        legend.key.width = unit(25, "pt"),
        legend.text = element_text(size = 4),
        legend.title = element_text(size = 6, margin = margin(l = 10, t = -8)),
        legend.position = c(-0.02, -0.04),
        legend.direction = "horizontal",
        plot.margin = margin(2, 2, 20, 2)) +
  coord_equal()
dev.off()
```

```{r corr_perm_test}
##### Check p-values by shuffling mat

ptest_df <- cor_meta
ptest_df[,] <- NA

for(i in seq_along(cor_meta)){
 ptest_df[,i] <- sample(cor_meta[,i])
}

pmat.ptest <- cor.mtest(cbind(ptest_df, np25.gapsResult@sampleFactors))
pmat.ptest <- pmat.ptest$p
rownames(pmat.ptest) <- colnames(cbind(ptest_df, np25.gapsResult@sampleFactors))
colnames(pmat.ptest) <- colnames(cbind(ptest_df, np25.gapsResult@sampleFactors))
pmat.ptest <- pmat.ptest[colnames(ptest_df), colnames(np25.gapsResult@sampleFactors)]

pmat.ptest_melted <- melt(as.data.frame(pmat.ptest))

ggplot(pmat.ptest_melted) +
  geom_histogram(aes(x = value), binwidth = 0.02)

#####run multiple permutations

ptest_df <- cor_meta
ptest_df[,] <- NA
bigMat <- list()

for(j in 1:100){
  
  for(i in seq_along(cor_meta)){
   ptest_df[,i] <- sample(cor_meta[,i])
  }

  pmat.ptest <- cor.mtest(cbind(ptest_df, np25.gapsResult@sampleFactors))
  pmat.ptest <- pmat.ptest$p
  rownames(pmat.ptest) <- colnames(cbind(ptest_df, np25.gapsResult@sampleFactors))
  colnames(pmat.ptest) <- colnames(cbind(ptest_df, np25.gapsResult@sampleFactors))
  pmat.ptest <- pmat.ptest[colnames(ptest_df), colnames(np25.gapsResult@sampleFactors)]
  
  bigMat[[j]] <- pmat.ptest
}
  
bigMat_melted <- melt(bigMat)

ggplot(bigMat_melted) +
  geom_histogram(aes(x = value), binwidth = 0.05)

ggplot(bigMat_melted) +
  geom_histogram(aes(x = value), binwidth = 0.05) +
  facet_wrap(~Var2)

ggplot(bigMat_melted) +
  geom_histogram(aes(x = value), binwidth = 0.05) +
  facet_wrap(~Var1)

melted_patterns <- melt(np25.gapsResult@sampleFactors)

ggplot(melted_patterns) +
  geom_density(aes(x = value, color = Var2))

```

```{r patternMarkers}
np25patternMarkers <- patternMarkers(np25.gapsResult)
names(np25patternMarkers$PatternMarkers) <- paste0("pattern_", seq(1:25))

for(i in seq_along(np25patternMarkers)){
  print(head(np25patternMarkers[[i]]))
}

save_patternmarkers <- function(np){
  object <- get(paste0("np", np, "patternMarkers"))
  object <- object$PatternMarkers
  nrow <- max(unlist(lapply(object, length)))
  mat <- matrix(NA, nrow = nrow, ncol = np)
  for(i in 1:np){
    object[[i]] <- lookupGeneName(dat.filtered, object[[i]])
    mat[,i] <- c(object[[i]], rep("", nrow - length(object[[i]])))
  }
  mat <- as.data.frame(mat)
  names(mat) <- paste("pattern", seq(1:np))
  write_excel_csv(mat, path = paste0("patternMarkers", np, ".csv"))
}

save_patternmarkers(25)
```

#Check this section
```{r GSEA_with_cogaps}
# example
library('limma')
library('org.Mm.eg.db')
library('KEGG.db')
library('GO.db')
library('DT')
library('ComplexHeatmap')
library('AnnotationDbi')

Z <- gsub("\\..*","", expressed_genes)

ENSEMBLToKEGG <- mapIds(org.Mm.eg.db,keys=Z,column='PATH',keytype = 'ENSEMBL',multiVals = list)
ENSEMBLToKEGG <- sapply(reverseSplit(ENSEMBLToKEGG),unique)
ENSEMBLToKEGG <- ENSEMBLToKEGG[sapply(ENSEMBLToKEGG,length)>5]
ENSEMBLToKEGG <- ENSEMBLToKEGG[sapply(ENSEMBLToKEGG,length)<=100]

ENSEMBLToGO <- AnnotationDbi::select(org.Mm.eg.db, keys = Z,columns = c("GO","ONTOLOGY"), keytype = "ENSEMBL")
ENSEMBLToGO <- ENSEMBLToGO[,c("ENSEMBL", "GO", "ONTOLOGY")]
ENSEMBLToGO <- unique(ENSEMBLToGO)
ENSEMBLToGO <- ENSEMBLToGO[ENSEMBLToGO$ONTOLOGY == "BP", ]
ENSEMBLToGO <- ENSEMBLToGO[,c("ENSEMBL", "GO")]
ENSEMBLToGO <- tapply(ENSEMBLToGO$ENSEMBL,ENSEMBLToGO$GO,list)
#ENSEMBLToGO <- sapply(reverseSplit(ENSEMBLToGO),unique)
ENSEMBLToGO <- ENSEMBLToGO[sapply(ENSEMBLToGO,length) > 10]
ENSEMBLToGO <- ENSEMBLToGO[sapply(ENSEMBLToGO,length)<=100]

np25_CGstats <- calcCoGAPSStat(np25.gapsResult, GStoGenes = ENSEMBLToGO, numPerm = 1000)

saveRDS(np25_CGstats, "np25_CGstats.rds")

write.csv(np25_CGstats,file=paste("CoGAPSStat.expressed_genes.HSCR.nP25.csv",sep=""))#,col.names=c("GSDownreg","GSUpreg"))
  ls("package:KEGG.db")
  
GOTerms = list()
for(i in 1:25){
  GSDownreg = stack(sort(np25_CGstats$GSDownreg[i,])[sort(np25_CGstats$GSDownreg[i,]) < 0.05])
  colnames(GSDownreg) = c("pval", "GOID")
  Term = AnnotationDbi::select(GO.db, keys = as.character(GSDownreg$GOID), column = 'TERM', keytype = 'GOID')
  GSDownreg = merge(GSDownreg, Term, by = "GOID", sort = F)
  
  GSUpreg = stack(sort(np25_CGstats$GSUpreg[i,])[sort(np25_CGstats$GSUpreg[i,]) < 0.05])
  colnames(GSUpreg) = c("pval", "GOID")
  Term = AnnotationDbi::select(GO.db, keys = as.character(GSUpreg$GOID), column = 'TERM', keytype = 'GOID')
  GSUpreg = merge(GSUpreg, Term, by = "GOID", sort = F)
  
  GOTerms[[paste0("Patt", i)]] = list("Downreg" = GSDownreg, "Upreg" = GSUpreg)
}

saveRDS(GOTerms, "np25_GOTerms.rds")

tmp <- lapply(GOTerms, function(x){
    x[[1]][,"TERM"]
  })

nrow <- max(unlist(lapply(tmp, length)))
mat <- matrix(NA, nrow = nrow, ncol = 25)

for(i in 1:25){
  mat[,i] <- c(tmp[[i]], 
               rep("", nrow - length(tmp[[i]])))
}

mat <- as.data.frame(mat)
names(mat) <- paste0("pattern_", seq(1:25))
write_excel_csv(mat, path = "HSCR_np25_GOTerms_Downreg.csv")

tmp <- lapply(GOTerms, function(x){
    x[[2]][,"TERM"]
  })

nrow <- max(unlist(lapply(tmp, length)))
mat <- matrix(NA, nrow = nrow, ncol = 25)

for(i in 1:25){
  mat[,i] <- c(tmp[[i]], 
               rep("", nrow - length(tmp[[i]])))
}

mat <- as.data.frame(mat)
names(mat) <- paste0("pattern_", seq(1:25))
write_excel_csv(mat, path = "HSCR_np25_GOTerms_Upreg.csv")
```

```{r clusterProfiler_highrank}
markers <- list()

apply(np25patternMarkers$PatternMarkerScores, 2, function(x){
  sum(x < 1.2)
})

for(i in 1:25){
  pattern_scores <- np25patternMarkers$PatternMarkerScores[,i,drop = F]
  index <- which(pattern_scores < 1.2)
  genes <- rownames(pattern_scores)[index]
  markers[[i]] <- genes
}

names(markers) <- paste0("pattern_", seq(1, 25, 1))

m <- lapply(markers, function(x){match(x, egENSEMBL$ensembl_id)})
marker_list <- lapply(m, function(x){egENSEMBL$gene_id[x]})
marker_list <- lapply(marker_list, function(x){x[!is.na(x)]}) #rm NA

marker_list <- marker_list[lapply(marker_list, length) > 10] # set min length to 10
marker_list <- lapply(marker_list, function(x){ # set max length to 100
  if(length(x) > 100){
    x <- x[1:100]
  }
  return(x)
})

library(clusterProfiler)
ego_markers_np25 <- lapply(marker_list, function(x){
  enrichGO(gene = x,
                OrgDb =       'org.Mm.eg.db',
                ont           = "BP",
                pAdjustMethod = "BH",
                pvalueCutoff  = 0.05,
                qvalueCutoff  = 1, 
                readable      = TRUE)
})


lapply(ego_markers_np25, function(x){
  head(x$Description)
})

saveRDS(ego_markers_np25, "ego_markers_np25.rds")
```

```{r}

# < 100 markger genes
# 6, 10, 13, 14, 16, 18, 20, 23
# < 3 GO terms
index <- c(6, 10, 13, 20, 23)
markers_subset <- list()

#subset to top 120 prior to converting to ENSEMBL ID to save computational time
#subset to 100 after conversion -- not all transcripts map to ENSEMBL IDs
j <- 1
for(i in index){
  pattern_scores <- np25patternMarkers$PatternMarkerScores[,i,drop = F]
  pattern_scores <- pattern_scores[order(pattern_scores),,drop = F]
  genes <- rownames(pattern_scores)[1:200] #top 200 genes for each pattern
  markers_subset[[j]] <- genes
  j=j+1
}

names(markers_subset) <- paste0("pattern_", index)

m <- lapply(markers_subset, function(x){match(x, egENSEMBL$ensembl_id)})
marker_subset_list <- lapply(m, function(x){egENSEMBL$gene_id[x]})
marker_subset_list <- lapply(marker_subset_list, function(x){x[!is.na(x)]}) #rm NA
marker_subset_list <- 
  lapply(marker_subset_list, function(x){x[1:100]}) #top 50 genes for each pattern

ego_markers_subset_np25 <- lapply(marker_subset_list, function(x){
  enrichGO(gene = x,
                OrgDb =       'org.Mm.eg.db',
                ont           = "BP",
                pAdjustMethod = "BH",
                pvalueCutoff  = 0.05,
                qvalueCutoff  = 1, 
                readable      = TRUE)
})

saveRDS(ego_markers_subset_np25, "ego_markers_subset_np25.rds")

lapply(ego_markers_subset_np25, function(x){
  head(x$Description)
})
```

```{r Linnarsson_projectR_np70}
linnarsson_gapsresult <- readRDS(file = "/Users/liz/Documents/Hopkins/Goff Lab/Chakravarti/Linnarsson_enteric_data/Linnarsson_np70.rds")
linnarsson_dat <- pData(readRDS(file = "../Linnarsson_enteric_data/dat.RDS"))

mat <- log10(exprs(dat.filtered[expressed_genes,]) + 1)
rownames(mat) <- gsub("\\..*", "", rownames(mat))

projection70 <- projectR(mat, linnarsson_gapsresult@featureLoadings, AnnotionObj = NA, IDcol=NA,full=TRUE)

# Plot Linnarsson patterns with HSCR projections 
tmp <- melt(linnarsson_gapsresult@sampleFactors)
tmp1 <- merge(tmp, linnarsson_dat[,c("X", "Y","parentCellType")], by.x = "Var1", by.y = 0)
names(tmp1) <- c("cell_ID", "pattern", "value", "x", "y", "subset")
tmp1$significant <- TRUE

tmp2 <- apply(t(projection70$projection), 2, function(x){
  x=x-min(x)
  x=x/max(x)
})
tmp2 <- melt(tmp2)
#tmp2 <- melt(pData(dat.filtered)[, c("cell_id", paste0("Linnarsson_Pattern_", c(1:70)))])
tmp3 <- merge(tmp2, pData(dat.filtered)[,c("UMAP1", "UMAP2", "project")], by.x = "Var1", by.y = 0)

names(tmp3) <- c("cell_ID", "pattern", "value", "x", "y", "subset")
tmp4 <- melt(t(projection70$pval) < 0.05/nrow(pData(dat.filtered)))
tmp5 <- merge(tmp3, tmp4, by.x = c("cell_ID", "pattern"), by.y = c("Var1", "Var2"), all = TRUE)
names(tmp5) <- c("cell_ID", "pattern", "value", "x", "y", "subset", "significant")

tmp6 <- rbind(tmp1, tmp5)
tmp6$subset <- as.factor(tmp6$subset)
tmp6$subset <- factor(tmp6$subset, levels(tmp6$subset)[c(1, 3, 2)])

pdf("HSCR_projected_into_Linnarsson_2.pdf", height = 140)
ggplot(tmp6[tmp6$pattern %in% paste0("Pattern_", c(1:70)),]) +
  geom_point(aes(x=x, y=y, 
     color = ifelse(significant, value, NA),
     alpha = ifelse(significant, 1, 0.1)), size = 0.1) +
  facet_wrap(pattern~subset, scales = "free", ncol = 3) +
  scale_color_viridis("Pattern weight", option = "inferno") +
  scale_alpha(guide = "none") +
  scale_x_continuous("Reduced dimension 1") +
  scale_y_continuous("Reduced dimension 2") +
  theme(strip.background = element_blank(), strip.text = element_blank(), axis.text = element_blank(), axis.ticks = element_blank(), legend.position = "bottom", legend.key.width = unit(60, "pt"))
dev.off()

plot_list = list()
for (i in 1:70) {
      p = ggplot(tmp6[tmp6$pattern == paste0("Pattern_", i),]) +
  geom_point(aes(x=x, y=y, color = value), size = 0.1) +
  facet_wrap(pattern~subset, scales = "free", ncol = 3) +
  scale_color_viridis("Pattern\nweight", option = "inferno", breaks = c(0, 0.5, 1)) +
  scale_x_continuous("Reduced dimension 1") +
  scale_y_continuous("Reduced dimension 2") +
  ggtitle(paste("Pattern", i)) +
  theme(strip.background = element_blank(), 
        strip.text = element_blank(), 
        axis.title = element_text(size = 6),
        axis.title.x = element_text(margin = margin(t = -2)),
        axis.title.y = element_text(margin = margin(r = -1)),
        axis.text = element_text(size = 6),
        axis.text.x = element_text(margin = margin(t = -2)),
        axis.text.y = element_text(margin = margin(r = -1)),
        axis.ticks = element_blank(),
        legend.position = c(0.94, 0.25), 
        legend.title = element_text(size = 6),
        legend.text = element_text(size = 6),
        legend.key.size = unit(0.25, "cm"))
      plot_list[[i]] = p
}

pdf(paste0("HSCR_proj_Linnarsson.pdf"), width = 7.25, height = 2.5)
  for (i in seq_along(plot_list)) {
      print(plot_list[[i]])
}
dev.off()

```

```{r projection_correlation}
projection.cor <- cor(cor_meta, t(projection70$projection))
projection.cor <- projection.cor[order.dendrogram(as.dendrogram(hclust(dist(projection.cor)))),]

pmat.projection <- cor.mtest(cbind(cor_meta, t(projection70$projection)))
pmat.projection <- pmat.projection$p
rownames(pmat.projection) <- colnames(cbind(cor_meta, t(projection70$projection)))
colnames(pmat.projection) <- colnames(cbind(cor_meta, t(projection70$projection)))
pmat.projection <- pmat.projection[rownames(projection.cor), colnames(projection.cor)]

cutoff <- .05/(dim(projection.cor)[1]*dim(projection.cor)[2])

tmp <- melt(projection.cor)
tmp2 <- melt(pmat.projection)
tmp3 <- merge(tmp, tmp2, by = c("Var1", "Var2"))
names(tmp3) <- c("metadata", "Linnarsson_pattern", "correlation", "pvalue")

tmp_meta_meta <- meta_meta[match(rownames(projection.cor), as.character(meta_meta$metadata)),]
tmp_meta_meta <- tmp_meta_meta[order(tmp_meta_meta$category),]
tmp_meta_meta$y <- c(1:37)

tmp4 <- merge(tmp3, tmp_meta_meta, by = "metadata")

tmp4$Linnarsson_pattern <- str_remove(tmp4$Linnarsson_pattern, "Pattern_")
tmp4$Linnarsson_pattern <- as.numeric(tmp4$Linnarsson_pattern)

pdf("Projection_metadata_cor.pdf", width = 7, height = 3.75)
ggplot(tmp4) +
  geom_tile(aes(x = Linnarsson_pattern,
                y = -y,
                fill = ifelse(tmp3$pvalue < cutoff, correlation, NA)), 
            color = "black") +
  scale_fill_viridis("Correlation", 
                     na.value = "grey80", 
                     limits = c(-1,1),
                     option = "inferno") +
  scale_x_continuous("Projection into Public Data Pattern", 
                     expand = c(0,0),
                     breaks = c(1:70),
                     position = "top") +
  scale_y_continuous("HSCR Metadata", 
                     breaks = -unique(tmp4$y), 
                     labels = unique(tmp4$label),
                     expand = c(0,0)) +
  theme(axis.ticks = element_blank(), 
        axis.line = element_blank(), 
        axis.text = element_text(size = 4), 
        axis.text.x.top = element_text(margin = margin(b = -2)),
        axis.text.y = element_text(margin = margin(r = -2)), 
        axis.title = element_text(size = 8),
        axis.title.x.top = element_text(margin = margin(b = 0)),
        axis.title.y = element_text(margin = margin(r = 1)),
        legend.key.height = unit(6, "pt"),
        legend.key.width = unit(25, "pt"),
        legend.text = element_text(size = 4),
        legend.title = element_text(size = 6, margin = margin(l = 10, t = -8)),
        legend.position = c(0.606,-0.045),
        legend.direction = "horizontal",
        plot.margin = margin(2, 2, 18, 2)
) +
  coord_equal()
dev.off()

ggplot(tmp4[tmp4$label %in% paste("Cluster", 1:9),]) +
  geom_tile(aes(x = Linnarsson_pattern,
                y = -y,
                fill = ifelse(tmp3[tmp3$metadata %in% paste0("cluster.", 1:9), "pvalue"] < cutoff, correlation, NA)), 
            color = "black") +
  scale_fill_viridis("Correlation", 
                     na.value = "grey90", 
                     #limits = c(-1,1),
                     option = "inferno") +
  scale_x_continuous("Projection into Public Data Pattern", 
                     expand = c(0,0),
                     breaks = c(1:70),
                     position = "top") +
  scale_y_continuous("HSCR Metadata", 
                     breaks = -unique(tmp4$y), 
                     labels = unique(tmp4$label),
                     expand = c(0,0)) +
  theme(axis.ticks = element_blank(), 
        axis.line = element_blank(), 
        axis.text = element_text(size = 4), 
        axis.text.x.top = element_text(margin = margin(b = -2)),
        axis.text.y = element_text(margin = margin(r = -2)), 
        axis.title = element_text(size = 8),
        axis.title.x.top = element_text(margin = margin(b = 0)),
        axis.title.y = element_text(margin = margin(r = 1)),
        legend.key.height = unit(6, "pt"),
        legend.key.width = unit(25, "pt"),
        legend.text = element_text(size = 4),
        legend.title = element_text(size = 6, margin = margin(l = 10, t = -8)),
        legend.position = c(0.606,-0.045),
        legend.direction = "horizontal",
        plot.margin = margin(2, 2, 18, 2)
) +
  coord_equal() +
  coord_flip()
```


```{r differential_pattern_usage}
# in HSCR
tmp <- melt(np25.gapsResult@sampleFactors)
tmp$genotype <- pData(dat.filtered)[tmp[,"Var1"], "genotype"]

wilcox.pvalue <- list()
ttest.pvalue <- list()
het.mean <- list()
hom.mean <- list()
pattern.mean <- list()

for(i in 1:25){
  res1 <- wilcox.test(value ~ genotype, data = tmp[tmp$Var2 == paste0("Pattern_", i),])
  wilcox.pvalue[[i]] <- res1$p.value
  res2 <- t.test(value ~ genotype, data = tmp[tmp$Var2 == paste0("Pattern_", i),])
  ttest.pvalue[[i]] <- res2$p.value
  het.mean[[i]] <- res2$estimate[1]
  hom.mean[[i]] <- res2$estimate[2]
  pattern.mean[[i]] <- mean(tmp[tmp$Var2 == paste0("Pattern_", i), "value"])
}

mylist <- c("pattern.mean", "het.mean", "hom.mean", "wilcox.pvalue", "ttest.pvalue")

myDat <- as.data.frame(c(1:25))
  
for(i in seq_along(mylist)){
  col <- as.data.frame(unlist(get(mylist[i])))
  myDat <- cbind(myDat, col)
}

colnames(myDat) <- c("pattern", mylist)
myDat[, 2:4] <- format(myDat[, 2:4], digits = 2)
myDat[,5:6] <- format(myDat[,5:6], format = "e", digits = 3)

table(myDat$wilcox.pvalue < myDat$ttest.pvalue)


myDat$wilcox.sig <- ifelse(as.numeric(myDat$wilcox.pvalue) < 0.01/25, "**", #if
       ifelse(as.numeric(myDat$wilcox.pvalue) < 0.05/25, "*", #else if
       "n.s.")) #else

myDat$ttest.sig <- ifelse(as.numeric(myDat$ttest.pvalue) < 0.01/25, "**", #if
       ifelse(as.numeric(myDat$ttest.pvalue) < 0.05/25, "*", #else if
       "n.s.")) #else
```

```{r}
sessionInfo()
```