---
title: "HSCR Neuron vs. Glia"
author: "Liz Vincent"
date: "`r format(Sys.time(), '%B %d, %Y')`"
output: pdf_document
---
```{r init,include=F}
knitr::opts_chunk$set(message=FALSE,warning=FALSE,error=FALSE,dev='pdf',fig.width=9, fig.height=6.5)

source('init.R')

#load dat.filtered and expressed_genes

#dat.filtered <- readRDS("datfiltered_postCTH.rds")
#expressed_genes <- readRDS("expressed_genes")
```

## Subset by cell type

#################
# Subset neurons
#################

```{r neuron_gene_selection}
dat.neurons <- dat.filtered[,!pData(dat.filtered)$final_celltype %in% c("Progenitor/Glia") & !pData(dat.filtered)$cluster == 8]
# 385 cells

fData(dat.neurons)$num_cells_expressed <- rowSums(exprs(dat.neurons) > 0)
fData(dat.neurons)$mean_expr <- esApply(dat.neurons,1,function(x){mean(x)})
fData(dat.neurons)$sd_expr <- esApply(dat.neurons,1,function(x){sd(x)})
fData(dat.neurons)$bcv <-
  (fData(dat.neurons)$sd_expr/fData(dat.neurons)$mean_expr)**2
fData(dat.neurons)$percent_detection <-
  (fData(dat.neurons)$num_cells_expressed/dim(dat.neurons)[2])*100

neuron_expressed_genes <- rownames(fData(dat.neurons)[fData(dat.neurons)$num_cells_expressed > 20 &fData(dat.neurons)$sd_expr > .Machine$double.eps,])
#Expressed in at least 20 cells (5%) at > 0 CPC with SD > 0
#11,892 genes
```

# Normalization by sequencing depth
```{r normalize_by_read_depth}
# sum across expressed genes for each cell and divide all gene values by sum to normalize by read depth
dat.neurons.expressed <- dat.neurons[neuron_expressed_genes,]
dat.neurons.expressed.normalized <- dat.neurons.expressed
exprs(dat.neurons.expressed.normalized) <- apply(exprs(dat.neurons.expressed), 2, function(x){x/sum(x)})*1e+5

fData(dat.neurons.expressed.normalized)$sd_expr <- esApply(dat.neurons.expressed.normalized, 1, function(x){sd(x)})
```

```{r neuron_PCA}
neurons.pca <- prcomp(t(log10(exprs(dat.neurons.expressed.normalized) + 1)), scale = T, center = T)

screeplot(neurons.pca, type = "lines", npcs = 20)

ggbiplot(neurons.pca, var.axes = F, groups = pData(dat.neurons[neuron_expressed_genes,])$batch, ellipse = T) + 
  scale_color_brewer("Batch", palette = "Set1") + 
  ggtitle("PCA on 13,403 expressed genes in the neuron subset\nPost-normalization") + 
  theme(aspect.ratio = 1)
```

### Differential test WRT genotype
```{r}
neuron_genotype_diff_test <-  differentialGeneTest(dat.neurons[neuron_expressed_genes,], fullModelFormulaStr = "~batch + sex + final_celltype + genotype", reducedModelFormulaStr = "~batch + sex + final_celltype", cores=detectCores()-1)

neuron_genotype_diff_test_sorted <- neuron_genotype_diff_test[order(neuron_genotype_diff_test$qval),]

neuron_genotype_diff_test_sigGenes <- rownames(neuron_genotype_diff_test_sorted[neuron_genotype_diff_test_sorted$qval < 0.01,])

tmp <- t(scale(t(exprs(dat.neurons[neuron_genotype_diff_test_sigGenes,])),scale = T, center = T))
heatmap_data <- log(tmp - min(tmp) + 1)

myLower <- (0)
myUpper <- (2.3)

heatmap_data[heatmap_data < myLower] <- myLower
heatmap_data[heatmap_data > myUpper] <- myUpper

rownames(heatmap_data) <- lookupGeneName(dat.neurons, rownames(heatmap_data))

heatmap_annotation <- pData(dat.neurons)[,c("final_celltype","genotype","sex","cluster")]

heatmap_colors <- list(
"Cell Type" =  c("Neuron" = "#377eb8", "Transition Cell" = "#FF7F00", "Acetylcholinergic Neuron" = "#e41a1c", "Noradrenergic Neuron" = "#4DAF4A"),
"Genotype" = c("het" = "#7FC97F", "hom" = "#BEAED4"),
"Sex" = c("female"="black","male"="gray70"),
"Cluster" = c("1" = "#E41A1C", "2" = "#377EB8", "3" = "#4DAF4A", "4" = "#984EA3", "5" = "#FF7F00", "6" = "#FFFF33", "7" = "#A65628", "8" = "#F781BF", "9" = "#999999"))

colnames(heatmap_annotation) <- c("Cell Type", "Genotype", "Sex", "Cluster")

pdf("Heatmap_neurons_wrt_genotype.pdf", height = 22, width = 17)
pheatmap(mat = heatmap_data,
         scale = "none",
         show_rownames = TRUE,
         show_colnames = FALSE,
         drop_levels = FALSE,
         #breaks = seq(myLower, myUpper, length = 100),
         annotation_col = heatmap_annotation,
         annotation_colors = heatmap_colors,
         annotation_names_col = TRUE,
         annotation_legend = TRUE,
         border_color = NA,
         main = "Neurons\n84 Genes, 1% FDR\nFull model: ~ genotype + final_celltype + sex + batch\nReduced model: ~ final_celltype + sex + batch"
         )
dev.off()

neuron_genotype_gene_names <- lookupGeneName(dat.filtered, neuron_genotype_diff_test_sigGenes)

neuron_het_means <- apply(exprs(dat.neurons)[neuron_genotype_diff_test_sigGenes, pData(dat.neurons)$genotype == "het"], 1, mean)
neuron_hom_means <- apply(exprs(dat.neurons)[neuron_genotype_diff_test_sigGenes,pData(dat.neurons)$genotype == "hom"], 1,mean)

neuron_genotype_relative_exprs <- neuron_het_means/neuron_hom_means

neuron_upregulated_in_het <- names(neuron_genotype_relative_exprs[neuron_genotype_relative_exprs > 1 | neuron_genotype_relative_exprs == "Inf"])

neuron_downregulated_in_het <- names(neuron_genotype_relative_exprs[neuron_genotype_relative_exprs < 1])

egENSEMBL <- toTable(org.Mm.egENSEMBL)

neuron_upregulated_in_het_entrez <- gsub("\\..*","", neuron_upregulated_in_het)
m<-match(neuron_upregulated_in_het_entrez,egENSEMBL$ensembl_id)
neuron_upregulated_in_het_entrez <- egENSEMBL$gene_id[m]
ego_neuron_upregulated <- enrichGO(gene = neuron_upregulated_in_het_entrez,
                OrgDb =       'org.Mm.eg.db',
                ont           = "BP",
                pAdjustMethod = "BH",
                pvalueCutoff  = 0.05,
                qvalueCutoff  = 1, 
                readable      = TRUE)

neuron_downregulated_in_het_entrez <- gsub("\\..*","",neuron_downregulated_in_het)
m<-match(neuron_downregulated_in_het_entrez,egENSEMBL$ensembl_id)
neuron_downregulated_in_het_entrez <- egENSEMBL$gene_id[m]
ego_neuron_downregulated <- enrichGO(gene = neuron_downregulated_in_het_entrez,
                OrgDb =       'org.Mm.eg.db',
                ont           = "BP",
                pAdjustMethod = "BH",
                pvalueCutoff  = 0.05,
                qvalueCutoff  = 1, 
                readable      = TRUE)
```

# Genotype:sex differential test
```{r}
neuron_genotype_sex_diff_test <-  differentialGeneTest(dat.neurons[neuron_expressed_genes,], fullModelFormulaStr = "~batch + final_celltype + genotype*sex", reducedModelFormulaStr = "~batch + final_celltype + genotype + sex", cores=detectCores()-1)

neuron_genotype_sex_diff_test_sorted <- neuron_genotype_sex_diff_test[order(neuron_genotype_sex_diff_test$qval),]

neuron_genotype_sex_diff_test_sigGenes <- rownames(neuron_genotype_sex_diff_test_sorted[neuron_genotype_sex_diff_test_sorted$qval < 0.01,])

tmp <- t(scale(t(exprs(dat.neurons[neuron_genotype_sex_diff_test_sigGenes,])),scale = T, center = T))
heatmap_data <- log(tmp - min(tmp) + 1)

myLower <- (0)
myUpper <- (2.3)

heatmap_data[heatmap_data < myLower] <- myLower
heatmap_data[heatmap_data > myUpper] <- myUpper

rownames(heatmap_data) <- lookupGeneName(dat.neurons, rownames(heatmap_data))

heatmap_annotation <- pData(dat.neurons)[,c("final_celltype","genotype","sex","cluster")]

heatmap_colors <- list(
"Cell Type" =  c("Neuron" = "#377eb8", "Transition Cell" = "#FF7F00", "Acetylcholinergic Neuron" = "#e41a1c", "Noradrenergic Neuron" = "#4DAF4A"),
"Genotype" = c("het" = "#7FC97F", "hom" = "#BEAED4"),
"Sex" = c("female"="black","male"="gray70"),
"Cluster" = c("1" = "#E41A1C", "2" = "#377EB8", "3" = "#4DAF4A", "4" = "#984EA3", "5" = "#FF7F00", "6" = "#FFFF33", "7" = "#A65628", "8" = "#F781BF", "9" = "#999999"))

colnames(heatmap_annotation) <- c("Cell Type", "Genotype", "Sex", "Cluster")

pdf("Heatmap_neurons_wrt_genotype.pdf", height = 22, width = 17)
pheatmap(mat = heatmap_data,
         scale = "none",
         show_rownames = TRUE,
         show_colnames = FALSE,
         drop_levels = FALSE,
         #breaks = seq(myLower, myUpper, length = 100),
         annotation_col = heatmap_annotation,
         annotation_colors = heatmap_colors,
         annotation_names_col = TRUE,
         annotation_legend = TRUE,
         border_color = NA,
         main = "Neurons\n84 Genes, 1% FDR\nFull model: ~ genotype + final_celltype + sex + batch\nReduced model: ~ final_celltype + sex + batch"
         )
dev.off()
```

### Differential test WRT cluster
```{r}
plotUMAP(dat.filtered, color = "cluster") +
  scale_color_manual("Cluster", values = colorRampPalette(brewer.pal(7, "Set1"))(11)[c(4:6, 1, 7:9, 2, 3, 10, 11)], guide = guide_legend(ncol=3)) +
  theme_classic() +
  theme(legend.position = c(0.8, 0.2), aspect.ratio = 1, legend.direction = "vertical")

plotUMAP(dat.filtered[,pData(dat.filtered)$cluster %in% c(4, 9)], color = "cluster") + 
  scale_color_brewer("Cluster", palette = "Set1") +
  theme_classic() +
  theme(legend.position = "bottom", aspect.ratio = 1)

bifurcation_diff_test <- differentialGeneTest(dat.filtered[expressed_genes,pData(dat.filtered)$cluster %in% c(4, 9)], fullModelFormulaStr = "~batch + sex + genotype + final_celltype + cluster", reducedModelFormulaStr = "~batch + sex + genotype + final_celltype", cores = detectCores()-1)

bifurcation_diff_test_sorted <- bifurcation_diff_test[order(bifurcation_diff_test$qval),]

sig_genes <- bifurcation_diff_test_sorted[bifurcation_diff_test_sorted$qval < 0.01, c("gene_short_name")]

to_plot <- c("Gal", "Cartpt", "Etv1", "Vip", "Th", "Gfra1", "Npy")
bifurcation_expr <- exprs(dat.filtered[lookupGeneId(dat.filtered, to_plot), pData(dat.filtered)$cluster %in% c(4, 9)])
bifurcation_expr <- t(bifurcation_expr)
tmp_df <- merge(bifurcation_expr, pData(dat.filtered)[pData(dat.filtered)$cluster %in% c(4, 9), "cluster", drop = F], by = 0)
row.names(tmp_df) <- tmp_df[,1]
tmp_df <- tmp_df[,-1]
melted_df <- melt(tmp_df)

ggplot(melted_df) +
  geom_col(aes(x = variable, y = log10(value + 1), fill = cluster), position = "dodge") +
  scale_fill_manual("Cluster", values = c("#E41A1C", "#3B87A2")) + 
  scale_x_discrete(labels = lookupGeneName(dat.filtered, as.character(melted_df$variable))) +
  scale_y_continuous("Log10(CPC + 1)", expand = c(0, 0)) +
  theme(axis.title.x = element_blank(), legend.position = c(0.8, 0.9))


neuron_cluster_diff_test <-  differentialGeneTest(dat.neurons[neuron_expressed_genes,], fullModelFormulaStr = "~batch + Cluster", reducedModelFormulaStr = "~batch", cores=detectCores()-1)

neuron_cluster_diff_test_sorted <- neuron_cluster_diff_test[order(neuron_cluster_diff_test$pval),]

fold_change <- esApply(dat.neurons[neuron_expressed_genes,pData(dat.neurons)$genotype == "hom"], 1, mean)/esApply(dat.neurons[neuron_expressed_genes,pData(dat.neurons)$genotype == "het"], 1, mean)

neuron_cluster_diff_test_merged <- merge(neuron_cluster_diff_test_sorted, as.data.frame(fold_change), by = 0)

ggplot(neuron_cluster_diff_test_merged) +
   geom_histogram(aes(x=pval))

ggplot(neuron_cluster_diff_test_merged) +
  geom_point(aes(y = -log10(pval), x=log10(fold_change), color = ifelse((log2(fold_change) > 2 | log2(fold_change) < -2) & -log10(pval) > 5, "significant","n.s.")))

neuron_cluster_diff_test_merged[(log2(neuron_cluster_diff_test_merged$fold_change) > 2 | log2(neuron_cluster_diff_test_merged$fold_change) < -2) & -log10(neuron_cluster_diff_test_merged$pval) > 5,]

neuron_cluster_diff_test_sigGenes <- neuron_cluster_diff_test_merged[(log2(neuron_cluster_diff_test_merged$fold_change) > 2 | log2(neuron_cluster_diff_test_merged$fold_change) < -2) & !neuron_cluster_diff_test_merged$fold_change == 0 & neuron_cluster_diff_test_merged$pval < .05/length(neuron_expressed_genes),"Row.names"]

tmp<-t(scale(t(exprs(dat.neurons[neuron_cluster_diff_test_sigGenes,])),scale=T,center=T))
heatmap_data<-log(tmp -min(tmp) + 1)

myLower<-(0)
myUpper<-(2.3)

heatmap_data[heatmap_data<myLower]<-myLower
heatmap_data[heatmap_data>myUpper]<-myUpper

rownames(heatmap_data)<-lookupGeneName(dat.neurons,rownames(heatmap_data))

heatmap_annotation<-pData(dat.neurons)[,c("final_celltype","genotype","sex","Cluster")]

heatmap_colors<-list(
"Cell Type" =  c("Acetylcholinergic Neuron"="#377eb8", "Ambiguous"="#999999", "Neuron"="#e41a1c", "Noradrenergic Neuron"="#984ea3"),
"Genotype" = c("het" = "#7FC97F", "hom" = "#BEAED4"),
"Sex" = c("female"="black","male"="gray70"),
"Cluster" = c("1" = "#E41A1C", "2" = "#377EB8", "3" = "#4DAF4A", "4" = "#984EA3", "5" = "#FF7F00", "6" = "#FFFF33", "7" = "#A65628", "1" = "#F781BF", "2" = "#999999"))

colnames(heatmap_annotation)<-c("Cell Type","Genotype","Sex","Cluster")

pdf("Heatmap_neurons_wrt_cluster.pdf",height=10,width=17)
pheatmap(mat = heatmap_data,
         scale="none",
         show_rownames=TRUE,
         show_colnames=FALSE,
         drop_levels=FALSE,
         breaks=seq(myLower,myUpper,length=100),
         annotation_col=heatmap_annotation,
         annotation_colors=heatmap_colors,
         annotation_names_col=TRUE,
         annotation_legend=TRUE,
         border_color = NA
         )
dev.off()
```

##################
# Subset Glia
##################

```{r glia_gene_selection}
dat.glia <- dat.filtered[, pData(dat.filtered)$final_celltype %in% c("Progenitor/Glia")]
# 588 cells

fData(dat.glia)$num_cells_expressed <- rowSums(exprs(dat.glia) > 0)
fData(dat.glia)$mean_expr <- esApply(dat.glia, 1, function(x){mean(x)})
fData(dat.glia)$sd_expr <- esApply(dat.glia, 1, function(x){sd(x)})
fData(dat.glia)$bcv <- 
  (fData(dat.glia)$sd_expr/fData(dat.glia)$mean_expr)**2
fData(dat.glia)$percent_detection <- 
  (fData(dat.glia)$num_cells_expressed/dim(dat.glia)[2])*100

glia_expressed_genes <- rownames(fData(dat.glia)[fData(dat.glia)$num_cells_expressed > 10 &fData(dat.glia)$sd_expr > 1e-20,])
#Expressed in at least 10 cells at > 0 CPC with SD > 0
#14088 genes
```

# Normalization by sequencing depth
```{r normalize_by_read_depth_glia}
# sum across expressed genes for each cell and divide all gene values by sum to normalize by read depth
dat.glia.expressed <- dat.glia[glia_expressed_genes,]
dat.glia.expressed.normalized <- dat.glia.expressed
exprs(dat.glia.expressed.normalized) <- apply(exprs(dat.glia.expressed), 2, function(x){x/sum(x)})*1e+5

fData(dat.glia.expressed.normalized)$sd_expr <- esApply(dat.glia.expressed.normalized, 1, function(x){sd(x)})
```

```{r glia_PCA}
glia.pca <- prcomp(t(log10(exprs(dat.glia[glia_expressed_genes,]) + 1)), center = T)

ggbiplot(glia.pca, var.axes = F, groups = pData(dat.glia)$batch, ellipse = T) + 
  scale_color_brewer(palette = "Set1") + 
  ggtitle("Batch") + 
  theme(aspect.ratio = 1)

glia.pca.unsup_genes <- prcomp(t(log10(exprs(dat.glia[unsup_clustering_expressed_genes.glia,]) + 1)), scale = T, center = T)

ggbiplot(glia.pca.unsup_genes, var.axes = F, groups = pData(dat.glia)$batch, ellipse = T) + 
  scale_color_brewer(palette = "Set1") + 
  ggtitle("Batch") + 
  theme(aspect.ratio = 1)
```

```{r glia_DE_wrt_genotype}
glia_genotype_diff_test <- differentialGeneTest(dat.glia[glia_expressed_genes,], fullModelFormulaStr = "~batch + sex + final_celltype +genotype", reducedModelFormulaStr = "~batch + sex + final_celltype", cores = detectCores()-1)

glia_genotype_diff_test_sorted <- glia_genotype_diff_test[order(glia_genotype_diff_test$qval),]

glia_genotype_diff_test_sigGenes <- rownames(glia_genotype_diff_test_sorted[glia_genotype_diff_test_sorted$qval <= 0.01,])

#heatmap_annotation <- pData(dat.glia)[, c("final_celltype", "genotype", "sex", "cluster")]
heatmap_annotation <- merge(pData(dat.glia)[, c("final_celltype", "genotype", "sex")], pData(dat.filtered)[,"cluster", drop = F], by = 0)
rownames(heatmap_annotation) <- heatmap_annotation[,1]
heatmap_annotation <- heatmap_annotation[,-1]

heatmap_colors <- list(
"Cell Type" =  c("Progenitor/Glia" = "#984EA3", "Transition Cell" = "#FF7F00"),
"Genotype" = c("het" = "#E41A1C", "hom" = "#377EB8"),
"Sex" = c("female"="black","male"="#999999"),
"Cluster" = c("1" = "#E41A1C", "2" = "#596A98", "3" = "#449B75", "4" = "#6B886D", "5" = "#AC5782", "6" = "#FF7F00", "7" = "#FFE528", "8" = "#C9992C", "9" = "#C66764", "10" = "#E485B7", "11" = "#999999")
)

writeLines(paste0('c("',paste(levels(pData(dat.filtered)$cluster),colorRampPalette(brewer.pal(9, "Set1"))(11),sep='" = "', collapse='", "'),'")'))


colnames(heatmap_annotation) <- c("Cell Type","Genotype","Sex","Cluster")
heatmap_annotation <- heatmap_annotation %>%
  select("Genotype","Cluster","Sex","Cell Type") #order the columns

tmp <- t(scale(t(exprs(dat.glia[glia_genotype_diff_test_sigGenes,])), scale = T, center = T))
heatmap_data <- log(tmp -min(tmp) + 1)
#heatmap_data<-tmp -min(tmp)
#myLower<-(0)
#myUpper<-(3)
#heatmap_data[heatmap_data<myLower]<-myLower
#heatmap_data[heatmap_data>myUpper]<-myUpper
rownames(heatmap_data) <- lookupGeneName(dat.glia, rownames(heatmap_data))

#pdf("Heatmap_glia_wrt_genotype.pdf",height=36,width=17)
pdf("posterFig_Heatmap_glia_wrt_genotype.pdf",height=36,width=17)
pheatmap(mat = heatmap_data,
         scale = "none",
         show_rownames = TRUE,
         show_colnames = FALSE,
         drop_levels = FALSE,
         #breaks = seq(myLower, myUpper, length = 100),
         annotation_col = heatmap_annotation,
         annotation_colors = heatmap_colors,
         annotation_names_col = TRUE,
         annotation_legend = TRUE,
         border_color = NA,
         fontsize = 8#,
         #main = "Neurons\n84 Genes, 1% FDR\nFull model: ~ genotype + final_celltype + sex + batch\nReduced model: ~ final_celltype + sex + batch"
         )
dev.off()
```

```{r GOenrichment}
library(org.Mm.eg.db)
library(clusterProfiler)
egENSEMBL <- toTable(org.Mm.egENSEMBL)

#All DEG
glia_genotype_diff_test_sigGenes_entrez <- gsub("\\..*","",glia_genotype_diff_test_sigGenes)
m<-match(glia_genotype_diff_test_sigGenes_entrez,egENSEMBL$ensembl_id)
glia_genotype_diff_test_sigGenes_entrez <- egENSEMBL$gene_id[m]
ego <- enrichGO(gene = glia_genotype_diff_test_sigGenes_entrez,
                OrgDb =       'org.Mm.eg.db',
                ont           = "BP",
                pAdjustMethod = "none",
                qvalueCutoff  = 0.01,
                pvalueCutoff  = 1, 
                readable      = TRUE)


# Subset by direction of change
glia_het_means <- apply(exprs(dat.glia)[glia_genotype_diff_test_sigGenes, pData(dat.glia)$genotype == "het"], 1, mean)
glia_hom_means <- apply(exprs(dat.glia)[glia_genotype_diff_test_sigGenes,pData(dat.glia)$genotype == "hom"], 1,mean)

glia_genotype_relative_exprs <- glia_het_means/glia_hom_means

glia_upregulated_in_het <- names(glia_genotype_relative_exprs[glia_genotype_relative_exprs > 1 | glia_genotype_relative_exprs == "Inf"])

glia_downregulated_in_het <- names(glia_genotype_relative_exprs[glia_genotype_relative_exprs < 1])



glia_upregulated_in_het_entrez <- gsub("\\..*", "", glia_upregulated_in_het)
m <- match(glia_upregulated_in_het_entrez, egENSEMBL$ensembl_id)
glia_upregulated_in_het_entrez <- egENSEMBL$gene_id[m]
ego_glia_upregulated <- enrichGO(gene = glia_upregulated_in_het_entrez,
                OrgDb =       'org.Mm.eg.db',
                ont           = "BP",
                pAdjustMethod = "none",
                qvalueCutoff  = 0.01,
                pvalueCutoff  = 1, 
                readable      = TRUE)

glia_downregulated_in_het_entrez <- gsub("\\..*", "", glia_downregulated_in_het)
m <- match(glia_downregulated_in_het_entrez, egENSEMBL$ensembl_id)
glia_downregulated_in_het_entrez <- egENSEMBL$gene_id[m]
ego_glia_downregulated <- enrichGO(gene = glia_downregulated_in_het_entrez,
                OrgDb =       'org.Mm.eg.db',
                ont           = "BP",
                pAdjustMethod = "none",
                qvalueCutoff  = 0.01,
                pvalueCutoff  = 1, 
                readable      = TRUE)


```

```{r glia_DE_wrt_Cluster}
glia_cluster_diff_test<-differentialGeneTest(dat.glia[glia_expressed_genes,], fullModelFormulaStr = "~batch + sex + genotype + final_celltype + cluster", reducedModelFormulaStr = "~batch + sex + genotype + final_celltype", cores=detectCores()-1)

glia_cluster_diff_test_sorted<-glia_cluster_diff_test[order(glia_cluster_diff_test$qval),]

glia_cluster_diff_test_sigGenes<-rownames(glia_cluster_diff_test_sorted[glia_cluster_diff_test_sorted$qval <= 1e-7,])

heatmap_annotation<-pData(dat.glia)[,c("final_celltype","genotype","sex","Cluster")]

heatmap_colors<-list(
"Cell Type" =  c("Progenitor/Glia"="#e41a1c","Ambiguous"="#999999"),
"Genotype" = c("het" = "#7FC97F", "hom" = "#BEAED4"),
"Sex" = c("female"="black","male"="#999999"),
"Cluster" = c("1" = "#E41A1C", "2" = "#377EB8", "3" = "#4DAF4A", "4" = "#984EA3"))

colnames(heatmap_annotation)<-c("Cell Type","Genotype","Sex","Cluster")
heatmap_annotation<-heatmap_annotation %>%
  select("Cluster","Genotype","Sex","Cell Type")

tmp<-t(scale(t(exprs(dat.glia[glia_cluster_diff_test_sigGenes,])),scale=T,center=T))
heatmap_data<-log(tmp -min(tmp) + 1)
#heatmap_data<-tmp -min(tmp)
myLower<-(0)
myUpper<-(3)
heatmap_data[heatmap_data<myLower]<-myLower
heatmap_data[heatmap_data>myUpper]<-myUpper
rownames(heatmap_data)<-lookupGeneName(dat.glia,rownames(heatmap_data))

pdf("Heatmap_glia_wrt_cluster.pdf",height=36,width=15)
pheatmap(mat = heatmap_data,
         scale="none",
         show_rownames=TRUE,
         show_colnames=FALSE,
         drop_levels=FALSE,
         color = colorRampPalette(rev(brewer.pal(n = 11, name = "RdYlBu")))(100),
#         color = colorRampPalette(c("blue","white","red"))(100),
         breaks=seq(myLower,myUpper,length=100),
         annotation_col=heatmap_annotation,
         annotation_colors=heatmap_colors,
         annotation_names_col=TRUE,
         annotation_legend=TRUE,
         border_color = NA
         )
dev.off()
```

```{r glia_DE_wrt_genotype*cluster}
glia_genotype_cluster_diff_test <- differentialGeneTest(dat.glia[glia_expressed_genes, pData(dat.glia)$cluster %in% c(1, 2, 5, 6, 8, 10)], fullModelFormulaStr = "~genotype*cluster + sex + batch + final_celltype", reducedModelFormulaStr = "~genotype + sex + cluster + batch + final_celltype", verbose = T)

glia_genotype_cluster_diff_test_sorted <- glia_genotype_cluster_diff_test[order(glia_genotype_cluster_diff_test$qval),]

glia_genotype_cluster_diff_test_sigGenes <- rownames(glia_genotype_cluster_diff_test_sorted[glia_genotype_cluster_diff_test_sorted$qval <= 0.01,])

heatmap_annotation <- pData(dat.glia)[, c("final_celltype", "genotype", "sex", "cluster")]

heatmap_colors <- list(
"Cell Type" =  c("Progenitor/Glia" = "#984EA3", "Transition Cell" = "#FF7F00"),
"Genotype" = c("het" = "#E41A1C", "hom" = "#377EB8"),
"Sex" = c("female"="black","male"="#999999"),
"Cluster" = c("1" = "#E41A1C", "2" = "#596A98", "3" = "#449B75", "4" = "#6B886D", "5" = "#AC5782", "6" = "#FF7F00", "7" = "#FFE528", "8" = "#C9992C", "9" = "#C66764", "10" = "#E485B7", "11" = "#999999")
)

colnames(heatmap_annotation) <- c("Cell Type","Genotype","Sex","Cluster")
heatmap_annotation <- heatmap_annotation %>%
  select("Cluster","Genotype","Sex","Cell Type")

tmp<-t(scale(t(exprs(dat.glia[glia_genotype_cluster_diff_test_sigGenes,])),scale=T,center=T))
heatmap_data<-log(tmp -min(tmp) + 1)
#heatmap_data<-tmp -min(tmp)
#myLower<-(0)
#myUpper<-(3)
#heatmap_data[heatmap_data<myLower]<-myLower
#heatmap_data[heatmap_data>myUpper]<-myUpper
rownames(heatmap_data)<-lookupGeneName(dat.glia,rownames(heatmap_data))

pdf("posterFig_Heatmap_glia_wrt_genotype_cluster.pdf",height=36,width=15)
pheatmap(mat = heatmap_data,
         scale="none",
         show_rownames=TRUE,
         show_colnames=FALSE,
         drop_levels=FALSE,
         color = colorRampPalette(rev(brewer.pal(n = 11, name = "RdYlBu")))(100),
#         color = colorRampPalette(c("blue","white","red"))(100),
#         breaks=seq(myLower,myUpper,length=100),
         annotation_col=heatmap_annotation,
         annotation_colors=heatmap_colors,
         annotation_names_col=TRUE,
         annotation_legend=TRUE,
         border_color = NA
         )
dev.off()
```

```{r Hox_gene_expr}
plot_genes_violin(dat.filtered[fData(dat.filtered)$gene_short_name %in% c("Hoxa1", "Hoxd1", "Hoxa2", "Hoxb2", "Hoxa3", "Hoxb3", "Hoxd3", "Hoxa4", "Hoxb4", "Hoxc4", "Hoxd4", "Hoxa5", "Hoxb5", "Hoxc5", "Hoxa6", "Hoxb6", "Hoxc6", "Hoxa7", "Hoxb7", "Hoxb8", "Hoxc8", "Hoxd8", "Hoxa9", "Hoxb9", "Hoxc9", "Hoxd9", "Hoxa10", "Hoxc10", "Hoxd10"),], grouping = "cluster", relative_expr = F, color_by = "cluster", nrow = 5, ncol = 6, panel_order = c("Hoxa1", "Hoxd1", "Hoxa2", "Hoxb2", "Hoxa3", "Hoxb3", "Hoxd3", "Hoxa4", "Hoxb4", "Hoxc4", "Hoxd4", "Hoxa5", "Hoxb5", "Hoxc5", "Hoxa6", "Hoxb6", "Hoxc6", "Hoxa7", "Hoxb7", "Hoxb8", "Hoxc8", "Hoxd8", "Hoxa9", "Hoxb9", "Hoxc9", "Hoxd9", "Hoxa10", "Hoxc10", "Hoxd10"))

tmp_expr <- exprs(dat.filtered.expressed.normalized)[fData(dat.filtered.expressed.normalized)$gene_short_name %in% c("Hoxa1", "Hoxd1", "Hoxa2", "Hoxb2", "Hoxa3", "Hoxb3", "Hoxd3", "Hoxa4", "Hoxb4", "Hoxc4", "Hoxd4", "Hoxa5", "Hoxb5", "Hoxc5", "Hoxa6", "Hoxb6", "Hoxc6", "Hoxa7", "Hoxb7", "Hoxb8", "Hoxc8", "Hoxd8", "Hoxa9", "Hoxb9", "Hoxc9", "Hoxd9", "Hoxa10", "Hoxc10", "Hoxd10"),]

#tmp_expr <- log10(tmp_expr*10000/pData(dat.filtered)$total_mRNAs + 1)

melted_df <- melt(tmp_expr)
colnames(melted_df) <- c("gene_id","cell_id","value")
melted_df <- merge(melted_df, pData(dat.filtered)[,c("cluster","genotype","final_celltype","sex","age")], by.x = "cell_id", by.y = 0)

#######
# CLUSTER
######

mean_expr_cluster <- melted_df %>%
  group_by(cluster, gene_id) %>%
  summarise(mean_expression = mean(value), sd_expression = sd(value)) %>%
  as.data.frame()

for(i in 1:11){
  mean_expr_cluster[mean_expr_cluster$cluster == i, "sem_expression"] <- mean_expr_cluster[mean_expr_cluster$cluster == i, "sd_expression"]/sqrt(table(pData(dat.filtered)$cluster)[i])
}

mean_expr_cluster$gene_short_name <- lookupGeneName(dat.filtered, as.character(mean_expr_cluster$gene_id))

levels(mean_expr_cluster$cluster)
mean_expr_cluster$cluster <- factor(mean_expr_cluster$cluster,levels(mean_expr_cluster$cluster)[c(10, 6, 1, 5, 3, 2, 7, 8, 9, 4, 11)])

ggplot(mean_expr_cluster, aes(x = cluster, y = mean_expression, fill = cluster)) +
  geom_col() + 
    facet_wrap(~gene_short_name, scales = "free") +
  geom_errorbar(aes(ymin=mean_expression-sem_expression, ymax=mean_expression+sem_expression))

gene_list <- c("Hoxa1", "Hoxd1", "Hoxa2", "Hoxb2", "Hoxa3", "Hoxb3", "Hoxd3", "Hoxa4", "Hoxb4", "Hoxc4", "Hoxd4", "Hoxa5", "Hoxb5", "Hoxc5", "Hoxa6", "Hoxb6", "Hoxc6", "Hoxa7", "Hoxb7", "Hoxb8", "Hoxc8", "Hoxd8", "Hoxa9", "Hoxb9", "Hoxc9", "Hoxd9", "Hoxa10", "Hoxc10", "Hoxd10")
gene_list <- gene_list[gene_list %in% lookupGeneName(dat.filtered, expressed_genes)]

for(i in seq_along(gene_list)){
  #print(gene_list[i])
  assign(paste0("p", i), 
         ggplot(mean_expr_cluster[mean_expr_cluster$gene_short_name == gene_list[i],], aes(x = cluster, y = mean_expression, fill = reorder(cluster, seq(1:11)))) +
          geom_col() + 
          geom_errorbar(aes(ymin = mean_expression-sem_expression, ymax = mean_expression+sem_expression)) +
          theme(legend.position = "none", plot.title = element_text(hjust= 0.5)) +
           ggtitle(gene_list[i])
  )
}

blank <- grid.rect(gp=gpar(col="white"))

library(cowplot)
Hox_by_cluster <- plot_grid(p1, blank, blank, p2, 
             p3, p4, blank, blank,
             p5, p6, blank, p7,
             p8, p9, p10, p11,
             p12, p13, p14, blank,
             blank, p15, p16, blank,
             p17, p18, blank, blank,
             blank, p19, blank, blank,
             p20, blank, p21, blank, 
             p22, blank, p23, blank,
             ncol = 4)

save_plot("Hox_genes_by_cluster.pdf", Hox_by_cluster, base_height = 40, base_width = 20)


mean_expr_cluster <- melted_df %>%
  group_by(cluster, genotype, gene_id) %>%
  summarise(mean_expression = mean(value), sd_expression = sd(value)) %>%
  as.data.frame()

for(i in 1:11){
  for(j in c("het", "hom")){
  mean_expr_cluster[mean_expr_cluster$cluster == i & mean_expr_cluster$genotype == j, "sem_expression"] <- mean_expr_cluster[mean_expr_cluster$cluster == i  & mean_expr_cluster$genotype == j, "sd_expression"]/sqrt(table(pData(dat.filtered)$cluster, pData(dat.filtered)$genotype)[i, j])
  }
}

mean_expr_cluster$gene_short_name <- lookupGeneName(dat.filtered, as.character(mean_expr_cluster$gene_id))

levels(mean_expr_cluster$cluster)
mean_expr_cluster$cluster <- factor(mean_expr_cluster$cluster,levels(mean_expr_cluster$cluster)[c(10, 6, 1, 5, 3, 2, 7, 8, 9, 4, 11)])

for(i in seq_along(gene_list)){
  assign(paste0("p", i), 
         ggplot(mean_expr_cluster[mean_expr_cluster$gene_short_name == gene_list[i],], aes(x = cluster, y = mean_expression, fill = genotype)) +
          geom_col(position = "dodge") + 
          geom_errorbar(aes(ymin = mean_expression-sem_expression, ymax = mean_expression+sem_expression), position = position_dodge()) +
          theme(legend.position = "none", plot.title = element_text(hjust= 0.5)) +
           scale_fill_brewer(palette = "Set1") +
           ggtitle(gene_list[i])
  )
}

Hox_by_genotype_and_cluster <- plot_grid(p1, blank, blank, p2, 
             p3, p4, blank, blank,
             p5, p6, blank, p7,
             p8, p9, p10, p11,
             p12, p13, p14, blank,
             blank, p15, p16, blank,
             p17, p18, blank, blank,
             blank, p19, blank, blank,
             p20, blank, p21, blank, 
             p22, blank, p23, blank,
             ncol = 4)

all_hox <- fData(dat.filtered)$gene_short_name[str_detect(fData(dat.filtered)$gene_short_name, "Hox[a-z][0-9]+$")]

tmp_expr <- exprs(dat.filtered)[fData(dat.filtered)$gene_short_name %in% all_hox,]

melted_df <- melt(tmp_expr)
colnames(melted_df) <- c("gene_id","cell_id","value")
melted_df <- merge(melted_df, pData(dat.filtered)[,c("cluster","genotype","final_celltype","sex","age")], by.x = "cell_id", by.y = 0)

mean_expr_cluster <- melted_df %>%
  group_by(cluster, gene_id) %>%
  summarise(mean_expression = mean(value), sd_expression = sd(value)) %>%
  as.data.frame()

for(i in 1:11){
  mean_expr_cluster[mean_expr_cluster$cluster == i, "sem_expression"] <- mean_expr_cluster[mean_expr_cluster$cluster == i, "sd_expression"]/sqrt(table(pData(dat.filtered)$cluster)[i])
}

mean_expr_cluster$gene_short_name <- lookupGeneName(dat.filtered, as.character(mean_expr_cluster$gene_id))

levels(mean_expr_cluster$cluster)
mean_expr_cluster$cluster <- factor(mean_expr_cluster$cluster,levels(mean_expr_cluster$cluster)[c(10, 6, 1, 5, 3, 2, 7, 8, 9, 4, 11)])

all_hox <- all_hox[order(as.numeric(str_extract(all_hox, "[0-9]+")))]

for(i in seq_along(all_hox)){
  #print(gene_list[i])
  assign(paste0("p", i), 
         ggplot(mean_expr_cluster[mean_expr_cluster$gene_short_name == all_hox[i],], aes(x = cluster, y = mean_expression, fill = reorder(cluster, seq(1:11)))) +
          geom_col() + 
          geom_errorbar(aes(ymin = mean_expression-sem_expression, ymax = mean_expression+sem_expression)) +
          theme(legend.position = "none", plot.title = element_text(hjust= 0.5)) +
           ggtitle(all_hox[i])
  )
}

paste0("p", 1:45, collapse = ", ")
grid.arrange(p1, p2, p3, p4, p5, p6, p7, p8, p9, p10, p11, p12, p13, p14, p15, p16, p17, p18, p19, p20, p21, p22, p23, p24, p25, p26, p27, p28, p29, p30, p31, p32, p33, p34, p35, p36, p37, p38, p39, p40, p41, p42, p43, p44, p45)
#blank <- grid.rect(gp=gpar(col="white"))

#library(cowplot)
#Hox_by_cluster <- plot_grid(p1, blank, blank, p2, 
#             p3, p4, blank, blank,
#             p5, p6, blank, p7,
#             p8, p9, p10, p11,
#             p12, p13, p14, blank,
#             blank, p15, p16, blank,
#             p17, p18, blank, blank,
#             blank, p19, blank, blank,
#             p20, blank, p21, blank, 
#             p22, blank, p23, blank,
#             ncol = 4)

#save_plot("Hox_genes_by_cluster.pdf", Hox_by_cluster, base_height = 40, base_width = 20)

#######
# GENOTYPE
######
mean_expr_genotype <- melted_df %>%
  group_by(genotype, gene_id) %>%
  summarise(mean_expression = mean(value), sd_expression = sd(value)) %>%
  as.data.frame()

for(i in c("het", "hom")){
  mean_expr_genotype[mean_expr_genotype$genotype == i, "sem_expression"] <- mean_expr_genotype[mean_expr_genotype$genotype == i, "sd_expression"]/sqrt(table(pData(dat.filtered)$genotype)[i])
}

mean_expr_genotype$gene_short_name <- lookupGeneName(dat.filtered, as.character(mean_expr_genotype$gene_id))

for(i in seq_along(gene_list)){
  #print(gene_list[i])
  assign(paste0("p", i), 
         ggplot(mean_expr_genotype[mean_expr_genotype$gene_short_name == gene_list[i],], aes(x = genotype, y = mean_expression, fill = genotype)) +
          geom_col() + 
          geom_errorbar(aes(ymin=mean_expression-sem_expression, ymax=mean_expression+sem_expression)) +
          theme(legend.position = "none", plot.title = element_text(hjust= 0.5)) +
           ggtitle(gene_list[i]) +
           scale_fill_brewer(palette = "Set1")
  )
}

Hox_by_genotype <- plot_grid(p1, blank, blank, p2, 
             p3, p4, blank, blank,
             p5, p6, blank, p7,
             p8, p9, p10, p11,
             p12, p13, p14, blank,
             blank, p15, p16, blank,
             p17, p18, blank, blank,
             blank, p19, blank, blank,
             p20, blank, p21, blank, 
             p22, blank, p23, blank,
             ncol = 4)

save_plot("Hox_genes_by_genotype.pdf", Hox_by_genotype, base_height = 40, base_width = 20)

######
# SEX
######

mean_expr_sex <- melted_df %>%
  group_by(sex, gene_id) %>%
  summarise(mean_expression = mean(value), sd_expression = sd(value)) %>%
  as.data.frame()

for(i in c("female", "male")){
  mean_expr_sex[mean_expr_sex$sex == i, "sem_expression"] <- mean_expr_sex[mean_expr_sex$sex == i, "sd_expression"]/sqrt(table(pData(dat.filtered)$sex)[i])
}

mean_expr_sex$gene_short_name <- lookupGeneName(dat.filtered, as.character(mean_expr_sex$gene_id))

for(i in seq_along(gene_list)){
  #print(gene_list[i])
  assign(paste0("p", i), 
         ggplot(mean_expr_sex[mean_expr_sex$gene_short_name == gene_list[i],], aes(x = sex, y = mean_expression, fill = sex)) +
          geom_col() + 
          geom_errorbar(aes(ymin=mean_expression-sem_expression, ymax=mean_expression+sem_expression)) +
          theme(legend.position = "none", plot.title = element_text(hjust= 0.5)) +
           ggtitle(gene_list[i]) +
           scale_fill_brewer(palette = "Set1")
  )
}

Hox_by_sex <- plot_grid(p1, blank, blank, p2, 
             p3, p4, blank, blank,
             p5, p6, blank, p7,
             p8, p9, p10, p11,
             p12, p13, p14, blank,
             p15, p16, p17, blank,
             p18, p19, blank, blank,
             blank, p20, p21, p22,
             p23, p24, p25, p26, 
             p27, blank, p28, p29,
             ncol = 4)

save_plot("Hox_genes_by_sex.pdf", Hox_by_sex, base_height = 40, base_width = 20)

######
# AGE
######

mean_expr_age <- melted_df %>%
  group_by(age, gene_id) %>%
  summarise(mean_expression = mean(value), sd_expression = sd(value)) %>%
  as.data.frame()

for(i in c("E12.5", "E14.5")){
  mean_expr_age[mean_expr_age$age == i, "sem_expression"] <- mean_expr_age[mean_expr_age$age == i, "sd_expression"]/sqrt(table(pData(dat.filtered)$age)[i])
}

mean_expr_age$gene_short_name <- lookupGeneName(dat.filtered, as.character(mean_expr_age$gene_id))

for(i in seq_along(gene_list)){
  #print(gene_list[i])
  assign(paste0("p", i), 
         ggplot(mean_expr_age[mean_expr_age$gene_short_name == gene_list[i],], aes(x = age, y = mean_expression, fill = age)) +
          geom_col() + 
          geom_errorbar(aes(ymin=mean_expression-sem_expression, ymax=mean_expression+sem_expression)) +
          theme(legend.position = "none", plot.title = element_text(hjust= 0.5)) +
           ggtitle(gene_list[i]) +
           scale_fill_brewer(palette = "Set1")
  )
}

Hox_by_age <- plot_grid(p1, blank, blank, p2, 
             p3, p4, blank, blank,
             p5, p6, blank, p7,
             p8, p9, p10, p11,
             p12, p13, p14, blank,
             p15, p16, p17, blank,
             p18, p19, blank, blank,
             blank, p20, p21, p22,
             p23, p24, p25, p26, 
             p27, blank, p28, p29,
             ncol = 4)

save_plot("Hox_genes_by_age.pdf", Hox_by_age, base_height = 40, base_width = 20)

######
# SEX X GENOTYPE
######


mean_expr_sexXgenotype <- melted_df %>%
  group_by(sex, genotype, gene_id) %>%
  summarise(mean_expression = mean(value), sd_expression = sd(value)) %>%
  as.data.frame()

for(i in c("female", "male")){
  for(j in c("het","hom"))
  mean_expr_sexXgenotype[mean_expr_sexXgenotype$sex == i & mean_expr_sexXgenotype$genotype == j, "sem_expression"] <- mean_expr_sexXgenotype[mean_expr_sexXgenotype$sex == i & mean_expr_sexXgenotype$genotype == j, "sd_expression"]/sqrt(table(pData(dat.filtered)$sex, pData(dat.filtered)$genotype)[i,j])
}

mean_expr_sexXgenotype$gene_short_name <- lookupGeneName(dat.filtered, as.character(mean_expr_sexXgenotype$gene_id))

for(i in seq_along(gene_list)){
  #print(gene_list[i])
  assign(paste0("p", i), 
         ggplot(mean_expr_sexXgenotype[mean_expr_sexXgenotype$gene_short_name == gene_list[i],], aes(x = sex, y = mean_expression, fill = sex)) +
          geom_col() + 
          geom_errorbar(aes(ymin=mean_expression-sem_expression, ymax=mean_expression+sem_expression)) +
           facet_wrap(~genotype) +
          theme(legend.position = "none", plot.title = element_text(hjust= 0.5)) +
           ggtitle(gene_list[i]) +
           scale_fill_brewer(palette = "Set1")
  )
}

Hox_by_sexXgenotype <- plot_grid(p1, blank, blank, p2, 
             p3, p4, blank, blank,
             p5, p6, blank, p7,
             p8, p9, p10, p11,
             p12, p13, p14, blank,
             p15, p16, p17, blank,
             p18, p19, blank, blank,
             blank, p20, p21, p22,
             p23, p24, p25, p26, 
             p27, blank, p28, p29,
             ncol = 4)

save_plot("Hox_genes_by_sexXgenotyep.pdf", Hox_by_sexXgenotype, base_height = 40, base_width = 20)

######
# SEX X GENOTYPE X AGE
######

mean_expr_sexXgenotypeXage <- melted_df %>%
  group_by(sex, genotype, age, gene_id) %>%
  summarise(mean_expression = mean(value), sd_expression = sd(value)) %>%
  as.data.frame()

for(i in c("female", "male")){
  for(j in c("het","hom")){
    for(k in c("E12.5","E14.5")){
  mean_expr_sexXgenotypeXage[mean_expr_sexXgenotypeXage$sex == i & mean_expr_sexXgenotypeXage$genotype == j & mean_expr_sexXgenotypeXage$age == k, "sem_expression"] <- mean_expr_sexXgenotypeXage[mean_expr_sexXgenotypeXage$sex == i & mean_expr_sexXgenotypeXage$genotype == j & mean_expr_sexXgenotypeXage$age == k, "sd_expression"]/sqrt(table(pData(dat.filtered)$sex, pData(dat.filtered)$genotype, pData(dat.filtered)$age)[i,j,k])
    }
  }
}

mean_expr_sexXgenotypeXage$gene_short_name <- lookupGeneName(dat.filtered, as.character(mean_expr_sexXgenotypeXage$gene_id))

for(i in seq_along(gene_list)){
  #print(gene_list[i])
  assign(paste0("p", i), 
         ggplot(mean_expr_sexXgenotypeXage[mean_expr_sexXgenotypeXage$gene_short_name == gene_list[i],], aes(x = genotype, y = mean_expression, fill = genotype)) +
          geom_col() + 
          geom_errorbar(aes(ymin=mean_expression-sem_expression, ymax=mean_expression+sem_expression)) +
           facet_wrap(~age+sex) +
          theme(legend.position = "none", plot.title = element_text(hjust= 0.5)) +
           ggtitle(gene_list[i]) +
           scale_fill_brewer(palette = "Set1")
  )
}

Hox_by_sexXgenotypeXage <- plot_grid(p1, blank, blank, p2, 
             p3, p4, blank, blank,
             p5, p6, blank, p7,
             p8, p9, p10, p11,
             p12, p13, p14, blank,
             p15, p16, p17, blank,
             p18, p19, blank, blank,
             blank, p20, p21, p22,
             p23, p24, p25, p26, 
             p27, blank, p28, p29,
             ncol = 4)

save_plot("Hox_genes_by_sexXgenotypeXage.pdf", Hox_by_sexXgenotypeXage, base_height = 60, base_width = 20)
```

# Session Information
```{r session}
sessionInfo()
```