---
title: "HSCR Differential Expression Analysis"
author: "Liz Vincent"
date: "`r format(Sys.time(), '%B %d, %Y')`"
output: pdf_document
---
```{r include=FALSE}
knitr::opts_chunk$set(message=FALSE,warning=FALSE,error=FALSE,fig.align="center")
source("init.R")
```

# Differential gene expression testing
## y ~ genotype
```{r genotype}
genotype_diff_test <- differentialGeneTest(dat.filtered[expressed_genes,], fullModelFormulaStr = "~genotype + batch + total_mRNAs", reducedModelFormulaStr = "~batch + total_mRNAs", verbose = T)

genotype_diff_test_sorted <- genotype_diff_test[order(genotype_diff_test$qval),]

write.csv(genotype_diff_test_sorted[genotype_diff_test_sorted$qval < 0.05, c("gene_short_name", "gene_id", "pval", "qval")], file = "DEG_all_cells_genotype.csv", quote = F, row.names = F)

genotype_diff_test_sig_genes <- rownames(genotype_diff_test_sorted[genotype_diff_test_sorted$qval < 0.05,])

heatmap_data <- log10(exprs(dat.filtered)[genotype_diff_test_sig_genes,] + 1) # log transform
heatmap_data <- t(apply(heatmap_data, 1, function(x){ # Normalize gene expr by dividing by max
  x/max(x)
}))

rownames(heatmap_data) <- lookupGeneName(dat.filtered, rownames(heatmap_data))

heatmap_data <- heatmap_data[order.dendrogram(as.dendrogram(hclust(dist(heatmap_data)))),order.dendrogram(as.dendrogram(hclust(dist(t(heatmap_data)))))]

#heatmap_data[heatmap_data == 0] <- NA

#myLower <- 0
myUpper <- 1

#heatmap_data[heatmap_data < myLower] <- myLower
heatmap_data[heatmap_data > myUpper] <- myUpper

heatmap_annotation <- pData(dat.filtered)[, "genotype", drop = F]
colnames(heatmap_annotation) <- "Genotype"

heatmap_colors <- list(
  "Genotype" = c("het" = "#E41A1C", "hom" = "#377EB8"))

pdf("DEG_all_cells_genotype_heatmap.pdf", height = 20, width = 20)
pheatmap(mat = heatmap_data,
         scale="none",
         show_rownames = TRUE,
         show_colnames = FALSE,
         drop_levels = FALSE,
         cluster_rows = F,
         cluster_cols = F,
         color = viridis(100),
         #breaks=seq(0, myUpper, length = 100),
         #breaks = seq(0, 2.8, length = 200),
         annotation_col = heatmap_annotation,
         annotation_colors = heatmap_colors,
         annotation_names_col = TRUE,
         annotation_legend = TRUE,
         fontsize_row = 5
#         treeheight_col = 0,
#         treeheight_row = 0
         #cutree_rows = 18,
         #cutree_cols = 9
         )
dev.off()

genotype_gene_names <- lookupGeneName(dat.filtered, genotype_diff_test_sorted_sig_genes)

het_means <- apply(exprs(dat.filtered)[genotype_diff_test_sorted_sig_genes,pData(dat.filtered)$genotype == "het"], 1,mean)
hom_means <- apply(exprs(dat.filtered)[genotype_diff_test_sorted_sig_genes,pData(dat.filtered)$genotype == "hom"], 1,mean)

genotype_relative_exprs <- hom_means/het_means

globally_upregulated_in_hom <- names(genotype_relative_exprs[genotype_relative_exprs > 1 | genotype_relative_exprs == "Inf"])

globally_downregulated_in_hom <- names(genotype_relative_exprs[genotype_relative_exprs < 1])

egENSEMBL <- toTable(org.Mm.egENSEMBL)

globally_upregulated_in_hom_entrez <- gsub("\\..*","",globally_upregulated_in_hom)
m<-match(globally_upregulated_in_hom_entrez,egENSEMBL$ensembl_id)
globally_upregulated_in_hom_entrez <- egENSEMBL$gene_id[m]
ego_globally_upregulated <- enrichGO(gene = globally_upregulated_in_hom_entrez,
                OrgDb =       'org.Mm.eg.db',
                ont           = "BP",
                pAdjustMethod = "BH",
                pvalueCutoff  = 1,
                qvalueCutoff  = 0.05, 
                readable      = TRUE)

globally_downregulated_in_hom_entrez <- gsub("\\..*","",globally_downregulated_in_hom)
m<-match(globally_downregulated_in_hom_entrez,egENSEMBL$ensembl_id)
globally_downregulated_in_hom_entrez <- egENSEMBL$gene_id[m]
ego_globally_downregulated <- enrichGO(gene = globally_downregulated_in_hom_entrez,
                OrgDb =       'org.Mm.eg.db',
                ont           = "BP",
                pAdjustMethod = "BH",
                pvalueCutoff  = 1,
                qvalueCutoff  = 0.05, 
                readable      = TRUE)


hom_glia_genes <- c("Vamp8", "Tspan12", "Lmna", "Dse", "Epb41l3", "Irf6", "Afap1l1", "Plp1", "Fabp7", "Timp3", "Mmp2", "Postn", "Slitrk2", "Pls3", "Gpr17", "Deptor", "Aldh1a3", "Fam198b", "Vgll3", "Cnn2", "Myl9", "Anxa2", "Fkbp10", "Adamts1", "Il1rap", "Cdk6", "Nkd2", "Tram2", "Lims2", "Kctd12", "Klk8", "Pcgf5", "Cyr61", "Col4a1", "F3", "Ctgf", "Hapln1", "Moxd1", "Tm4sf1", "Cryab", "Col3a1", "Afap1l2", "Limch1", "Sox5", "Pros1", "Sorbs2", "Abcg2", "Trnau1ap", "Slc44a2", "B2m")

hom_glia_genes <- lookupGeneId(dat.filtered, hom_glia_genes)
hom_glia_genes <- gsub("\\..*","", hom_glia_genes)
m<-match(hom_glia_genes, egENSEMBL$ensembl_id)
hom_glia_genes_entrez <- egENSEMBL$gene_id[m]
ego_hom_glia <- enrichGO(gene = hom_glia_genes_entrez,
                OrgDb =       'org.Mm.eg.db',
                ont           = "BP",
                pAdjustMethod = "BH",
                pvalueCutoff  = 1,
                qvalueCutoff  = 0.05, 
                readable      = TRUE)
```

```{r genotype:sex}
genotype_sex_diff_test <- differentialGeneTest(dat.filtered[expressed_genes,], fullModelFormulaStr = "~genotype*sex + batch + total_mRNAs", reducedModelFormulaStr = "~genotype + sex + batch + total_mRNAs", verbose = T)

genotype_sex_diff_test_sorted <- genotype_sex_diff_test[order(genotype_sex_diff_test$qval),]

write.csv(genotype_sex_diff_test_sorted[genotype_sex_diff_test_sorted$qval < 0.05, c("gene_short_name", "gene_id", "pval", "qval")], file = "DEG_all_cells_genotypeXsex.csv", quote = F, row.names = F)

genotype_sex_diff_test_sig_genes <- rownames(genotype_sex_diff_test_sorted[genotype_sex_diff_test_sorted$qval < 0.05,])

length(genotype_sex_diff_test_sig_genes)

#tmp <- t(scale(t(exprs(dat.filtered[genotype_sex_diff_test_sorted_sig_genes,])),scale=T,center=T))
#heatmap_data <- log(tmp -min(tmp) + 1)

heatmap_data <- log10(exprs(dat.filtered)[genotype_sex_diff_test_sig_genes,] + 1) # log transform
heatmap_data <- t(apply(heatmap_data, 1, function(x){ # Normalize gene expr by dividing by SD
  x/sd(x)
}))


myLower <- 0
myUpper <- 10

heatmap_data[heatmap_data < myLower] <- myLower
heatmap_data[heatmap_data > myUpper] <- myUpper

rownames(heatmap_data) <- lookupGeneName(dat.filtered, rownames(heatmap_data))
heatmap_data <- heatmap_data[order.dendrogram(as.dendrogram(hclust(dist(heatmap_data)))),order.dendrogram(as.dendrogram(hclust(dist(t(heatmap_data)))))]

heatmap_annotation <- pData(dat.filtered)[, c("cluster", "genotype","sex")]

#heatmap_colors <- list(
#"Cell Type" =  c("Acetylcholinergic Neuron" = "#E41A1C", "Neuron" = "#377EB8", "Noradrenergic Neuron" = "#4DAF4A", #"Progenitor/Glia" = "#984EA3", "Transition Cell" = "#FF7F00"),
#"Genotype" = c("het" = "black", "hom" = "gray70"),
#"Sex" = c("female"="black","male"="gray70"),
#"Age" = c("E12.5" = "#7FC97F", "E14.5" = "#BEAED4"))

colnames(heatmap_annotation) <- c("Cluster","Genotype","Sex")

pdf("heatmap_DEG_genotype-sex.pdf", height = 20, width = 20)
pheatmap(mat = heatmap_data,
         scale="none",
         show_rownames = TRUE,
         show_colnames = FALSE,
         drop_levels = FALSE,
         cluster_rows = F,
         cluster_cols = F,
         color = viridis(100),
         annotation_col = heatmap_annotation,
         #annotation_colors = heatmap_colors,
         annotation_names_col = TRUE,
         annotation_legend = TRUE,
         fontsize_row = 5,
         treeheight_col = 0,
         treeheight_row = 0
         )
dev.off()

genotype_sex_gene_names <- lookupGeneName(dat.filtered, genotype_sex_diff_test_sorted_sig_genes)

het_means2 <- apply(exprs(dat.filtered)[genotype_sex_diff_test_sorted_sig_genes,pData(dat.filtered)$genotype == "het"], 1,mean)
hom_means2 <- apply(exprs(dat.filtered)[genotype_sex_diff_test_sorted_sig_genes,pData(dat.filtered)$genotype == "hom"], 1,mean)

genotype_relative_exprs2 <- hom_means2/het_means2

globally_upregulated_in_hom2 <- names(genotype_relative_exprs2[genotype_relative_exprs2 > 1 | genotype_relative_exprs2 == "Inf"])

globally_downregulated_in_hom2 <- names(genotype_relative_exprs2[genotype_relative_exprs2 < 1])

egENSEMBL <- toTable(org.Mm.egENSEMBL)

globally_upregulated_in_hom_entrez2 <- gsub("\\..*","",globally_upregulated_in_hom2)
m<-match(globally_upregulated_in_hom_entrez2,egENSEMBL$ensembl_id)
globally_upregulated_in_hom_entrez2 <- egENSEMBL$gene_id[m]
ego_globally_upregulated2 <- enrichGO(gene = globally_upregulated_in_hom_entrez2,
                OrgDb =       'org.Mm.eg.db',
                ont           = "BP",
                pAdjustMethod = "BH",
                pvalueCutoff  = 1,
                qvalueCutoff  = 0.05, 
                readable      = TRUE)

globally_downregulated_in_hom_entrez2 <- gsub("\\..*","",globally_downregulated_in_hom2)
m<-match(globally_downregulated_in_hom_entrez2,egENSEMBL$ensembl_id)
globally_downregulated_in_hom_entrez2 <- egENSEMBL$gene_id[m]
ego_globally_downregulated2 <- enrichGO(gene = globally_downregulated_in_hom_entrez2,
                OrgDb =       'org.Mm.eg.db',
                ont           = "BP",
                pAdjustMethod = "BH",
                pvalueCutoff  = 1,
                qvalueCutoff  = 0.05, 
                readable      = TRUE)
```

```{r genotype:celltype}
genotypeCellType_diff_test <- differentialGeneTest(dat.filtered[expressed_genes,], fullModelFormulaStr = "~genotype*celltype + batch + total_mRNAs", reducedModelFormulaStr = "~genotype + celltype + batch + total_mRNAs", cores = 1, verbose = T)

genotypeCellType_diff_test_sorted <- genotypeCellType_diff_test[order(genotypeCellType_diff_test$qval),]

write.csv(genotypeCellType_diff_test_sorted[genotypeCellType_diff_test_sorted$qval < 0.05, c("gene_short_name", "gene_id", "pval", "qval")], file = "DEG_all_cells_genotypeXcelltype.csv", quote = F, row.names = F)

genotypeCellType_diff_test_sig_genes<-rownames(genotypeCellType_diff_test_sorted[genotypeCellType_diff_test_sorted$qval < 0.05,])

length(genotypeCellType_diff_test_sig_genes)

heatmap_data <- log10(exprs(dat.filtered)[genotypeCellType_diff_test_sig_genes,] + 1) # log transform
heatmap_data <- t(apply(heatmap_data, 1, function(x){ # Normalize gene expr by dividing by SD
  x/sd(x)
}))
#tmp <- t(scale(t(exprs(dat.filtered[genotypeCellType_diff_test_sorted_sig_genes,])),scale = T,center = T))
#heatmap_data <- log(tmp -min(tmp) + 1)

myLower <- 0
myUpper <- 6

heatmap_data[heatmap_data < myLower] <- myLower
heatmap_data[heatmap_data > myUpper] <- myUpper

rownames(heatmap_data) <- lookupGeneName(dat.filtered,rownames(heatmap_data))
heatmap_data <- heatmap_data[order.dendrogram(as.dendrogram(hclust(dist(heatmap_data)))),order.dendrogram(as.dendrogram(hclust(dist(t(heatmap_data)))))]

heatmap_annotation <- pData(dat.filtered)[,c("celltype","genotype")]

heatmap_colors <- list(
"Cell Type" =  c("Acetylcholinergic Neuron" = "#E41A1C", "Neuron" = "#377EB8", "Noradrenergic Neuron" = "#4DAF4A", "Progenitor/Glia" = "#984EA3", "Transition Cell" = "#FF7F00"),
"Genotype" = c("het" = "black", "hom" = "gray70"),
"Age" = c("E12.5" = "#7FC97F", "E14.5" = "#BEAED4"))

colnames(heatmap_annotation) <- c("Cell Type","Genotype")

pdf("heatmap_DEG_genotype-celltype.pdf", height = 20, width = 20)
pheatmap(mat = heatmap_data,
         scale="none",
         show_rownames = TRUE,
         show_colnames = FALSE,
         drop_levels = FALSE,
         cluster_rows = F,
         cluster_cols = F,
         color = viridis(100),
         annotation_col = heatmap_annotation,
         #annotation_colors = heatmap_colors,
         annotation_names_col = TRUE,
         annotation_legend = TRUE,
         fontsize_row = 5,
         treeheight_col = 0,
         treeheight_row = 0
         )
dev.off()
```

```{r neuron_bifurcation_diff_test}
neuron_bifuracation_difftest <- differentialGeneTest(dat.filtered[expressed_genes,pData(dat.filtered)$cluster %in% 7:8], fullModelFormulaStr = "~cluster + batch + total_mRNAs", reducedModelFormulaStr = "~batch + total_mRNAs", cores = 1, verbose = T)

neuron_bifuracation_difftest_sorted <- neuron_bifuracation_difftest[order(neuron_bifuracation_difftest$qval),]

write.csv(neuron_bifuracation_difftest_sorted[neuron_bifuracation_difftest_sorted$qval < 0.05, c("gene_short_name", "gene_id", "pval", "qval")], file = "DEG_clusters7&8_cluster.csv", quote = F, row.names = F)

neuron_bifuracation_difftest_sig_genes <- rownames(neuron_bifuracation_difftest_sorted[neuron_bifuracation_difftest_sorted$qval < 0.05,])

bifurcation_gene_names <- lookupGeneName(dat.filtered, neuron_bifuracation_difftest_sig_genes)

# het-only = 7, mixed pop = 8

pop7_means <- apply(exprs(dat.filtered)[neuron_bifuracation_difftest_sig_genes, pData(dat.filtered)$cluster == 7], 1,mean)
pop8_means <- apply(exprs(dat.filtered)[neuron_bifuracation_difftest_sig_genes, pData(dat.filtered)$cluster == 8], 1,mean)

relative_exprs <- pop7_means/pop8_means

upregulated_in_het <- names(relative_exprs[relative_exprs > 1])

downregulated_in_het <- names(relative_exprs[relative_exprs < 1])

egENSEMBL <- toTable(org.Mm.egENSEMBL)

bifurcation_entrez <- gsub("\\..*","",neuron_bifuracation_difftest_sig_genes)
m<-match(bifurcation_entrez,egENSEMBL$ensembl_id)
bifurcation_entrez <- egENSEMBL$gene_id[m]
ego_bifurcation <- enrichGO(gene = bifurcation_entrez,
                OrgDb =       'org.Mm.eg.db',
                ont           = "BP",
                pAdjustMethod = "BH",
                pvalueCutoff  = 1,
                qvalueCutoff  = 0.05, 
                readable      = TRUE)

upregulated_in_het_entrez <- gsub("\\..*","",upregulated_in_het)
m<-match(upregulated_in_het_entrez,egENSEMBL$ensembl_id)
upregulated_in_het_entrez <- egENSEMBL$gene_id[m]
ego_upregulated <- enrichGO(gene = upregulated_in_het_entrez,
                OrgDb =       'org.Mm.eg.db',
                ont           = "BP",
                pAdjustMethod = "BH",
                pvalueCutoff  = 1,
                qvalueCutoff  = 0.05, 
                readable      = TRUE)

downregulated_in_het_entrez <- gsub("\\..*","",downregulated_in_het)
m<-match(downregulated_in_het_entrez,egENSEMBL$ensembl_id)
downregulated_in_het_entrez <- egENSEMBL$gene_id[m]
ego_downregulated <- enrichGO(gene = downregulated_in_het_entrez,
                OrgDb =       'org.Mm.eg.db',
                ont           = "BP",
                pAdjustMethod = "BH",
                pvalueCutoff  = 1,
                qvalueCutoff  = 0.05, 
                readable      = TRUE)


#tmp <- t(scale(t(exprs(dat.filtered[neuron_bifuracation_difftest_sorted_sig_genes, pData(dat.filtered)$cluster %in% 7:8])), scale = T, center = T))
#heatmap_data <- log10(tmp - min(tmp) + 1)

heatmap_data <- log10(exprs(dat.filtered)[neuron_bifuracation_difftest_sig_genes, pData(dat.filtered)$cluster %in% 7:8] + 1) # log transform
heatmap_data <- t(apply(heatmap_data, 1, function(x){ # Normalize gene expr by dividing by SD
  x/max(x)
}))

rownames(heatmap_data) <- lookupGeneName(dat.filtered, rownames(heatmap_data))

heatmap_data <- heatmap_data[order.dendrogram(as.dendrogram(hclust(dist(heatmap_data)))),order.dendrogram(as.dendrogram(hclust(dist(t(heatmap_data)))))]

#myLower <- 0.1
#myUpper <- 0.9

#heatmap_data[heatmap_data < myLower] <- myLower
#heatmap_data[heatmap_data > myUpper] <- myUpper

heatmap_annotation <- pData(dat.filtered)[pData(dat.filtered)$cluster %in% 7:8, c("cluster", "genotype")]
heatmap_annotation$cluster <- droplevels(heatmap_annotation$cluster)

heatmap_colors <- list(
  "Cluster" = c("7" = brewer.pal(8, "Set2")[7],
                "8" = brewer.pal(8, "Set2")[8]),
  "Genotype" = c("het" = brewer.pal(3, "Set1")[1], 
                 "hom" = brewer.pal(3, "Set1")[2]))

colnames(heatmap_annotation)<-c("Cluster","Genotype")

pdf("heatmap_DEG_cluster7vs8_clustered.pdf", height = 10, width = 14)
pheatmap(mat = heatmap_data,
         scale="none",
         show_rownames = TRUE,
         show_colnames = FALSE,
         drop_levels = FALSE,
         cluster_rows = T,
         cluster_cols = T,
         color = viridis(100),
         annotation_col = heatmap_annotation,
         annotation_colors = heatmap_colors,
         annotation_names_col = TRUE,
         annotation_legend = TRUE,
         fontsize_row = 5,
         treeheight_col = 0,
         treeheight_row = 20,
         cutree_rows = 10
         )
dev.off()

```

```{r Ret_independent_genotype_diff}
Ret_independent_difftest <- differentialGeneTest(dat.filtered[expressed_genes,pData(dat.filtered)$cluster == 8], fullModelFormulaStr = "~genotype + batch + total_mRNAs", reducedModelFormulaStr = "~batch + total_mRNAs", cores = 1, verbose = T)

Ret_independent_difftest_sorted <- Ret_independent_difftest[order(Ret_independent_difftest$qval),]

write.csv(Ret_independent_difftest_sorted[Ret_independent_difftest_sorted$qval < 0.05, c("gene_short_name", "gene_id", "pval", "qval")], file = "DEG_cluster8_genotype.csv", quote = F, row.names = F)

Ret_independent_difftest_sig_genes <- rownames(Ret_independent_difftest_sorted[Ret_independent_difftest_sorted$qval < 0.05,])

het_mean <- apply(exprs(dat.filtered[Ret_independent_difftest_sig_genes, pData(dat.filtered)$cluster == 8 & pData(dat.filtered)$genotype == "het"]), 1, mean)

hom_mean <- apply(exprs(dat.filtered[Ret_independent_difftest_sig_genes, pData(dat.filtered)$cluster == 8 & pData(dat.filtered)$genotype == "hom"]), 1, mean)


lookupGeneName(dat.filtered, names(which(het_mean/hom_mean < 1)))
# "Tac1" "Pcp4" "Ddc" "Th" "A730017C20Rik" "Rgs4" "Agtr2" "Spock1"  higher in null

lookupGeneName(dat.filtered, names(which(het_mean/hom_mean > 1)))
#"Gm27747" "Akr1e1"  "Gm27553" "Hoxb5" higher in het

Ret_independent_gene_names <- lookupGeneName(dat.filtered, Ret_independent_difftest_sig_genes)


#tmp <- t(scale(t(exprs(dat.filtered[Ret_independent_difftest_sorted_sig_genes, pData(dat.filtered)$cluster %in% c(4, 6)])), scale = T, center = T))
#heatmap_data <- log10(tmp - min(tmp) + 1)

heatmap_data <- log10(exprs(dat.filtered)[Ret_independent_difftest_sig_genes, pData(dat.filtered)$cluster == 8] + 1) # log transform
heatmap_data <- t(apply(heatmap_data, 1, function(x){ # Normalize gene expr by dividing by max
  x/max(x)
}))

rownames(heatmap_data) <- lookupGeneName(dat.filtered, rownames(heatmap_data))

heatmap_data <- heatmap_data[order.dendrogram(as.dendrogram(hclust(dist(heatmap_data)))),order.dendrogram(as.dendrogram(hclust(dist(t(heatmap_data)))))]

#myLower <- 0.1
#myUpper <- 0.9

#heatmap_data[heatmap_data < myLower] <- myLower
#heatmap_data[heatmap_data > myUpper] <- myUpper

heatmap_annotation <- pData(dat.filtered)[pData(dat.filtered)$cluster == 8, c("genotype"), drop = F]


heatmap_colors <- list(
  "Genotype" = c("het" = brewer.pal(3, "Set1")[1], 
                 "hom" = brewer.pal(3, "Set1")[2]))

colnames(heatmap_annotation)<-c("Genotype")

pdf("heatmap_DEG_genotype_cluster8.pdf", height = 3, width = 8)
pheatmap(mat = heatmap_data,
         scale="none",
         show_rownames = TRUE,
         show_colnames = FALSE,
         drop_levels = FALSE,
         cluster_rows = F,
         cluster_cols = F,
         color = viridis(100),
         annotation_col = heatmap_annotation,
         annotation_colors = heatmap_colors,
         annotation_names_col = TRUE,
         annotation_legend = TRUE,
         fontsize_row = 7,
         treeheight_col = 0,
         treeheight_row = 0
         )
dev.off()

egENSEMBL <- toTable(org.Mm.egENSEMBL)

Ret_independent_entrez <- gsub("\\..*","",Ret_independent_difftest_sig_genes)
m<-match(Ret_independent_entrez,egENSEMBL$ensembl_id)
Ret_independent_entrez <- egENSEMBL$gene_id[m]
ego_Ret_independent <- enrichGO(gene = Ret_independent_entrez,
                OrgDb =       'org.Mm.eg.db',
                ont           = "BP",
                pAdjustMethod = "BH",
                pvalueCutoff  = 1,
                qvalueCutoff  = 0.05, 
                readable      = TRUE)
```

```{r genotype_glia}
glia_genotype_diff_test <- differentialGeneTest(dat.filtered[expressed_genes, pData(dat.filtered)$celltype == "progenitor/glia" & !pData(dat.filtered)$cluster == 9], fullModelFormulaStr = "~genotype + batch + total_mRNAs", reducedModelFormulaStr = "~batch + total_mRNAs", verbose = T)

glia_genotype_diff_test_sorted <- glia_genotype_diff_test[order(glia_genotype_diff_test$qval),]

write.csv(glia_genotype_diff_test_sorted[glia_genotype_diff_test_sorted$qval < 0.05, c("gene_short_name", "gene_id", "pval", "qval")], file = "DEG_glia!cluster9_genotype.csv", quote = F, row.names = F)

genotype_diff_test_sig_genes <- rownames(genotype_diff_test_sorted[genotype_diff_test_sorted$qval < 0.05,])

```


# Cluster Specificity
```{r specificity}
melted_df <- melt(exprs(dat.filtered[expressed_genes,]))
colnames(melted_df) <- c("gene_id","cell_id","value")
melted_df <- merge(melted_df, pData(dat.filtered)[, c("genotype", "sex", "age", "cluster")], by.x = "cell_id", by.y = 0)

mean_expr_cluster <- melted_df %>%
  group_by(cluster, gene_id) %>%
  summarise(mean_expression_cluster = mean(value)) %>%
  as.data.frame()

mean_expr_cluster <- dcast(mean_expr_cluster, gene_id ~ cluster)
rownames(mean_expr_cluster) <- mean_expr_cluster$gene_id
mean_expr_cluster <- mean_expr_cluster[,-1]

#Data frame of mean expression for each gene in each cluster

cluster_specificity <- cummeRbund:::.specificity(mean_expr_cluster)

melted_cluster_specificity <- melt(cluster_specificity)
mean_expr_cluster_melted <- melt(mean_expr_cluster)
names(mean_expr_cluster_melted) <- c("cluster","mean_expr")
mean_expr_cluster_melted$gene_id <- rep(row.names(mean_expr_cluster), nlevels(pData(dat.filtered)$cluster))
melted_cluster_specificity$Var2 <- as.numeric(str_replace(melted_cluster_specificity$Var2, pattern = "_spec", replacement = ""))
names(melted_cluster_specificity) <- c("gene_id", "cluster", "specificity")
melted_df <- merge(melted_cluster_specificity, mean_expr_cluster_melted, by = c("gene_id", "cluster"))
melted_df$gene_short_name <- lookupGeneName(dat.filtered, as.character(melted_df$gene_id), unique = F)

index <- apply(cluster_specificity, 1, which.max)
gene_specificity <- list()
for(i in 1:nlevels(pData(dat.filtered)$cluster)){
  gene_specificity[[i]] <- lookupGeneName(dat.filtered, names(index[index == i]))
}

apply(cluster_specificity, 2, function(x){
  lookupGeneName(dat.filtered, names(x)[order(x)[1:10]])
})


plot_specificity <- function(x){
  ggplot(melted_df[melted_df$cluster == x,], aes(x = log10(mean_expr + 1), 
                                        y = specificity, 
                                        label = ifelse(melted_df[melted_df$cluster == x,]$gene_short_name %in% gene_specificity[[x]],
                                                       melted_df[melted_df$cluster == x,]$gene_short_name, 
                                                       ""), 
                                        color = ifelse(ifelse(melted_df[melted_df$cluster == x,]$gene_short_name %in% gene_specificity[[x]],
                                                       melted_df[melted_df$cluster == x,]$gene_short_name, 
                                                       "") == "", "no", "yes"), 
                                        #alpha = ifelse(ifelse(melted_df[melted_df$cluster == x,]$gene_short_name %in% genes[[x]],
                                        #               melted_df[melted_df$cluster == x,]$gene_short_name, 
                                        #               "") == "", 0.1, 1)
                                        )) +
#  ggtitle(paste("Cluster", x)) +
      geom_point(size = 0.5) +
      #geom_smooth(lwd = 0.2, color = "red") +
      scale_color_manual(values = c("no" = "black", "yes" = "red")) +
      scale_x_continuous("log10(mean expression + 1)", expand = c(0,0), limits = c(0, 3.5)) +
      scale_y_continuous("Specificity", expand = c(0,0), limits = c(0,1)) +
      geom_text(size = 1.4, 
                hjust = 0, 
                vjust = 0, 
                nudge_x = 0.03, 
                nudge_y = -0.005, 
                color = "black") +
      theme(legend.position = "none",
            plot.title = element_text(size = 10, hjust = 0.005, vjust = -6, face = "plain"),
            plot.margin = margin(-10, 2, 2, 2),
            axis.title = element_text(size= 8),
            axis.title.x = element_text(margin = margin(t = 0)),
            axis.title.y = element_text(margin = margin(r = 0)),
            axis.text = element_text(size = 6),
            #axis.title = element_blank(),
            #axis.text = element_blank(),
            aspect.ratio = 1
            )
}


plot_specificity(4) + ggtitle("Cluster 4: Ache+")
plot_specificity(5) + ggtitle("Cluster 5: Nos1+")
plot_specificity(7) + ggtitle("Cluster 7: Maturing Glia")
plot_specificity(8) + ggtitle("Cluster 8: Th+, Dbh+")
p2 <- plot_specificity(2)
p3 <- plot_specificity(3)
p4 <- plot_specificity(4)
p5 <- plot_specificity(5)
p6 <- plot_specificity(6)
p7 <- plot_specificity(7)
p8 <- plot_specificity(8)

pg1 <- cowplot::plot_grid(p1, p3, p5, p7, rel_widths = 1, ncol = 1)
pg2 <- cowplot::plot_grid(p2, p4, p6, p8, rel_widths = 1, ncol = 1)

cowplot::plot_grid(pg1, pg2, ncol = 2)      

grid.arrange(p1, p2, p3, p4, p5, p6, p7, p8, ncol = 2)

```