---
title: "HSCR Differential Expression Analysis"
author: "Liz Vincent"
date: "`r format(Sys.time(), '%B %d, %Y')`"
output: pdf_document
---
```{r include=FALSE}
knitr::opts_chunk$set(message=FALSE,warning=FALSE,error=FALSE,fig.align="center")
source("init.R")
```

# Differential gene expression testing
## y ~ genotype
```{r genotype}
genotype_diff_test <- differentialGeneTest(dat.filtered[expressed_genes,], fullModelFormulaStr = "~batch+sex+CellType+genotype", reducedModelFormulaStr = "~batch+sex+CellType", verbose = T)
#include total_mRNAs

genotype_diff_test_sorted <- genotype_diff_test[order(genotype_diff_test$qval),]

genotype_diff_test_sorted_sig_genes <- rownames(genotype_diff_test_sorted[genotype_diff_test_sorted$qval < 0.01,])

tmp <- t(scale(t(exprs(dat.filtered[genotype_diff_test_sorted_sig_genes,])), scale = T, center = T))
heatmap_data <- log10(tmp - min(tmp) + 1)

rownames(heatmap_data) <- lookupGeneName(dat.filtered, rownames(heatmap_data))

heatmap_data <- heatmap_data[order.dendrogram(as.dendrogram(hclust(dist(heatmap_data)))),order.dendrogram(as.dendrogram(hclust(dist(t(heatmap_data)))))]

myLower <- 0.1
myUpper <- 0.9

heatmap_data[heatmap_data < myLower] <- myLower
heatmap_data[heatmap_data > myUpper] <- myUpper

heatmap_annotation <- pData(dat.filtered)[, c("CellType", "genotype", "age")]

heatmap_colors <- list(
  "Cell Type" =  c("Acetylcholinergic Neuron" = "#E41A1C", 
                   "Neuron" = "#377EB8",
                   "Noradrenergic Neuron" = "#4DAF4A", 
                   "Progenitor/Glia" = "#984EA3",
                   "Transition Cell" = "#FF7F00"),
  "Genotype" = c("het" = "black", "hom" = "gray70"),
  "Age" = c("E12.5" = "#7FC97F", "E14.5" = "#BEAED4"))

colnames(heatmap_annotation)<-c("Cell Type","Genotype","Age")

#pdf("Heatmap_all_cells_wrt_genotype.pdf",height=25,width=18)
#png("Heatmap_all_cells_wrt_genotype.png",height=15,width=20, units = "in", res = 400)
pheatmap(mat = heatmap_data,
         scale="none",
         show_rownames = TRUE,
         show_colnames = FALSE,
         drop_levels = FALSE,
         cluster_rows = F,
         cluster_cols = F,
         color = colorRampPalette(colors = c("#377EB8", "white", "#E41A1C"))(100),
         breaks=seq(myLower, myUpper, length = 100),
         #breaks = seq(0, 2.8, length = 200),
         annotation_col = heatmap_annotation,
         annotation_colors = heatmap_colors,
         annotation_names_col = TRUE,
         annotation_legend = TRUE,
         fontsize_row = 5,
         treeheight_col = 0,
         treeheight_row = 0,
         main = "All Cells\n254 Genes, 1% FDR\nFull model: ~ genotype + batch  + sex + cell type\nReduced model: ~ batch + sex + cell type"
         )
#dev.off()

genotype_gene_names <- lookupGeneName(dat.filtered, genotype_diff_test_sorted_sig_genes)

het_means <- apply(exprs(dat.filtered)[genotype_diff_test_sorted_sig_genes,pData(dat.filtered)$genotype == "het"], 1,mean)
hom_means <- apply(exprs(dat.filtered)[genotype_diff_test_sorted_sig_genes,pData(dat.filtered)$genotype == "hom"], 1,mean)

genotype_relative_exprs <- hom_means/het_means

globally_upregulated_in_hom <- names(genotype_relative_exprs[genotype_relative_exprs > 1 | genotype_relative_exprs == "Inf"])

globally_downregulated_in_hom <- names(genotype_relative_exprs[genotype_relative_exprs < 1])

egENSEMBL <- toTable(org.Mm.egENSEMBL)

globally_upregulated_in_hom_entrez <- gsub("\\..*","",globally_upregulated_in_hom)
m<-match(globally_upregulated_in_hom_entrez,egENSEMBL$ensembl_id)
globally_upregulated_in_hom_entrez <- egENSEMBL$gene_id[m]
ego_globally_upregulated <- enrichGO(gene = globally_upregulated_in_hom_entrez,
                OrgDb =       'org.Mm.eg.db',
                ont           = "BP",
                pAdjustMethod = "BH",
                pvalueCutoff  = 1,
                qvalueCutoff  = 0.01, 
                readable      = TRUE)

globally_downregulated_in_hom_entrez <- gsub("\\..*","",globally_downregulated_in_hom)
m<-match(globally_downregulated_in_hom_entrez,egENSEMBL$ensembl_id)
globally_downregulated_in_hom_entrez <- egENSEMBL$gene_id[m]
ego_globally_downregulated <- enrichGO(gene = globally_downregulated_in_hom_entrez,
                OrgDb =       'org.Mm.eg.db',
                ont           = "BP",
                pAdjustMethod = "BH",
                pvalueCutoff  = 1,
                qvalueCutoff  = 0.01, 
                readable      = TRUE)
```

```{r genotype:sex}
genotype_sex_diff_test <- differentialGeneTest(dat.filtered[expressed_genes,], fullModelFormulaStr = "~genotype*sex + batch + CellType", reducedModelFormulaStr = "~genotype + sex + batch + CellType", verbose = T)

genotype_sex_diff_test_sorted <- genotype_sex_diff_test[order(genotype_sex_diff_test$qval),]

genotype_sex_diff_test_sorted_sig_genes <- rownames(genotype_sex_diff_test_sorted[genotype_sex_diff_test_sorted$qval < 0.01,])

length(genotype_sex_diff_test_sorted_sig_genes)

tmp <- t(scale(t(exprs(dat.filtered[genotype_sex_diff_test_sorted_sig_genes,])),scale=T,center=T))
heatmap_data <- log(tmp -min(tmp) + 1)

myLower <- 0
myUpper <- 2.8

heatmap_data[heatmap_data < myLower] <- myLower
heatmap_data[heatmap_data > myUpper] <- myUpper

rownames(heatmap_data) <- lookupGeneName(dat.filtered, rownames(heatmap_data))

heatmap_annotation <- pData(dat.filtered)[, c("CellType","genotype","sex", "age")]

heatmap_colors <- list(
"Cell Type" =  c("Acetylcholinergic Neuron" = "#E41A1C", "Neuron" = "#377EB8", "Noradrenergic Neuron" = "#4DAF4A", "Progenitor/Glia" = "#984EA3", "Transition Cell" = "#FF7F00"),
"Genotype" = c("het" = "black", "hom" = "gray70"),
"Sex" = c("female"="black","male"="gray70"),
"Age" = c("E12.5" = "#7FC97F", "E14.5" = "#BEAED4"))

colnames(heatmap_annotation) <- c("Cell Type","Genotype","Sex", "Age")

pheatmap(mat = heatmap_data,
         scale = "none",
         show_rownames = TRUE,
         show_colnames = FALSE,
         drop_levels = FALSE,
         breaks = seq(myLower, myUpper, length = 100),
         annotation_col = heatmap_annotation,
         annotation_colors = heatmap_colors,
         annotation_names_col = TRUE,
         annotation_legend = TRUE,
         fontsize_row = 8,
         #cutree_rows = 15,
         #cutree_col = 25,
         main = "All Cells\n30 Genes, 1% FDR\nFull model: ~ genotype*sex + batch  + cell type\nReduced model: ~ genotype + sex + batch + cell type"
         )

genotype_sex_gene_names <- lookupGeneName(dat.filtered, genotype_sex_diff_test_sorted_sig_genes)

het_means2 <- apply(exprs(dat.filtered)[genotype_sex_diff_test_sorted_sig_genes,pData(dat.filtered)$genotype == "het"], 1,mean)
hom_means2 <- apply(exprs(dat.filtered)[genotype_sex_diff_test_sorted_sig_genes,pData(dat.filtered)$genotype == "hom"], 1,mean)

genotype_relative_exprs2 <- hom_means2/het_means2

globally_upregulated_in_hom2 <- names(genotype_relative_exprs2[genotype_relative_exprs2 > 1 | genotype_relative_exprs2 == "Inf"])

globally_downregulated_in_hom2 <- names(genotype_relative_exprs2[genotype_relative_exprs2 < 1])

egENSEMBL <- toTable(org.Mm.egENSEMBL)

globally_upregulated_in_hom_entrez2 <- gsub("\\..*","",globally_upregulated_in_hom2)
m<-match(globally_upregulated_in_hom_entrez2,egENSEMBL$ensembl_id)
globally_upregulated_in_hom_entrez2 <- egENSEMBL$gene_id[m]
ego_globally_upregulated2 <- enrichGO(gene = globally_upregulated_in_hom_entrez2,
                OrgDb =       'org.Mm.eg.db',
                ont           = "BP",
                pAdjustMethod = "BH",
                pvalueCutoff  = 1,
                qvalueCutoff  = 0.01, 
                readable      = TRUE)

globally_downregulated_in_hom_entrez2 <- gsub("\\..*","",globally_downregulated_in_hom2)
m<-match(globally_downregulated_in_hom_entrez2,egENSEMBL$ensembl_id)
globally_downregulated_in_hom_entrez2 <- egENSEMBL$gene_id[m]
ego_globally_downregulated2 <- enrichGO(gene = globally_downregulated_in_hom_entrez2,
                OrgDb =       'org.Mm.eg.db',
                ont           = "BP",
                pAdjustMethod = "BH",
                pvalueCutoff  = 1,
                qvalueCutoff  = 0.01, 
                readable      = TRUE)
```

```{r genotype:celltype}
genotypeCellType_diff_test <- differentialGeneTest(dat.filtered[expressed_genes,], fullModelFormulaStr = "~batch + sex + genotype*CellType", reducedModelFormulaStr = "~batch + sex + genotype + CellType", cores = 1, verbose = T)

genotypeCellType_diff_test_sorted<-genotypeCellType_diff_test[order(genotypeCellType_diff_test$qval),]

genotypeCellType_diff_test_sorted_sig_genes<-rownames(genotypeCellType_diff_test_sorted[genotypeCellType_diff_test_sorted$qval < 0.01,])

length(genotypeCellType_diff_test_sorted_sig_genes)

tmp <- t(scale(t(exprs(dat.filtered[genotypeCellType_diff_test_sorted_sig_genes,])),scale = T,center = T))
heatmap_data <- log(tmp -min(tmp) + 1)

myLower <- 0
myUpper <- 2.5

heatmap_data[heatmap_data < myLower] <- myLower
heatmap_data[heatmap_data > myUpper] <- myUpper

rownames(heatmap_data) <- lookupGeneName(dat.filtered,rownames(heatmap_data))

heatmap_annotation <- pData(dat.filtered)[,c("CellType","genotype","age")]

heatmap_colors <- list(
"Cell Type" =  c("Acetylcholinergic Neuron" = "#E41A1C", "Neuron" = "#377EB8", "Noradrenergic Neuron" = "#4DAF4A", "Progenitor/Glia" = "#984EA3", "Transition Cell" = "#FF7F00"),
"Genotype" = c("het" = "black", "hom" = "gray70"),
"Age" = c("E12.5" = "#7FC97F", "E14.5" = "#BEAED4"))

colnames(heatmap_annotation) <- c("Cell Type","Genotype","Age")

pheatmap(mat = heatmap_data,
         scale = "none",
         show_rownames = TRUE,
         show_colnames = FALSE,
         drop_levels = FALSE,
         breaks = seq(myLower, myUpper, length = 100),
         annotation_col = heatmap_annotation,
         annotation_colors = heatmap_colors,
         annotation_names_col = TRUE,
         annotation_legend = TRUE,
         fontsize_row = 8,
         main = "All Cells\n79 Genes, 1% FDR\nFull model: ~ genotype*CellType + batch + sex\nReduced model: ~ genotype + CellType + batch + sex"
         )

genotypeCellType_diff_test_sorted_sig_genes_entrez <- gsub("\\..*","",genotypeCellType_diff_test_sorted_sig_genes)
m <- match(genotypeCellType_diff_test_sorted_sig_genes_entrez,egENSEMBL$ensembl_id)
genotypeCellType_diff_test_sorted_sig_genes_entrez <- egENSEMBL$gene_id[m]
ego_genotype_celltype <- enrichGO(gene = genotypeCellType_diff_test_sorted_sig_genes_entrez,
                OrgDb         = 'org.Mm.eg.db',
                ont           = "BP",
                pAdjustMethod = "BH",
                pvalueCutoff  = 0.05,
                qvalueCutoff  = 1, 
                readable      = TRUE)


#By celltype

ache_hom2het <- apply(exprs(dat.filtered)[genotypeCellType_diff_test_sorted_sig_genes,pData(dat.filtered)$genotype == "hom" & pData(dat.filtered)$CellType == "Acetylcholinergic Neuron"], 1,mean)/apply(exprs(dat.filtered)[genotypeCellType_diff_test_sorted_sig_genes,pData(dat.filtered)$genotype == "het" & pData(dat.filtered)$CellType == "Acetylcholinergic Neuron"], 1,mean)

ache_hom2het_up <- names(ache_hom2het[ache_hom2het > 1 | ache_hom2het == "Inf"])
ache_hom2het_down <- names(ache_hom2het[ache_hom2het < 1 | ache_hom2het == "Nan"])

ache_hom2het_up_entrez <- gsub("\\..*","",ache_hom2het_up)
m <- match(ache_hom2het_up_entrez,egENSEMBL$ensembl_id)
ache_hom2het_up_entrez <- egENSEMBL$gene_id[m]
ego_ache_upregulated <- enrichGO(gene = ache_hom2het_up_entrez,
                OrgDb =       'org.Mm.eg.db',
                ont           = "BP",
                pAdjustMethod = "BH",
                pvalueCutoff  = 0.05,
                qvalueCutoff  = 1, 
                readable      = TRUE)

ache_hom2het_down_entrez <- gsub("\\..*","",ache_hom2het_down)
m <- match(ache_hom2het_down_entrez,egENSEMBL$ensembl_id)
ache_hom2het_down_entrez <- egENSEMBL$gene_id[m]
ego_ache_downregulated <- enrichGO(gene = ache_hom2het_down_entrez,
                OrgDb =       'org.Mm.eg.db',
                ont           = "BP",
                pAdjustMethod = "BH",
                pvalueCutoff  = 0.05,
                qvalueCutoff  = 1, 
                readable      = TRUE)




neuron_hom2het <- apply(exprs(dat.filtered)[genotypeCellType_diff_test_sorted_sig_genes,pData(dat.filtered)$genotype == "hom" & pData(dat.filtered)$CellType == "Neuron"], 1,mean)/apply(exprs(dat.filtered)[genotypeCellType_diff_test_sorted_sig_genes,pData(dat.filtered)$genotype == "het" & pData(dat.filtered)$CellType == "Neuron"], 1,mean)

neuron_hom2het_up <- names(neuron_hom2het[neuron_hom2het > 1 | neuron_hom2het == "Inf"])
neuron_hom2het_down <- names(neuron_hom2het[neuron_hom2het < 1 | neuron_hom2het == "Nan"])

neuron_hom2het_up_entrez <- gsub("\\..*","",neuron_hom2het_up)
m<-match(neuron_hom2het_up_entrez,egENSEMBL$ensembl_id)
neuron_hom2het_up_entrez <- egENSEMBL$gene_id[m]
ego_neuron_upregulated <- enrichGO(gene = neuron_hom2het_up_entrez,
                OrgDb =       'org.Mm.eg.db',
                ont           = "BP",
                pAdjustMethod = "BH",
                pvalueCutoff  = 0.05,
                qvalueCutoff  = 1, 
                readable      = TRUE)

neuron_hom2het_down_entrez <- gsub("\\..*","",neuron_hom2het_down)
m<-match(neuron_hom2het_down_entrez,egENSEMBL$ensembl_id)
neuron_hom2het_down_entrez <- egENSEMBL$gene_id[m]
ego_neuron_downregulated <- enrichGO(gene = neuron_hom2het_down_entrez,
                OrgDb =       'org.Mm.eg.db',
                ont           = "BP",
                pAdjustMethod = "BH",
                pvalueCutoff  = 0.05,
                qvalueCutoff  = 1, 
                readable      = TRUE)



norad_hom2het <- apply(exprs(dat.filtered)[genotypeCellType_diff_test_sorted_sig_genes,pData(dat.filtered)$genotype == "hom" & pData(dat.filtered)$CellType == "Noradrenergic Neuron"], 1,mean)/apply(exprs(dat.filtered)[genotypeCellType_diff_test_sorted_sig_genes,pData(dat.filtered)$genotype == "het" & pData(dat.filtered)$CellType == "Noradrenergic Neuron"], 1,mean)

norad_hom2het_up <- names(norad_hom2het[norad_hom2het > 1 | norad_hom2het == "Inf"])
norad_hom2het_down <- names(norad_hom2het[norad_hom2het < 1 | norad_hom2het == "Nan"])

norad_hom2het_up_entrez <- gsub("\\..*","",norad_hom2het_up)
m<-match(norad_hom2het_up_entrez,egENSEMBL$ensembl_id)
norad_hom2het_up_entrez <- egENSEMBL$gene_id[m]
ego_norad_upregulated <- enrichGO(gene = norad_hom2het_up_entrez,
                OrgDb =       'org.Mm.eg.db',
                ont           = "BP",
                pAdjustMethod = "BH",
                pvalueCutoff  = 0.05,
                qvalueCutoff  = 1, 
                readable      = TRUE)

norad_hom2het_down_entrez <- gsub("\\..*","",norad_hom2het_down)
m<-match(norad_hom2het_down_entrez,egENSEMBL$ensembl_id)
norad_hom2het_down_entrez <- egENSEMBL$gene_id[m]
ego_norad_downregulated <- enrichGO(gene = norad_hom2het_down_entrez,
                OrgDb =       'org.Mm.eg.db',
                ont           = "BP",
                pAdjustMethod = "BH",
                pvalueCutoff  = 0.05,
                qvalueCutoff  = 1, 
                readable      = TRUE)




glia_hom2het <- apply(exprs(dat.filtered)[genotypeCellType_diff_test_sorted_sig_genes,pData(dat.filtered)$genotype == "hom" & pData(dat.filtered)$CellType == "Progenitor/Glia"], 1,mean)/apply(exprs(dat.filtered)[genotypeCellType_diff_test_sorted_sig_genes,pData(dat.filtered)$genotype == "het" & pData(dat.filtered)$CellType == "Progenitor/Glia"], 1,mean)

glia_hom2het_up <- names(glia_hom2het[glia_hom2het > 1 | glia_hom2het == "Inf"])
glia_hom2het_down <- names(glia_hom2het[glia_hom2het < 1 | glia_hom2het == "Nan"])

glia_hom2het_up_entrez <- gsub("\\..*","",glia_hom2het_up)
m<-match(glia_hom2het_up_entrez,egENSEMBL$ensembl_id)
glia_hom2het_up_entrez <- egENSEMBL$gene_id[m]
ego_glia_upregulated <- enrichGO(gene = glia_hom2het_up_entrez,
                OrgDb =       'org.Mm.eg.db',
                ont           = "BP",
                pAdjustMethod = "BH",
                pvalueCutoff  = 0.05,
                qvalueCutoff  = 1, 
                readable      = TRUE)

glia_hom2het_down_entrez <- gsub("\\..*","",glia_hom2het_down)
m<-match(glia_hom2het_down_entrez,egENSEMBL$ensembl_id)
glia_hom2het_down_entrez <- egENSEMBL$gene_id[m]
ego_glia_downregulated <- enrichGO(gene = glia_hom2het_down_entrez,
                OrgDb =       'org.Mm.eg.db',
                ont           = "BP",
                pAdjustMethod = "BH",
                pvalueCutoff  = 0.05,
                qvalueCutoff  = 1, 
                readable      = TRUE)




transition_hom2het <- apply(exprs(dat.filtered)[genotypeCellType_diff_test_sorted_sig_genes,pData(dat.filtered)$genotype == "hom" & pData(dat.filtered)$CellType == "Transition Cell"], 1,mean)/apply(exprs(dat.filtered)[genotypeCellType_diff_test_sorted_sig_genes,pData(dat.filtered)$genotype == "het" & pData(dat.filtered)$CellType == "Transition Cell"], 1,mean)

transition_hom2het_up <- names(transition_hom2het[transition_hom2het > 1 | transition_hom2het == "Inf"])
transition_hom2het_down <- names(transition_hom2het[transition_hom2het < 1 | transition_hom2het == "Nan"])

transition_hom2het_up_entrez <- gsub("\\..*","",transition_hom2het_up)
m<-match(transition_hom2het_up_entrez,egENSEMBL$ensembl_id)
transition_hom2het_up_entrez <- egENSEMBL$gene_id[m]
ego_transition_upregulated <- enrichGO(gene = transition_hom2het_up_entrez,
                OrgDb =       'org.Mm.eg.db',
                ont           = "BP",
                pAdjustMethod = "BH",
                pvalueCutoff  = 0.05,
                qvalueCutoff  = 1, 
                readable      = TRUE)

transition_hom2het_down_entrez <- gsub("\\..*","",transition_hom2het_down)
m<-match(transition_hom2het_down_entrez,egENSEMBL$ensembl_id)
transition_hom2het_down_entrez <- egENSEMBL$gene_id[m]
ego_transition_downregulated <- enrichGO(gene = transition_hom2het_down_entrez,
                OrgDb =       'org.Mm.eg.db',
                ont           = "BP",
                pAdjustMethod = "BH",
                pvalueCutoff  = 0.05,
                qvalueCutoff  = 1, 
                readable      = TRUE)
```

```{r genotype*sex*CellType_diff_test}
genotype_sex_celltype_diff_test <- differentialGeneTest(dat.filtered[expressed_genes,], fullModelFormulaStr = "~batch + CellType*sex*genotype", reducedModelFormulaStr = "~batch + CellType + sex + genotype + genotype*CellType + genotype*sex", cores = detectCores()-1)

genotype_sex_celltype_diff_test_sorted <- na.omit(genotype_sex_celltype_diff_test[order(genotype_sex_celltype_diff_test$qval),])

genotype_sex_celltype_diff_test_sorted_sig_genes <- rownames(genotype_sex_celltype_diff_test_sorted[genotype_sex_celltype_diff_test_sorted$qval < 0.01,])

length(genotype_sex_celltype_diff_test_sorted_sig_genes)

tmp <- t(scale(t(exprs(dat.filtered[genotype_sex_celltype_diff_test_sorted_sig_genes,])), scale = T, center = T))
heatmap_data<-log(tmp -min(tmp) + 1)

myLower <- 0
myUpper <- 2

heatmap_data[heatmap_data < myLower] <- myLower
heatmap_data[heatmap_data > myUpper] <- myUpper

rownames(heatmap_data) <- lookupGeneName(dat.filtered,rownames(heatmap_data))

heatmap_annotation <- pData(dat.filtered)[,c("CellType", "genotype", "age")]

heatmap_colors <- list(
"Cell Type" =  c("Acetylcholinergic Neuron" = "#E41A1C", "Neuron" = "#377EB8", "Noradrenergic Neuron" = "#4DAF4A", "Progenitor/Glia" = "#984EA3", "Transition Cell" = "#FF7F00"),
"Genotype" = c("het" = "black", "hom" = "gray70"),
"Age" = c("E12.5" = "#7FC97F", "E14.5" = "#BEAED4"))

colnames(heatmap_annotation) <- c("Cell Type","Genotype","Age")

pheatmap(mat = heatmap_data,
         scale="none",
         show_rownames=TRUE,
         show_colnames=FALSE,
         drop_levels=FALSE,
         breaks=seq(myLower,myUpper,length=100),
         annotation_col=heatmap_annotation,
         annotation_colors=heatmap_colors,
         annotation_names_col=TRUE,
         annotation_legend=TRUE,
         fontsize_row = 8,
         main = "All Cells\n61 Genes, 1% FDR\nFull model: ~ genotype*CellType*sex + batch\nReduced model: ~ genotype + CellType + sex + genotype*CellType + genotype*sex + batch"
         )

```

```{r CellType_diff_test}
CellType_diff_test <- differentialGeneTest(dat.filtered[expressed_genes,], fullModelFormulaStr = "~batch+CellType", reducedModelFormulaStr = "~batch", cores = detectCores() - 1)

CellType_diff_test_sorted <- CellType_diff_test[order(CellType_diff_test$qval),]

CellType_diff_test_sorted_sig_genes <- rownames(CellType_diff_test_sorted[CellType_diff_test_sorted$qval < 1e-30,])

length(CellType_diff_test_sorted_sig_genes)

tmp <- t(scale(t(exprs(dat.filtered[CellType_diff_test_sorted_sig_genes,])), scale = T, center = T))
heatmap_data <- log(tmp - min(tmp) + 1)

myLower <- 0
myUpper <- 4

heatmap_data[heatmap_data < myLower] <- myLower
heatmap_data[heatmap_data > myUpper] <- myUpper

rownames(heatmap_data) <- lookupGeneName(dat.filtered, rownames(heatmap_data))

heatmap_annotation <- pData(dat.filtered)[,c("CellType","genotype","sex","Cluster","age")]

heatmap_colors <- list(
"Cell Type" =  c("Acetylcholinergic Neuron" = "#E41A1C", "Ambiguous" = "#377EB8", "Neuron" = "#4DAF4A", "Noradrenergic Neuron" = "#984EA3", "Progenitor/Glia" = "#FF7F00", "Unknown" = "#FFFF33", "Acetylcholinergic Neuron" = "#A65628", "Ambiguous" = "#F781BF", "Neuron" = "#999999"),
"Genotype" = c("het" = "#7FC97F", "hom" = "#BEAED4"),
"Sex" = c("female"="black","male"="gray70"),
"Cluster" = c("1" = "#E41A1C", "2" = "#377EB8", "3" = "#4DAF4A", "4" = "#984EA3", "5" = "#FF7F00", "6" = "#FFFF33", "7" = "#A65628", "8" = "#F781BF", "9" = "#999999", "10" = "#E41A1C", "11" = "#377EB8"),
"Age" = c("E12.5" = "#7FC97F", "E14.5" = "#BEAED4"))

colnames(heatmap_annotation) <- c("Cell Type", "Genotype", "Sex", "Cluster", "Age")

pheatmap(mat = heatmap_data,
         scale="none",
         show_rownames=TRUE,
         show_colnames=FALSE,
         drop_levels=FALSE,
         breaks=seq(myLower,myUpper,length=100),
         annotation_col=heatmap_annotation,
         annotation_colors=heatmap_colors,
         annotation_names_col=TRUE,
         annotation_legend=TRUE,
         main = "All Cells\n315 Genes, 1e-30 FDR\nFull model: ~ batch + CellType\nReduced model: ~ batch"
         )
```

```{r neuron_bifurcation_diff_test}
neuron_bifuracation_difftest <- differentialGeneTest(dat.filtered[expressed_genes,pData(dat.filtered)$cluster %in% c(4, 5)], fullModelFormulaStr = "~batch + cluster", reducedModelFormulaStr = "~batch", cores = detectCores()-1)

neuron_bifuracation_difftest_sorted <- neuron_bifuracation_difftest[order(neuron_bifuracation_difftest$pval),]

neuron_bifuracation_difftest_sorted_sig_genes <- rownames(neuron_bifuracation_difftest_sorted[neuron_bifuracation_difftest_sorted$qval < 0.05/length(expressed_genes),])

bifurcation_gene_names <- lookupGeneName(dat.filtered, neuron_bifuracation_difftest_sorted_sig_genes)

# het-only = 5, mixed pop = 4

pop4_means <- apply(exprs(dat.filtered)[neuron_bifuracation_difftest_sorted_sig_genes, pData(dat.filtered)$cluster == 4], 1,mean)
pop5_means <- apply(exprs(dat.filtered)[neuron_bifuracation_difftest_sorted_sig_genes, pData(dat.filtered)$cluster == 5], 1,mean)

relative_exprs <- pop5_means/pop4_means
# 5 = het 

upregulated_in_het <- names(relative_exprs[relative_exprs > 1 | relative_exprs == "Inf"])

downregulated_in_het <- names(relative_exprs[relative_exprs < 1])

egENSEMBL <- toTable(org.Mm.egENSEMBL)

upregulated_in_het_entrez <- gsub("\\..*","",upregulated_in_het)
m<-match(upregulated_in_het_entrez,egENSEMBL$ensembl_id)
upregulated_in_het_entrez <- egENSEMBL$gene_id[m]
ego_upregulated <- enrichGO(gene = upregulated_in_het_entrez,
                OrgDb =       'org.Mm.eg.db',
                ont           = "BP",
                pAdjustMethod = "BH",
                pvalueCutoff  = 0.05,
                qvalueCutoff  = 1, 
                readable      = TRUE)

downregulated_in_het_entrez <- gsub("\\..*","",downregulated_in_het)
m<-match(downregulated_in_het_entrez,egENSEMBL$ensembl_id)
downregulated_in_het_entrez <- egENSEMBL$gene_id[m]
ego_downregulated <- enrichGO(gene = downregulated_in_het_entrez,
                OrgDb =       'org.Mm.eg.db',
                ont           = "BP",
                pAdjustMethod = "BH",
                pvalueCutoff  = 0.05,
                qvalueCutoff  = 1, 
                readable      = TRUE)


tmp <- t(scale(t(exprs(dat.filtered[neuron_bifuracation_difftest_sorted_sig_genes, pData(dat.filtered)$cluster %in% c(4, 5)])), scale = T, center = T))
heatmap_data <- log10(tmp - min(tmp) + 1)

rownames(heatmap_data) <- lookupGeneName(dat.filtered, rownames(heatmap_data))

heatmap_data <- heatmap_data[order.dendrogram(as.dendrogram(hclust(dist(heatmap_data)))),order.dendrogram(as.dendrogram(hclust(dist(t(heatmap_data)))))]

#myLower <- 0.1
#myUpper <- 0.9

#heatmap_data[heatmap_data < myLower] <- myLower
#heatmap_data[heatmap_data > myUpper] <- myUpper

heatmap_annotation <- pData(dat.filtered)[, c("cluster", "genotype", "age")]

heatmap_colors <- list(
  "Cluster" = c("4" = "#984EA3",
                "5" = "#FF7F00"),
  "Genotype" = c("het" = "#E41A1C", "hom" = "#377EB8"),
  "Age" = c("E12.5" = "black", "E14.5" = "gray"))

colnames(heatmap_annotation)<-c("Cluster","Genotype","Age")

#pdf("Heatmap_all_cells_wrt_genotype.pdf",height=25,width=18)
#png("Heatmap_all_cells_wrt_genotype.png",height=15,width=20, units = "in", res = 400)
pheatmap(mat = heatmap_data,
         scale="none",
         show_rownames = TRUE,
         show_colnames = FALSE,
         drop_levels = FALSE,
         cluster_rows = F,
         cluster_cols = F,
         color = colorRampPalette(colors = c("#377EB8", "white", "#E41A1C"))(100),
         #breaks=seq(myLower, myUpper, length = 100),
         #breaks = seq(0, 2.8, length = 200),
         annotation_col = heatmap_annotation,
         annotation_colors = heatmap_colors,
         annotation_names_col = TRUE,
         annotation_legend = TRUE,
         fontsize_row = 5#,
         #treeheight_col = 0,
         #treeheight_row = 0,
         #main = "All Cells\n254 Genes, 1% FDR\nFull model: ~ genotype + batch  + sex + cell type\nReduced model: ~ batch + sex + cell type"
         )
#dev.off()

```

```{r Ret_independent_genotype_diff}
Ret_independent_difftest <- differentialGeneTest(dat.filtered[expressed_genes,pData(dat.filtered)$cluster %in% c(4, 6)], fullModelFormulaStr = "~batch + genotype", reducedModelFormulaStr = "~batch", cores = detectCores()-1)

Ret_independent_difftest_sorted <- Ret_independent_difftest[order(Ret_independent_difftest$pval),]

Ret_independent_difftest_sorted_sig_genes <- rownames(Ret_independent_difftest_sorted[Ret_independent_difftest_sorted$qval < 0.05/length(expressed_genes),])

Ret_independent_gene_names <- lookupGeneName(dat.filtered, Ret_independent_difftest_sorted_sig_genes)



tmp <- t(scale(t(exprs(dat.filtered[Ret_independent_difftest_sorted_sig_genes, pData(dat.filtered)$cluster %in% c(4, 6)])), scale = T, center = T))
heatmap_data <- log10(tmp - min(tmp) + 1)

rownames(heatmap_data) <- lookupGeneName(dat.filtered, rownames(heatmap_data))

heatmap_data <- heatmap_data[order.dendrogram(as.dendrogram(hclust(dist(heatmap_data)))),order.dendrogram(as.dendrogram(hclust(dist(t(heatmap_data)))))]

#myLower <- 0.1
#myUpper <- 0.9

#heatmap_data[heatmap_data < myLower] <- myLower
#heatmap_data[heatmap_data > myUpper] <- myUpper

heatmap_annotation <- pData(dat.filtered)[, c("cluster", "genotype")]

heatmap_colors <- list(
  "Cluster" = c("4" = "#984EA3",
                "6" = "#FFFF33"),
  "Genotype" = c("het" = "#E41A1C", "hom" = "#377EB8")#,
#  "Age" = c("E12.5" = "black", "E14.5" = "gray")
)

#colnames(heatmap_annotation)<-c("Cluster","Genotype","Age")
colnames(heatmap_annotation)<-c("Cluster","Genotype")

#pdf("Heatmap_all_cells_wrt_genotype.pdf",height=25,width=18)
#png("Heatmap_all_cells_wrt_genotype.png",height=15,width=20, units = "in", res = 400)
pheatmap(mat = heatmap_data,
         scale="none",
         show_rownames = TRUE,
         show_colnames = FALSE,
         drop_levels = FALSE,
         cluster_rows = F,
         cluster_cols = F,
         color = colorRampPalette(colors = c("#377EB8", "white", "#E41A1C"))(100),
         #breaks=seq(myLower, myUpper, length = 100),
         #breaks = seq(0, 2.8, length = 200),
         annotation_col = heatmap_annotation,
         annotation_colors = heatmap_colors,
         annotation_names_col = TRUE,
         annotation_legend = TRUE,
         fontsize = 7,
         fontsize_row = 6#,
         #treeheight_col = 0,
         #treeheight_row = 0,
         #main = "All Cells\n254 Genes, 1% FDR\nFull model: ~ genotype + batch  + sex + cell type\nReduced model: ~ batch + sex + cell type"
         )
#dev.off()

```


# Cluster Specificity
```{r specificity}
# Look at both cluster and state specificity (should see a fair amount of overlap)

melted_df <- melt(exprs(dat.filtered[expressed_genes,]))
colnames(melted_df) <- c("gene_id","cell_id","value")
melted_df <- merge(melted_df, pData(dat.filtered)[, c("genotype", "sex", "age", "cluster")], by.x = "cell_id", by.y = 0)

mean_expr_cluster <- melted_df %>%
  group_by(cluster, gene_id) %>%
  summarise(mean_expression_cluster = mean(value)) %>%
  as.data.frame()

mean_expr_cluster <- dcast(mean_expr_cluster, gene_id ~ cluster)
rownames(mean_expr_cluster) <- mean_expr_cluster$gene_id
mean_expr_cluster <- mean_expr_cluster[,-1]

#Data frame of mean expression for each gene in each cluster

cluster_specificity <- cummeRbund:::.specificity(mean_expr_cluster)

melted_cluster_specificity <- melt(cluster_specificity)
mean_expr_cluster_melted <- melt(mean_expr_cluster)
names(mean_expr_cluster_melted) <- c("cluster","mean_expr")
mean_expr_cluster_melted$gene_id <- rep(row.names(mean_expr_cluster), nlevels(pData(dat.filtered)$cluster))
melted_cluster_specificity$Var2 <- as.numeric(str_replace(melted_cluster_specificity$Var2, pattern = "_spec", replacement = ""))
names(melted_cluster_specificity) <- c("gene_id", "cluster", "specificity")
melted_df <- merge(melted_cluster_specificity, mean_expr_cluster_melted, by = c("gene_id", "cluster"))
melted_df$gene_short_name <- lookupGeneName(dat.filtered, as.character(melted_df$gene_id), unique = F)

#cluster 2 is mixed cluster
#Do GO enrichment for genes specific to cluster 6
#What are the genes specific to clusters 9 and 10?

index <- apply(cluster_specificity, 1, which.max)
gene_specificity <- list()
for(i in 1:nlevels(pData(dat.filtered)$cluster)){
  gene_specificity[[i]] <- lookupGeneName(dat.filtered, names(index[index == i]))
}

genes <- list()
genes[[1]] <- c("Igf1", "Aurka", "Aurkb", "Top2a", "Ube2c", "Hist1h2ai", "Hist1h2ap", "Hist1h2ah", "Epcam", "Ccnb1", "Cdk1", "Sox10", "Hist1h2ao", "Ccna2")
genes[[2]] <- c("Sox6", "Notch2", "Mast4")
genes[[3]] <- c("Cdc20", "Cenpa", "Cenpf", "Ccn2", "Hoxa2")
genes[[4]] <- c("Lancl3", "Tac1", "Ecel1", "Chgb", "Mapt", "Stmn2", "Stmn3", "Rtn1", "Meg3", "Gabrg3", "Ache", "Npy", "Myl1", "Snap25")
genes[[5]] <- c("Nos1", "Ntng1", "Nrg1", "Etv1", "Sncg", "Gal", "Ass1", "Tubb3", "Elavl4", "Kcnab1")
genes[[6]] <- c("H2-Ab1", "Irgm1", "Arg1", "Cck", "Apoe", "Pantr1", "Grp", "St18", "Dll1", "Dll3")
genes[[7]] <- c("Tagln", "Myl9", "Postn", "Plp1", "Gas1", "Mest", "Ednrb", "Fabp7")
genes[[8]] <- c("Hoxc10", "Hoxa10", "Hoxa7", "Hoxc9", "Hoxa9", "Hand1", "Gata2", "Gfra3", "Agtr2", "Cartpt", "Th", "Dbh", "Nefl", "Isl1", "Pmp22", "Slc18a3")


for(i in 1:nlevels(pData(dat.filtered)$cluster)){
  print(paste("Cluster", i))
  print(genes[[i]][!genes[[i]] %in% gene_specificity[[i]]])
}


plot_specificity <- function(x){
  ggplot(melted_df[melted_df$cluster == x,], aes(x = log10(mean_expr + 1), 
                                        y = specificity, 
                                        label = ifelse(melted_df[melted_df$cluster == x,]$gene_short_name %in% genes[[x]],
                                                       melted_df[melted_df$cluster == x,]$gene_short_name, 
                                                       ""), 
                                        color = ifelse(ifelse(melted_df[melted_df$cluster == x,]$gene_short_name %in% genes[[x]],
                                                       melted_df[melted_df$cluster == x,]$gene_short_name, 
                                                       "") == "", "no", "yes"), 
                                        alpha = ifelse(ifelse(melted_df[melted_df$cluster == x,]$gene_short_name %in% genes[[x]],
                                                       melted_df[melted_df$cluster == x,]$gene_short_name, 
                                                       "") == "", 0.1, 1))) +
#  ggtitle(paste("Cluster", x)) +
      geom_point(size = 0.5) +
      geom_smooth(lwd = 0.2, color = "red") +
      scale_color_manual(values = c("no" = "black", "yes" = "red")) +
      scale_x_continuous("log10(mean expression + 1)", expand = c(0,0), limits = c(0, 3.5)) +
      scale_y_continuous("Specificity", expand = c(0,0), limits = c(0,1)) +
      geom_text(size = 1.4, 
                hjust = 0, 
                vjust = 0, 
                nudge_x = 0.03, 
                nudge_y = -0.005, 
                color = "black") +
      theme(legend.position = "none",
            plot.title = element_text(size = 10, hjust = 0.005, vjust = -6, face = "plain"),
            plot.margin = margin(-10, 2, 2, 2),
            axis.title = element_text(size= 8),
            axis.title.x = element_text(margin = margin(t = 0)),
            axis.title.y = element_text(margin = margin(r = 0)),
            axis.text = element_text(size = 6),
            #axis.title = element_blank(),
            #axis.text = element_blank(),
            aspect.ratio = 1
            )
}


plot_specificity(4) + ggtitle("Cluster 4: Ache+")
plot_specificity(5) + ggtitle("Cluster 5: Nos1+")
plot_specificity(7) + ggtitle("Cluster 7: Maturing Glia")
plot_specificity(8) + ggtitle("Cluster 8: Th+, Dbh+")
p2 <- plot_specificity(2)
p3 <- plot_specificity(3)
p4 <- plot_specificity(4)
p5 <- plot_specificity(5)
p6 <- plot_specificity(6)
p7 <- plot_specificity(7)
p8 <- plot_specificity(8)

pg1 <- cowplot::plot_grid(p1, p3, p5, p7, rel_widths = 1, ncol = 1)
pg2 <- cowplot::plot_grid(p2, p4, p6, p8, rel_widths = 1, ncol = 1)

cowplot::plot_grid(pg1, pg2, ncol = 2)      

grid.arrange(p1, p2, p3, p4, p5, p6, p7, p8, ncol = 2)

```